

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>API Documentation &mdash; MHCflurry 1.1.0 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="search" title="Search" href="search.html"/>
    <link rel="top" title="MHCflurry 1.1.0 documentation" href="index.html"/>
        <link rel="prev" title="Command-line reference" href="commandline_tools.html"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> MHCflurry
          

          
          </a>

          
            
            
              <div class="version">
                1.1.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="intro.html">Introduction and setup</a></li>
<li class="toctree-l1"><a class="reference internal" href="commandline_tutorial.html">Command-line tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="python_tutorial.html">Python library tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="models_supported_alleles.html">Supported alleles and peptide lengths</a></li>
<li class="toctree-l1"><a class="reference internal" href="models.html">Details on the released models</a></li>
<li class="toctree-l1"><a class="reference internal" href="commandline_tools.html">Command-line reference</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">API Documentation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.allele_encoding">mhcflurry.allele_encoding module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.amino_acid">mhcflurry.amino_acid module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.class1_affinity_predictor">mhcflurry.class1_affinity_predictor module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.class1_neural_network">mhcflurry.class1_neural_network module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.common">mhcflurry.common module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.custom_loss">mhcflurry.custom_loss module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.downloads">mhcflurry.downloads module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.downloads_command">mhcflurry.downloads_command module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.encodable_sequences">mhcflurry.encodable_sequences module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.ensemble_centrality">mhcflurry.ensemble_centrality module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.hyperparameters">mhcflurry.hyperparameters module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.percent_rank_transform">mhcflurry.percent_rank_transform module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.predict_command">mhcflurry.predict_command module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.regression_target">mhcflurry.regression_target module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.scoring">mhcflurry.scoring module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.train_allele_specific_models_command">mhcflurry.train_allele_specific_models_command module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.version">mhcflurry.version module</a></li>
</ul>
</li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">MHCflurry</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>API Documentation</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/api.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-mhcflurry">
<span id="id1"></span><span id="api-documentation"></span><h1>API Documentation<a class="headerlink" href="#module-mhcflurry" title="Permalink to this headline">¶</a></h1>
<dl class="class">
<dt id="mhcflurry.Class1AffinityPredictor">
<em class="property">class </em><code class="descclassname">mhcflurry.</code><code class="descname">Class1AffinityPredictor</code><span class="sig-paren">(</span><em>allele_to_allele_specific_models=None</em>, <em>class1_pan_allele_models=None</em>, <em>allele_to_fixed_length_sequence=None</em>, <em>manifest_df=None</em>, <em>allele_to_percent_rank_transform=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1AffinityPredictor" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<p>High-level interface for peptide/MHC I binding affinity prediction.</p>
<p>This class manages low-level <a class="reference internal" href="#mhcflurry.Class1NeuralNetwork" title="mhcflurry.Class1NeuralNetwork"><code class="xref py py-obj docutils literal"><span class="pre">Class1NeuralNetwork</span></code></a> instances, each of which
wraps a single Keras network. The purpose of <a class="reference internal" href="#mhcflurry.Class1AffinityPredictor" title="mhcflurry.Class1AffinityPredictor"><code class="xref py py-obj docutils literal"><span class="pre">Class1AffinityPredictor</span></code></a> is to
implement ensembles, handling of multiple alleles, and predictor loading and
saving.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>allele_to_allele_specific_models</strong> : dict of string -&gt; list of <a class="reference internal" href="#mhcflurry.Class1NeuralNetwork" title="mhcflurry.Class1NeuralNetwork"><code class="xref py py-obj docutils literal"><span class="pre">Class1NeuralNetwork</span></code></a></p>
<blockquote>
<div><p>Ensemble of single-allele models to use for each allele.</p>
</div></blockquote>
<p><strong>class1_pan_allele_models</strong> : list of <a class="reference internal" href="#mhcflurry.Class1NeuralNetwork" title="mhcflurry.Class1NeuralNetwork"><code class="xref py py-obj docutils literal"><span class="pre">Class1NeuralNetwork</span></code></a></p>
<blockquote>
<div><p>Ensemble of pan-allele models.</p>
</div></blockquote>
<p><strong>allele_to_fixed_length_sequence</strong> : dict of string -&gt; string</p>
<blockquote>
<div><p>Required only if class1_pan_allele_models is specified.</p>
</div></blockquote>
<p><strong>manifest_df</strong> : <code class="xref py py-obj docutils literal"><span class="pre">pandas.DataFrame</span></code>, optional</p>
<blockquote>
<div><p>Must have columns: model_name, allele, config_json, model.
Only required if you want to update an existing serialization of a
Class1AffinityPredictor. Otherwise this dataframe will be generated
automatically based on the supplied models.</p>
</div></blockquote>
<p><strong>allele_to_percent_rank_transform</strong> : dict of string -&gt; <code class="xref py py-obj docutils literal"><span class="pre">PercentRankTransform</span></code>, optional</p>
<blockquote class="last">
<div><p><code class="xref py py-obj docutils literal"><span class="pre">PercentRankTransform</span></code> instances to use for each allele</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="mhcflurry.Class1AffinityPredictor.calibrate_percentile_ranks">
<code class="descname">calibrate_percentile_ranks</code><span class="sig-paren">(</span><em>peptides=None</em>, <em>num_peptides_per_length=100000</em>, <em>alleles=None</em>, <em>bins=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.calibrate_percentile_ranks"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1AffinityPredictor.calibrate_percentile_ranks" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute the cumulative distribution of ic50 values for a set of alleles
over a large universe of random peptides, to enable computing quantiles in
this distribution later.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>peptides</strong> : sequence of string or EncodableSequences, optional</p>
<blockquote>
<div><p>Peptides to use</p>
</div></blockquote>
<p><strong>num_peptides_per_length</strong> : int, optional</p>
<blockquote>
<div><p>If peptides argument is not specified, then num_peptides_per_length
peptides are randomly sampled from a uniform distribution for each
supported length</p>
</div></blockquote>
<p><strong>alleles</strong> : sequence of string, optional</p>
<blockquote>
<div><p>Alleles to perform calibration for. If not specified all supported
alleles will be calibrated.</p>
</div></blockquote>
<p><strong>bins</strong> : object</p>
<blockquote>
<div><p>Anything that can be passed to numpy.histogram’s “bins” argument
can be used here, i.e. either an integer or a sequence giving bin
edges. This is in ic50 space.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last"><strong>EncodableSequences</strong> : peptides used for calibration</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1AffinityPredictor.fit_allele_specific_predictors">
<code class="descname">fit_allele_specific_predictors</code><span class="sig-paren">(</span><em>n_models</em>, <em>architecture_hyperparameters_list</em>, <em>allele</em>, <em>peptides</em>, <em>affinities</em>, <em>inequalities=None</em>, <em>models_dir_for_save=None</em>, <em>verbose=0</em>, <em>progress_preamble=''</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.fit_allele_specific_predictors"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1AffinityPredictor.fit_allele_specific_predictors" title="Permalink to this definition">¶</a></dt>
<dd><p>Fit one or more allele specific predictors for a single allele using a
single neural network architecture.</p>
<p>The new predictors are saved in the Class1AffinityPredictor instance
and will be used on subsequent calls to <a class="reference internal" href="#mhcflurry.Class1AffinityPredictor.predict" title="mhcflurry.Class1AffinityPredictor.predict"><code class="xref py py-obj docutils literal"><span class="pre">predict</span></code></a>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>n_models</strong> : int</p>
<blockquote>
<div><p>Number of neural networks to fit</p>
</div></blockquote>
<p><strong>architecture_hyperparameters_list</strong> : list of dict</p>
<blockquote>
<div><p>List of hyperparameters. If more than one set of hyperparameters
are specified, model selection over them is performed and model
with the lowest validation loss is selected for each fold.</p>
</div></blockquote>
<p><strong>allele</strong> : string</p>
<p><strong>peptides</strong> : <code class="xref py py-obj docutils literal"><span class="pre">EncodableSequences</span></code> or list of string</p>
<p><strong>affinities</strong> : list of float</p>
<blockquote>
<div><p>nM affinities</p>
</div></blockquote>
<p><strong>inequalities</strong> : list of string, each element one of “&gt;”, “&lt;”, or “=”</p>
<blockquote>
<div><p>See Class1NeuralNetwork.fit for details.</p>
</div></blockquote>
<p><strong>models_dir_for_save</strong> : string, optional</p>
<blockquote>
<div><p>If specified, the Class1AffinityPredictor is (incrementally) written
to the given models dir after each neural network is fit.</p>
</div></blockquote>
<p><strong>verbose</strong> : int</p>
<blockquote>
<div><p>Keras verbosity</p>
</div></blockquote>
<p><strong>progress_preamble</strong> : string</p>
<blockquote>
<div><p>Optional string of information to include in each progress update</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">list of <a class="reference internal" href="#mhcflurry.Class1NeuralNetwork" title="mhcflurry.Class1NeuralNetwork"><code class="xref py py-obj docutils literal"><span class="pre">Class1NeuralNetwork</span></code></a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1AffinityPredictor.fit_class1_pan_allele_models">
<code class="descname">fit_class1_pan_allele_models</code><span class="sig-paren">(</span><em>n_models</em>, <em>architecture_hyperparameters</em>, <em>alleles</em>, <em>peptides</em>, <em>affinities</em>, <em>inequalities</em>, <em>models_dir_for_save=None</em>, <em>verbose=1</em>, <em>progress_preamble=''</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.fit_class1_pan_allele_models"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1AffinityPredictor.fit_class1_pan_allele_models" title="Permalink to this definition">¶</a></dt>
<dd><p>Fit one or more pan-allele predictors using a single neural network
architecture.</p>
<p>The new predictors are saved in the Class1AffinityPredictor instance
and will be used on subsequent calls to <a class="reference internal" href="#mhcflurry.Class1AffinityPredictor.predict" title="mhcflurry.Class1AffinityPredictor.predict"><code class="xref py py-obj docutils literal"><span class="pre">predict</span></code></a>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>n_models</strong> : int</p>
<blockquote>
<div><p>Number of neural networks to fit</p>
</div></blockquote>
<p><strong>architecture_hyperparameters</strong> : dict</p>
<p><strong>alleles</strong> : list of string</p>
<blockquote>
<div><p>Allele names (not sequences) corresponding to each peptide</p>
</div></blockquote>
<p><strong>peptides</strong> : <code class="xref py py-obj docutils literal"><span class="pre">EncodableSequences</span></code> or list of string</p>
<p><strong>affinities</strong> : list of float</p>
<blockquote>
<div><p>nM affinities</p>
</div></blockquote>
<p><strong>inequalities</strong> : list of string, each element one of “&gt;”, “&lt;”, or “=”</p>
<blockquote>
<div><p>See Class1NeuralNetwork.fit for details.</p>
</div></blockquote>
<p><strong>models_dir_for_save</strong> : string, optional</p>
<blockquote>
<div><p>If specified, the Class1AffinityPredictor is (incrementally) written
to the given models dir after each neural network is fit.</p>
</div></blockquote>
<p><strong>verbose</strong> : int</p>
<blockquote>
<div><p>Keras verbosity</p>
</div></blockquote>
<p><strong>progress_preamble</strong> : string</p>
<blockquote>
<div><p>Optional string of information to include in each progress update</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">list of <a class="reference internal" href="#mhcflurry.Class1NeuralNetwork" title="mhcflurry.Class1NeuralNetwork"><code class="xref py py-obj docutils literal"><span class="pre">Class1NeuralNetwork</span></code></a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="staticmethod">
<dt id="mhcflurry.Class1AffinityPredictor.load">
<em class="property">static </em><code class="descname">load</code><span class="sig-paren">(</span><em>models_dir=None</em>, <em>max_models=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.load"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1AffinityPredictor.load" title="Permalink to this definition">¶</a></dt>
<dd><p>Deserialize a predictor from a directory on disk.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>models_dir</strong> : string</p>
<blockquote>
<div><p>Path to directory</p>
</div></blockquote>
<p><strong>max_models</strong> : int, optional</p>
<blockquote>
<div><p>Maximum number of <a class="reference internal" href="#mhcflurry.Class1NeuralNetwork" title="mhcflurry.Class1NeuralNetwork"><code class="xref py py-obj docutils literal"><span class="pre">Class1NeuralNetwork</span></code></a> instances to load</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last"><a class="reference internal" href="#mhcflurry.Class1AffinityPredictor" title="mhcflurry.Class1AffinityPredictor"><code class="xref py py-obj docutils literal"><span class="pre">Class1AffinityPredictor</span></code></a> instance</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="staticmethod">
<dt id="mhcflurry.Class1AffinityPredictor.load_weights">
<em class="property">static </em><code class="descname">load_weights</code><span class="sig-paren">(</span><em>filename</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.load_weights"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1AffinityPredictor.load_weights" title="Permalink to this definition">¶</a></dt>
<dd><p>Restore model weights from the given filename, which should have been
created with <a class="reference internal" href="#mhcflurry.Class1AffinityPredictor.save_weights" title="mhcflurry.Class1AffinityPredictor.save_weights"><code class="xref py py-obj docutils literal"><span class="pre">save_weights</span></code></a>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>filename</strong> : string</p>
<blockquote>
<div><p>Should end in “.npz”.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">list of array</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="classmethod">
<dt id="mhcflurry.Class1AffinityPredictor.merge">
<em class="property">classmethod </em><code class="descname">merge</code><span class="sig-paren">(</span><em>predictors</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.merge"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1AffinityPredictor.merge" title="Permalink to this definition">¶</a></dt>
<dd><p>Merge the ensembles of two or more <a class="reference internal" href="#mhcflurry.Class1AffinityPredictor" title="mhcflurry.Class1AffinityPredictor"><code class="xref py py-obj docutils literal"><span class="pre">Class1AffinityPredictor</span></code></a> instances.</p>
<p>Note: the resulting merged predictor will NOT have calibrated percentile
ranks. Call <a class="reference internal" href="#mhcflurry.Class1AffinityPredictor.calibrate_percentile_ranks" title="mhcflurry.Class1AffinityPredictor.calibrate_percentile_ranks"><code class="xref py py-obj docutils literal"><span class="pre">calibrate_percentile_ranks</span></code></a> on it if these are needed.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>predictors</strong> : sequence of <a class="reference internal" href="#mhcflurry.Class1AffinityPredictor" title="mhcflurry.Class1AffinityPredictor"><code class="xref py py-obj docutils literal"><span class="pre">Class1AffinityPredictor</span></code></a></td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><a class="reference internal" href="#mhcflurry.Class1AffinityPredictor" title="mhcflurry.Class1AffinityPredictor"><code class="xref py py-obj docutils literal"><span class="pre">Class1AffinityPredictor</span></code></a> instance</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="staticmethod">
<dt id="mhcflurry.Class1AffinityPredictor.model_name">
<em class="property">static </em><code class="descname">model_name</code><span class="sig-paren">(</span><em>allele</em>, <em>num</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.model_name"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1AffinityPredictor.model_name" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate a model name</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>allele</strong> : string</p>
<p><strong>num</strong> : int</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">string</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.Class1AffinityPredictor.neural_networks">
<code class="descname">neural_networks</code><a class="headerlink" href="#mhcflurry.Class1AffinityPredictor.neural_networks" title="Permalink to this definition">¶</a></dt>
<dd><p>List of the neural networks in the ensemble.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">list of <a class="reference internal" href="#mhcflurry.Class1NeuralNetwork" title="mhcflurry.Class1NeuralNetwork"><code class="xref py py-obj docutils literal"><span class="pre">Class1NeuralNetwork</span></code></a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1AffinityPredictor.percentile_ranks">
<code class="descname">percentile_ranks</code><span class="sig-paren">(</span><em>affinities</em>, <em>allele=None</em>, <em>alleles=None</em>, <em>throw=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.percentile_ranks"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1AffinityPredictor.percentile_ranks" title="Permalink to this definition">¶</a></dt>
<dd><p>Return percentile ranks for the given ic50 affinities and alleles.</p>
<p>The ‘allele’ and ‘alleles’ argument are as in the <a class="reference internal" href="#mhcflurry.Class1AffinityPredictor.predict" title="mhcflurry.Class1AffinityPredictor.predict"><code class="xref py py-obj docutils literal"><span class="pre">predict</span></code></a> method.
Specify one of these.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>affinities</strong> : sequence of float</p>
<blockquote>
<div><p>nM affinities</p>
</div></blockquote>
<p><strong>allele</strong> : string</p>
<p><strong>alleles</strong> : sequence of string</p>
<p><strong>throw</strong> : boolean</p>
<blockquote>
<div><p>If True, a ValueError will be raised in the case of unsupported
alleles. If False, a warning will be logged and NaN will be returned
for those percentile ranks.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">numpy.array of float</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1AffinityPredictor.predict">
<code class="descname">predict</code><span class="sig-paren">(</span><em>peptides</em>, <em>alleles=None</em>, <em>allele=None</em>, <em>throw=True</em>, <em>centrality_measure='robust_mean'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.predict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1AffinityPredictor.predict" title="Permalink to this definition">¶</a></dt>
<dd><p>Predict nM binding affinities.</p>
<p>If multiple predictors are available for an allele, the predictions are
the geometric means of the individual model predictions.</p>
<p>One of ‘allele’ or ‘alleles’ must be specified. If ‘allele’ is specified
all predictions will be for the given allele. If ‘alleles’ is specified
it must be the same length as ‘peptides’ and give the allele
corresponding to each peptide.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>peptides</strong> : <code class="xref py py-obj docutils literal"><span class="pre">EncodableSequences</span></code> or list of string</p>
<p><strong>alleles</strong> : list of string</p>
<p><strong>allele</strong> : string</p>
<p><strong>throw</strong> : boolean</p>
<blockquote>
<div><p>If True, a ValueError will be raised in the case of unsupported
alleles or peptide lengths. If False, a warning will be logged and
the predictions for the unsupported alleles or peptides will be NaN.</p>
</div></blockquote>
<p><strong>centrality_measure</strong> : string or callable</p>
<blockquote>
<div><p>Measure of central tendency to use to combine predictions in the
ensemble. Options include: mean, median, robust_mean.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">numpy.array of predictions</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1AffinityPredictor.predict_to_dataframe">
<code class="descname">predict_to_dataframe</code><span class="sig-paren">(</span><em>peptides</em>, <em>alleles=None</em>, <em>allele=None</em>, <em>throw=True</em>, <em>include_individual_model_predictions=False</em>, <em>include_percentile_ranks=True</em>, <em>centrality_measure='robust_mean'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.predict_to_dataframe"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1AffinityPredictor.predict_to_dataframe" title="Permalink to this definition">¶</a></dt>
<dd><p>Predict nM binding affinities. Gives more detailed output than <a class="reference internal" href="#mhcflurry.Class1AffinityPredictor.predict" title="mhcflurry.Class1AffinityPredictor.predict"><code class="xref py py-obj docutils literal"><span class="pre">predict</span></code></a>
method, including 5-95% prediction intervals.</p>
<p>If multiple predictors are available for an allele, the predictions are
the geometric means of the individual model predictions.</p>
<p>One of ‘allele’ or ‘alleles’ must be specified. If ‘allele’ is specified
all predictions will be for the given allele. If ‘alleles’ is specified
it must be the same length as ‘peptides’ and give the allele
corresponding to each peptide.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>peptides</strong> : <code class="xref py py-obj docutils literal"><span class="pre">EncodableSequences</span></code> or list of string</p>
<p><strong>alleles</strong> : list of string</p>
<p><strong>allele</strong> : string</p>
<p><strong>throw</strong> : boolean</p>
<blockquote>
<div><p>If True, a ValueError will be raised in the case of unsupported
alleles or peptide lengths. If False, a warning will be logged and
the predictions for the unsupported alleles or peptides will be NaN.</p>
</div></blockquote>
<p><strong>include_individual_model_predictions</strong> : boolean</p>
<blockquote>
<div><p>If True, the predictions of each individual model are included as
columns in the result dataframe.</p>
</div></blockquote>
<p><strong>include_percentile_ranks</strong> : boolean, default True</p>
<blockquote>
<div><p>If True, a “prediction_percentile” column will be included giving the
percentile ranks. If no percentile rank information is available,
this will be ignored with a warning.</p>
</div></blockquote>
<p><strong>centrality_measure</strong> : string or callable</p>
<blockquote>
<div><p>Measure of central tendency to use to combine predictions in the
ensemble. Options include: mean, median, robust_mean.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last"><code class="xref py py-obj docutils literal"><span class="pre">pandas.DataFrame</span></code> of predictions</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1AffinityPredictor.save">
<code class="descname">save</code><span class="sig-paren">(</span><em>models_dir</em>, <em>model_names_to_write=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.save"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1AffinityPredictor.save" title="Permalink to this definition">¶</a></dt>
<dd><p>Serialize the predictor to a directory on disk. If the directory does
not exist it will be created.</p>
<p>The serialization format consists of a file called “manifest.csv” with
the configurations of each Class1NeuralNetwork, along with per-network
files giving the model weights. If there are pan-allele predictors in
the ensemble, the allele sequences are also stored in the
directory. There is also a small file “index.txt” with basic metadata:
when the models were trained, by whom, on what host.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>models_dir</strong> : string</p>
<blockquote>
<div><p>Path to directory</p>
</div></blockquote>
<p><strong>model_names_to_write</strong> : list of string, optional</p>
<blockquote class="last">
<div><p>Only write the weights for the specified models. Useful for
incremental updates during training.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="staticmethod">
<dt id="mhcflurry.Class1AffinityPredictor.save_weights">
<em class="property">static </em><code class="descname">save_weights</code><span class="sig-paren">(</span><em>weights_list</em>, <em>filename</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.save_weights"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1AffinityPredictor.save_weights" title="Permalink to this definition">¶</a></dt>
<dd><p>Save the model weights to the given filename using numpy’s “.npz”
format.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>weights_list</strong> : list of array</p>
<p><strong>filename</strong> : string</p>
<blockquote class="last">
<div><p>Should end in “.npz”.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.Class1AffinityPredictor.supported_alleles">
<code class="descname">supported_alleles</code><a class="headerlink" href="#mhcflurry.Class1AffinityPredictor.supported_alleles" title="Permalink to this definition">¶</a></dt>
<dd><p>Alleles for which predictions can be made.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">list of string</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.Class1AffinityPredictor.supported_peptide_lengths">
<code class="descname">supported_peptide_lengths</code><a class="headerlink" href="#mhcflurry.Class1AffinityPredictor.supported_peptide_lengths" title="Permalink to this definition">¶</a></dt>
<dd><p>(minimum, maximum) lengths of peptides supported by <em>all models</em>,
inclusive.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">(int, int) tuple</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="staticmethod">
<dt id="mhcflurry.Class1AffinityPredictor.weights_path">
<em class="property">static </em><code class="descname">weights_path</code><span class="sig-paren">(</span><em>models_dir</em>, <em>model_name</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.weights_path"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1AffinityPredictor.weights_path" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate the path to the weights file for a model</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>models_dir</strong> : string</p>
<p><strong>model_name</strong> : string</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">string</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="mhcflurry.Class1NeuralNetwork">
<em class="property">class </em><code class="descclassname">mhcflurry.</code><code class="descname">Class1NeuralNetwork</code><span class="sig-paren">(</span><em>**hyperparameters</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<p>Low level class I predictor consisting of a single neural network.</p>
<p>Both single allele and pan-allele prediction are supported, but pan-allele
is in development and not yet well performing.</p>
<p>Users will generally use Class1AffinityPredictor, which gives a higher-level
interface and supports ensembles.</p>
<dl class="attribute">
<dt id="mhcflurry.Class1NeuralNetwork.KERAS_MODELS_CACHE">
<code class="descname">KERAS_MODELS_CACHE</code><em class="property"> = {}</em><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.KERAS_MODELS_CACHE" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1NeuralNetwork.allele_encoding_to_network_input">
<code class="descname">allele_encoding_to_network_input</code><span class="sig-paren">(</span><em>allele_encoding</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.allele_encoding_to_network_input"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.allele_encoding_to_network_input" title="Permalink to this definition">¶</a></dt>
<dd><p>Encode alleles to the fixed-length encoding expected by the neural
network (which depends on the architecture).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>allele_encoding</strong> : AlleleEncoding</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">numpy.array</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="classmethod">
<dt id="mhcflurry.Class1NeuralNetwork.apply_hyperparameter_renames">
<em class="property">classmethod </em><code class="descname">apply_hyperparameter_renames</code><span class="sig-paren">(</span><em>hyperparameters</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.apply_hyperparameter_renames"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.apply_hyperparameter_renames" title="Permalink to this definition">¶</a></dt>
<dd><p>Handle hyperparameter renames.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>hyperparameters</strong> : dict</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><strong>dict</strong> : updated hyperparameters</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="classmethod">
<dt id="mhcflurry.Class1NeuralNetwork.borrow_cached_network">
<em class="property">classmethod </em><code class="descname">borrow_cached_network</code><span class="sig-paren">(</span><em>klass</em>, <em>network_json</em>, <em>network_weights</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.borrow_cached_network"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.borrow_cached_network" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a keras Model with the specified architecture and weights.
As an optimization, when possible this will reuse architectures from a
process-wide cache.</p>
<p>The returned object is “borrowed” in the sense that its weights can
change later after subsequent calls to this method from other objects.</p>
<p>If you’re using this from a parallel implementation you’ll need to
hold a lock while using the returned object.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>network_json</strong> : string of JSON</p>
<p><strong>network_weights</strong> : list of numpy.array</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">keras.models.Model</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="classmethod">
<dt id="mhcflurry.Class1NeuralNetwork.clear_model_cache">
<em class="property">classmethod </em><code class="descname">clear_model_cache</code><span class="sig-paren">(</span><em>klass</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.clear_model_cache"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.clear_model_cache" title="Permalink to this definition">¶</a></dt>
<dd><p>Clear the Keras model cache.</p>
</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.Class1NeuralNetwork.compile_hyperparameter_defaults">
<code class="descname">compile_hyperparameter_defaults</code><em class="property"> = &lt;mhcflurry.hyperparameters.HyperparameterDefaults object&gt;</em><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.compile_hyperparameter_defaults" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="mhcflurry.Class1NeuralNetwork.early_stopping_hyperparameter_defaults">
<code class="descname">early_stopping_hyperparameter_defaults</code><em class="property"> = &lt;mhcflurry.hyperparameters.HyperparameterDefaults object&gt;</em><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.early_stopping_hyperparameter_defaults" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1NeuralNetwork.fit">
<code class="descname">fit</code><span class="sig-paren">(</span><em>peptides</em>, <em>affinities</em>, <em>allele_encoding=None</em>, <em>inequalities=None</em>, <em>sample_weights=None</em>, <em>shuffle_permutation=None</em>, <em>verbose=1</em>, <em>progress_preamble=''</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.fit"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.fit" title="Permalink to this definition">¶</a></dt>
<dd><p>Fit the neural network.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>peptides</strong> : EncodableSequences or list of string</p>
<p><strong>affinities</strong> : list of float</p>
<blockquote>
<div><p>nM affinities. Must be same length of as peptides.</p>
</div></blockquote>
<p><strong>allele_encoding</strong> : AlleleEncoding, optional</p>
<blockquote>
<div><p>If not specified, the model will be a single-allele predictor.</p>
</div></blockquote>
<p><strong>inequalities</strong> : list of string, each element one of “&gt;”, “&lt;”, or “=”.</p>
<blockquote>
<div><p>Inequalities to use for fitting. Same length as affinities.
Each element must be one of “&gt;”, “&lt;”, or “=”. For example, a “&gt;”
will train on y_pred &gt; y_true for that element in the training set.
Requires using a custom losses that support inequalities (e.g.
mse_with_ineqalities).
If None all inequalities are taken to be “=”.</p>
</div></blockquote>
<p><strong>sample_weights</strong> : list of float, optional</p>
<blockquote>
<div><p>If not specified, all samples (including random negatives added
during training) will have equal weight. If specified, the random
negatives will be assigned weight=1.0.</p>
</div></blockquote>
<p><strong>shuffle_permutation</strong> : list of int, optional</p>
<blockquote>
<div><p>Permutation (integer list) of same length as peptides and affinities
If None, then a random permutation will be generated.</p>
</div></blockquote>
<p><strong>verbose</strong> : int</p>
<blockquote>
<div><p>Keras verbosity level</p>
</div></blockquote>
<p><strong>progress_preamble</strong> : string</p>
<blockquote class="last">
<div><p>Optional string of information to include in each progress update</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.Class1NeuralNetwork.fit_hyperparameter_defaults">
<code class="descname">fit_hyperparameter_defaults</code><em class="property"> = &lt;mhcflurry.hyperparameters.HyperparameterDefaults object&gt;</em><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.fit_hyperparameter_defaults" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="classmethod">
<dt id="mhcflurry.Class1NeuralNetwork.from_config">
<em class="property">classmethod </em><code class="descname">from_config</code><span class="sig-paren">(</span><em>config</em>, <em>weights=None</em>, <em>weights_loader=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.from_config"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.from_config" title="Permalink to this definition">¶</a></dt>
<dd><p>deserialize from a dict returned by get_config().</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>config</strong> : dict</p>
<p><strong>weights</strong> : list of array, optional</p>
<blockquote>
<div><p>Network weights to restore</p>
</div></blockquote>
<p><strong>weights_loader</strong> : callable, optional</p>
<blockquote>
<div><p>Function to call (no arguments) to load weights when needed</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">Class1NeuralNetwork</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1NeuralNetwork.get_config">
<code class="descname">get_config</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.get_config"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.get_config" title="Permalink to this definition">¶</a></dt>
<dd><p>serialize to a dict all attributes except model weights</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">dict</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1NeuralNetwork.get_weights">
<code class="descname">get_weights</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.get_weights"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.get_weights" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the network weights</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">list of numpy.array giving weights for each layer</p>
<p class="last">or None if there is no network</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.Class1NeuralNetwork.hyperparameter_defaults">
<code class="descname">hyperparameter_defaults</code><em class="property"> = &lt;mhcflurry.hyperparameters.HyperparameterDefaults object&gt;</em><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.hyperparameter_defaults" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="mhcflurry.Class1NeuralNetwork.hyperparameter_renames">
<code class="descname">hyperparameter_renames</code><em class="property"> = {'use_embedding': None, 'pseudosequence_use_embedding': None}</em><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.hyperparameter_renames" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="mhcflurry.Class1NeuralNetwork.input_encoding_hyperparameter_defaults">
<code class="descname">input_encoding_hyperparameter_defaults</code><em class="property"> = &lt;mhcflurry.hyperparameters.HyperparameterDefaults object&gt;</em><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.input_encoding_hyperparameter_defaults" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1NeuralNetwork.load_weights">
<code class="descname">load_weights</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.load_weights"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.load_weights" title="Permalink to this definition">¶</a></dt>
<dd><p>Load weights by evaluating self.network_weights_loader, if needed.</p>
<p>After calling this, self.network_weights_loader will be None and
self.network_weights will be the weights list, if available.</p>
</dd></dl>

<dl class="staticmethod">
<dt id="mhcflurry.Class1NeuralNetwork.make_network">
<em class="property">static </em><code class="descname">make_network</code><span class="sig-paren">(</span><em>allele_encoding_dims</em>, <em>kmer_size</em>, <em>peptide_amino_acid_encoding</em>, <em>embedding_input_dim</em>, <em>embedding_output_dim</em>, <em>allele_dense_layer_sizes</em>, <em>peptide_dense_layer_sizes</em>, <em>peptide_allele_merge_method</em>, <em>peptide_allele_merge_activation</em>, <em>layer_sizes</em>, <em>dense_layer_l1_regularization</em>, <em>dense_layer_l2_regularization</em>, <em>activation</em>, <em>init</em>, <em>output_activation</em>, <em>dropout_probability</em>, <em>batch_normalization</em>, <em>embedding_init_method</em>, <em>locally_connected_layers</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.make_network"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.make_network" title="Permalink to this definition">¶</a></dt>
<dd><p>Helper function to make a keras network for class1 affinity prediction.</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1NeuralNetwork.network">
<code class="descname">network</code><span class="sig-paren">(</span><em>borrow=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.network"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.network" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the keras model associated with this predictor.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>borrow</strong> : bool</p>
<blockquote>
<div><p>Whether to return a cached model if possible. See
borrow_cached_network for details</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">keras.models.Model</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.Class1NeuralNetwork.network_hyperparameter_defaults">
<code class="descname">network_hyperparameter_defaults</code><em class="property"> = &lt;mhcflurry.hyperparameters.HyperparameterDefaults object&gt;</em><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.network_hyperparameter_defaults" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1NeuralNetwork.peptides_to_network_input">
<code class="descname">peptides_to_network_input</code><span class="sig-paren">(</span><em>peptides</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.peptides_to_network_input"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.peptides_to_network_input" title="Permalink to this definition">¶</a></dt>
<dd><p>Encode peptides to the fixed-length encoding expected by the neural
network (which depends on the architecture).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>peptides</strong> : EncodableSequences or list of string</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">numpy.array</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1NeuralNetwork.predict">
<code class="descname">predict</code><span class="sig-paren">(</span><em>peptides</em>, <em>allele_encoding=None</em>, <em>batch_size=4096</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.predict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.predict" title="Permalink to this definition">¶</a></dt>
<dd><p>Predict affinities</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>peptides</strong> : EncodableSequences or list of string</p>
<p><strong>allele_pseudosequences</strong> : AlleleEncoding, optional</p>
<blockquote>
<div><p>Only required when this model is a pan-allele model</p>
</div></blockquote>
<p><strong>batch_size</strong> : int</p>
<blockquote>
<div><p>batch_size passed to Keras</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">numpy.array of nM affinity predictions</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.Class1NeuralNetwork.supported_peptide_lengths">
<code class="descname">supported_peptide_lengths</code><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.supported_peptide_lengths" title="Permalink to this definition">¶</a></dt>
<dd><p>(minimum, maximum) lengths of peptides supported, inclusive.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">(int, int) tuple</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1NeuralNetwork.update_network_description">
<code class="descname">update_network_description</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.update_network_description"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.update_network_description" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-mhcflurry.allele_encoding">
<span id="mhcflurry-allele-encoding-module"></span><h2>mhcflurry.allele_encoding module<a class="headerlink" href="#module-mhcflurry.allele_encoding" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="mhcflurry.allele_encoding.AlleleEncoding">
<em class="property">class </em><code class="descclassname">mhcflurry.allele_encoding.</code><code class="descname">AlleleEncoding</code><span class="sig-paren">(</span><em>alleles</em>, <em>allele_to_fixed_length_sequence=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/allele_encoding.html#AlleleEncoding"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.allele_encoding.AlleleEncoding" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<p>A place to cache encodings for a (potentially large) sequence of alleles.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>alleles</strong> : list of string</p>
<blockquote>
<div><p>Allele names</p>
</div></blockquote>
<p><strong>allele_to_fixed_length_sequence</strong> : dict of str -&gt; str</p>
<blockquote class="last">
<div><p>Allele name to fixed lengths sequence (“pseudosequence”)</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="mhcflurry.allele_encoding.AlleleEncoding.fixed_length_vector_encoded_sequences">
<code class="descname">fixed_length_vector_encoded_sequences</code><span class="sig-paren">(</span><em>vector_encoding_name</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/allele_encoding.html#AlleleEncoding.fixed_length_vector_encoded_sequences"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.allele_encoding.AlleleEncoding.fixed_length_vector_encoded_sequences" title="Permalink to this definition">¶</a></dt>
<dd><p>Encode alleles.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>vector_encoding_name</strong> : string</p>
<blockquote>
<div><p>How to represent amino acids.
One of “BLOSUM62”, “one-hot”, etc. Full list of supported vector
encodings is given by available_vector_encodings() in amino_acid.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">numpy.array with shape (num sequences, sequence length, m) where m is</p>
<p class="last">vector_encoding_length(vector_encoding_name)</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-mhcflurry.amino_acid">
<span id="mhcflurry-amino-acid-module"></span><h2>mhcflurry.amino_acid module<a class="headerlink" href="#module-mhcflurry.amino_acid" title="Permalink to this headline">¶</a></h2>
<p>Functions for encoding fixed length sequences of amino acids into various
vector representations, such as one-hot and BLOSUM62.</p>
<dl class="function">
<dt id="mhcflurry.amino_acid.available_vector_encodings">
<code class="descclassname">mhcflurry.amino_acid.</code><code class="descname">available_vector_encodings</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/amino_acid.html#available_vector_encodings"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.amino_acid.available_vector_encodings" title="Permalink to this definition">¶</a></dt>
<dd><p>Return list of supported amino acid vector encodings.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">list of string</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.amino_acid.vector_encoding_length">
<code class="descclassname">mhcflurry.amino_acid.</code><code class="descname">vector_encoding_length</code><span class="sig-paren">(</span><em>name</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/amino_acid.html#vector_encoding_length"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.amino_acid.vector_encoding_length" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the length of the given vector encoding.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>name</strong> : string</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">int</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.amino_acid.index_encoding">
<code class="descclassname">mhcflurry.amino_acid.</code><code class="descname">index_encoding</code><span class="sig-paren">(</span><em>sequences</em>, <em>letter_to_index_dict</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/amino_acid.html#index_encoding"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.amino_acid.index_encoding" title="Permalink to this definition">¶</a></dt>
<dd><p>Encode a sequence of same-length strings to a matrix of integers of the
same shape. The map from characters to integers is given by
<code class="xref py py-obj docutils literal"><span class="pre">letter_to_index_dict</span></code>.</p>
<p>Given a sequence of <code class="xref py py-obj docutils literal"><span class="pre">n</span></code> strings all of length <code class="xref py py-obj docutils literal"><span class="pre">k</span></code>, return a <code class="xref py py-obj docutils literal"><span class="pre">k</span> <span class="pre">*</span> <span class="pre">n</span></code> array where
the (<code class="xref py py-obj docutils literal"><span class="pre">i</span></code>, <code class="xref py py-obj docutils literal"><span class="pre">j</span></code>)th element is <code class="xref py py-obj docutils literal"><span class="pre">letter_to_index_dict[sequence[i][j]]</span></code>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>sequences</strong> : list of length n of strings of length k</p>
<p><strong>letter_to_index_dict</strong> : dict</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">numpy.array of integers with shape (<code class="xref py py-obj docutils literal"><span class="pre">k</span></code>, <code class="xref py py-obj docutils literal"><span class="pre">n</span></code>)</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.amino_acid.fixed_vectors_encoding">
<code class="descclassname">mhcflurry.amino_acid.</code><code class="descname">fixed_vectors_encoding</code><span class="sig-paren">(</span><em>index_encoded_sequences</em>, <em>letter_to_vector_df</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/amino_acid.html#fixed_vectors_encoding"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.amino_acid.fixed_vectors_encoding" title="Permalink to this definition">¶</a></dt>
<dd><p>Given a <code class="xref py py-obj docutils literal"><span class="pre">n</span></code> x <code class="xref py py-obj docutils literal"><span class="pre">k</span></code> matrix of integers such as that returned by <a class="reference internal" href="#mhcflurry.amino_acid.index_encoding" title="mhcflurry.amino_acid.index_encoding"><code class="xref py py-obj docutils literal"><span class="pre">index_encoding()</span></code></a> and
a dataframe mapping each index to an arbitrary vector, return a <code class="xref py py-obj docutils literal"><span class="pre">n</span> <span class="pre">*</span> <span class="pre">k</span> <span class="pre">*</span> <span class="pre">m</span></code>
array where the (<code class="xref py py-obj docutils literal"><span class="pre">i</span></code>, <code class="xref py py-obj docutils literal"><span class="pre">j</span></code>)’th element is <code class="xref py py-obj docutils literal"><span class="pre">letter_to_vector_df.iloc[sequence[i][j]]</span></code>.</p>
<p>The dataframe index and columns names are ignored here; the indexing is done
entirely by integer position in the dataframe.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>index_encoded_sequences</strong> : <code class="xref py py-obj docutils literal"><span class="pre">n</span></code> x <code class="xref py py-obj docutils literal"><span class="pre">k</span></code> array of integers</p>
<p><strong>letter_to_vector_df</strong> : pandas.DataFrame of shape (<code class="xref py py-obj docutils literal"><span class="pre">alphabet</span> <span class="pre">size</span></code>, <code class="xref py py-obj docutils literal"><span class="pre">m</span></code>)</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">numpy.array of integers with shape (<code class="xref py py-obj docutils literal"><span class="pre">n</span></code>, <code class="xref py py-obj docutils literal"><span class="pre">k</span></code>, <code class="xref py py-obj docutils literal"><span class="pre">m</span></code>)</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="module-mhcflurry.class1_affinity_predictor">
<span id="mhcflurry-class1-affinity-predictor-module"></span><h2>mhcflurry.class1_affinity_predictor module<a class="headerlink" href="#module-mhcflurry.class1_affinity_predictor" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor">
<em class="property">class </em><code class="descclassname">mhcflurry.class1_affinity_predictor.</code><code class="descname">Class1AffinityPredictor</code><span class="sig-paren">(</span><em>allele_to_allele_specific_models=None</em>, <em>class1_pan_allele_models=None</em>, <em>allele_to_fixed_length_sequence=None</em>, <em>manifest_df=None</em>, <em>allele_to_percent_rank_transform=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<p>High-level interface for peptide/MHC I binding affinity prediction.</p>
<p>This class manages low-level <code class="xref py py-obj docutils literal"><span class="pre">Class1NeuralNetwork</span></code> instances, each of which
wraps a single Keras network. The purpose of <a class="reference internal" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor" title="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor"><code class="xref py py-obj docutils literal"><span class="pre">Class1AffinityPredictor</span></code></a> is to
implement ensembles, handling of multiple alleles, and predictor loading and
saving.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>allele_to_allele_specific_models</strong> : dict of string -&gt; list of <code class="xref py py-obj docutils literal"><span class="pre">Class1NeuralNetwork</span></code></p>
<blockquote>
<div><p>Ensemble of single-allele models to use for each allele.</p>
</div></blockquote>
<p><strong>class1_pan_allele_models</strong> : list of <code class="xref py py-obj docutils literal"><span class="pre">Class1NeuralNetwork</span></code></p>
<blockquote>
<div><p>Ensemble of pan-allele models.</p>
</div></blockquote>
<p><strong>allele_to_fixed_length_sequence</strong> : dict of string -&gt; string</p>
<blockquote>
<div><p>Required only if class1_pan_allele_models is specified.</p>
</div></blockquote>
<p><strong>manifest_df</strong> : <code class="xref py py-obj docutils literal"><span class="pre">pandas.DataFrame</span></code>, optional</p>
<blockquote>
<div><p>Must have columns: model_name, allele, config_json, model.
Only required if you want to update an existing serialization of a
Class1AffinityPredictor. Otherwise this dataframe will be generated
automatically based on the supplied models.</p>
</div></blockquote>
<p><strong>allele_to_percent_rank_transform</strong> : dict of string -&gt; <code class="xref py py-obj docutils literal"><span class="pre">PercentRankTransform</span></code>, optional</p>
<blockquote class="last">
<div><p><code class="xref py py-obj docutils literal"><span class="pre">PercentRankTransform</span></code> instances to use for each allele</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<dl class="attribute">
<dt id="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.neural_networks">
<code class="descname">neural_networks</code><a class="headerlink" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.neural_networks" title="Permalink to this definition">¶</a></dt>
<dd><p>List of the neural networks in the ensemble.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">list of <code class="xref py py-obj docutils literal"><span class="pre">Class1NeuralNetwork</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="classmethod">
<dt id="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.merge">
<em class="property">classmethod </em><code class="descname">merge</code><span class="sig-paren">(</span><em>predictors</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.merge"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.merge" title="Permalink to this definition">¶</a></dt>
<dd><p>Merge the ensembles of two or more <a class="reference internal" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor" title="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor"><code class="xref py py-obj docutils literal"><span class="pre">Class1AffinityPredictor</span></code></a> instances.</p>
<p>Note: the resulting merged predictor will NOT have calibrated percentile
ranks. Call <a class="reference internal" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.calibrate_percentile_ranks" title="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.calibrate_percentile_ranks"><code class="xref py py-obj docutils literal"><span class="pre">calibrate_percentile_ranks</span></code></a> on it if these are needed.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>predictors</strong> : sequence of <a class="reference internal" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor" title="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor"><code class="xref py py-obj docutils literal"><span class="pre">Class1AffinityPredictor</span></code></a></td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><a class="reference internal" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor" title="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor"><code class="xref py py-obj docutils literal"><span class="pre">Class1AffinityPredictor</span></code></a> instance</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.supported_alleles">
<code class="descname">supported_alleles</code><a class="headerlink" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.supported_alleles" title="Permalink to this definition">¶</a></dt>
<dd><p>Alleles for which predictions can be made.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">list of string</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.supported_peptide_lengths">
<code class="descname">supported_peptide_lengths</code><a class="headerlink" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.supported_peptide_lengths" title="Permalink to this definition">¶</a></dt>
<dd><p>(minimum, maximum) lengths of peptides supported by <em>all models</em>,
inclusive.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">(int, int) tuple</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.save">
<code class="descname">save</code><span class="sig-paren">(</span><em>models_dir</em>, <em>model_names_to_write=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.save"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.save" title="Permalink to this definition">¶</a></dt>
<dd><p>Serialize the predictor to a directory on disk. If the directory does
not exist it will be created.</p>
<p>The serialization format consists of a file called “manifest.csv” with
the configurations of each Class1NeuralNetwork, along with per-network
files giving the model weights. If there are pan-allele predictors in
the ensemble, the allele sequences are also stored in the
directory. There is also a small file “index.txt” with basic metadata:
when the models were trained, by whom, on what host.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>models_dir</strong> : string</p>
<blockquote>
<div><p>Path to directory</p>
</div></blockquote>
<p><strong>model_names_to_write</strong> : list of string, optional</p>
<blockquote class="last">
<div><p>Only write the weights for the specified models. Useful for
incremental updates during training.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="staticmethod">
<dt id="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.load">
<em class="property">static </em><code class="descname">load</code><span class="sig-paren">(</span><em>models_dir=None</em>, <em>max_models=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.load"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.load" title="Permalink to this definition">¶</a></dt>
<dd><p>Deserialize a predictor from a directory on disk.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>models_dir</strong> : string</p>
<blockquote>
<div><p>Path to directory</p>
</div></blockquote>
<p><strong>max_models</strong> : int, optional</p>
<blockquote>
<div><p>Maximum number of <code class="xref py py-obj docutils literal"><span class="pre">Class1NeuralNetwork</span></code> instances to load</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last"><a class="reference internal" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor" title="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor"><code class="xref py py-obj docutils literal"><span class="pre">Class1AffinityPredictor</span></code></a> instance</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="staticmethod">
<dt id="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.model_name">
<em class="property">static </em><code class="descname">model_name</code><span class="sig-paren">(</span><em>allele</em>, <em>num</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.model_name"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.model_name" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate a model name</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>allele</strong> : string</p>
<p><strong>num</strong> : int</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">string</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="staticmethod">
<dt id="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.weights_path">
<em class="property">static </em><code class="descname">weights_path</code><span class="sig-paren">(</span><em>models_dir</em>, <em>model_name</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.weights_path"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.weights_path" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate the path to the weights file for a model</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>models_dir</strong> : string</p>
<p><strong>model_name</strong> : string</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">string</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.fit_allele_specific_predictors">
<code class="descname">fit_allele_specific_predictors</code><span class="sig-paren">(</span><em>n_models</em>, <em>architecture_hyperparameters_list</em>, <em>allele</em>, <em>peptides</em>, <em>affinities</em>, <em>inequalities=None</em>, <em>models_dir_for_save=None</em>, <em>verbose=0</em>, <em>progress_preamble=''</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.fit_allele_specific_predictors"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.fit_allele_specific_predictors" title="Permalink to this definition">¶</a></dt>
<dd><p>Fit one or more allele specific predictors for a single allele using a
single neural network architecture.</p>
<p>The new predictors are saved in the Class1AffinityPredictor instance
and will be used on subsequent calls to <a class="reference internal" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.predict" title="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.predict"><code class="xref py py-obj docutils literal"><span class="pre">predict</span></code></a>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>n_models</strong> : int</p>
<blockquote>
<div><p>Number of neural networks to fit</p>
</div></blockquote>
<p><strong>architecture_hyperparameters_list</strong> : list of dict</p>
<blockquote>
<div><p>List of hyperparameters. If more than one set of hyperparameters
are specified, model selection over them is performed and model
with the lowest validation loss is selected for each fold.</p>
</div></blockquote>
<p><strong>allele</strong> : string</p>
<p><strong>peptides</strong> : <code class="xref py py-obj docutils literal"><span class="pre">EncodableSequences</span></code> or list of string</p>
<p><strong>affinities</strong> : list of float</p>
<blockquote>
<div><p>nM affinities</p>
</div></blockquote>
<p><strong>inequalities</strong> : list of string, each element one of “&gt;”, “&lt;”, or “=”</p>
<blockquote>
<div><p>See Class1NeuralNetwork.fit for details.</p>
</div></blockquote>
<p><strong>models_dir_for_save</strong> : string, optional</p>
<blockquote>
<div><p>If specified, the Class1AffinityPredictor is (incrementally) written
to the given models dir after each neural network is fit.</p>
</div></blockquote>
<p><strong>verbose</strong> : int</p>
<blockquote>
<div><p>Keras verbosity</p>
</div></blockquote>
<p><strong>progress_preamble</strong> : string</p>
<blockquote>
<div><p>Optional string of information to include in each progress update</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">list of <code class="xref py py-obj docutils literal"><span class="pre">Class1NeuralNetwork</span></code></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.fit_class1_pan_allele_models">
<code class="descname">fit_class1_pan_allele_models</code><span class="sig-paren">(</span><em>n_models</em>, <em>architecture_hyperparameters</em>, <em>alleles</em>, <em>peptides</em>, <em>affinities</em>, <em>inequalities</em>, <em>models_dir_for_save=None</em>, <em>verbose=1</em>, <em>progress_preamble=''</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.fit_class1_pan_allele_models"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.fit_class1_pan_allele_models" title="Permalink to this definition">¶</a></dt>
<dd><p>Fit one or more pan-allele predictors using a single neural network
architecture.</p>
<p>The new predictors are saved in the Class1AffinityPredictor instance
and will be used on subsequent calls to <a class="reference internal" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.predict" title="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.predict"><code class="xref py py-obj docutils literal"><span class="pre">predict</span></code></a>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>n_models</strong> : int</p>
<blockquote>
<div><p>Number of neural networks to fit</p>
</div></blockquote>
<p><strong>architecture_hyperparameters</strong> : dict</p>
<p><strong>alleles</strong> : list of string</p>
<blockquote>
<div><p>Allele names (not sequences) corresponding to each peptide</p>
</div></blockquote>
<p><strong>peptides</strong> : <code class="xref py py-obj docutils literal"><span class="pre">EncodableSequences</span></code> or list of string</p>
<p><strong>affinities</strong> : list of float</p>
<blockquote>
<div><p>nM affinities</p>
</div></blockquote>
<p><strong>inequalities</strong> : list of string, each element one of “&gt;”, “&lt;”, or “=”</p>
<blockquote>
<div><p>See Class1NeuralNetwork.fit for details.</p>
</div></blockquote>
<p><strong>models_dir_for_save</strong> : string, optional</p>
<blockquote>
<div><p>If specified, the Class1AffinityPredictor is (incrementally) written
to the given models dir after each neural network is fit.</p>
</div></blockquote>
<p><strong>verbose</strong> : int</p>
<blockquote>
<div><p>Keras verbosity</p>
</div></blockquote>
<p><strong>progress_preamble</strong> : string</p>
<blockquote>
<div><p>Optional string of information to include in each progress update</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">list of <code class="xref py py-obj docutils literal"><span class="pre">Class1NeuralNetwork</span></code></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.percentile_ranks">
<code class="descname">percentile_ranks</code><span class="sig-paren">(</span><em>affinities</em>, <em>allele=None</em>, <em>alleles=None</em>, <em>throw=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.percentile_ranks"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.percentile_ranks" title="Permalink to this definition">¶</a></dt>
<dd><p>Return percentile ranks for the given ic50 affinities and alleles.</p>
<p>The ‘allele’ and ‘alleles’ argument are as in the <a class="reference internal" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.predict" title="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.predict"><code class="xref py py-obj docutils literal"><span class="pre">predict</span></code></a> method.
Specify one of these.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>affinities</strong> : sequence of float</p>
<blockquote>
<div><p>nM affinities</p>
</div></blockquote>
<p><strong>allele</strong> : string</p>
<p><strong>alleles</strong> : sequence of string</p>
<p><strong>throw</strong> : boolean</p>
<blockquote>
<div><p>If True, a ValueError will be raised in the case of unsupported
alleles. If False, a warning will be logged and NaN will be returned
for those percentile ranks.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">numpy.array of float</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.predict">
<code class="descname">predict</code><span class="sig-paren">(</span><em>peptides</em>, <em>alleles=None</em>, <em>allele=None</em>, <em>throw=True</em>, <em>centrality_measure='robust_mean'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.predict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.predict" title="Permalink to this definition">¶</a></dt>
<dd><p>Predict nM binding affinities.</p>
<p>If multiple predictors are available for an allele, the predictions are
the geometric means of the individual model predictions.</p>
<p>One of ‘allele’ or ‘alleles’ must be specified. If ‘allele’ is specified
all predictions will be for the given allele. If ‘alleles’ is specified
it must be the same length as ‘peptides’ and give the allele
corresponding to each peptide.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>peptides</strong> : <code class="xref py py-obj docutils literal"><span class="pre">EncodableSequences</span></code> or list of string</p>
<p><strong>alleles</strong> : list of string</p>
<p><strong>allele</strong> : string</p>
<p><strong>throw</strong> : boolean</p>
<blockquote>
<div><p>If True, a ValueError will be raised in the case of unsupported
alleles or peptide lengths. If False, a warning will be logged and
the predictions for the unsupported alleles or peptides will be NaN.</p>
</div></blockquote>
<p><strong>centrality_measure</strong> : string or callable</p>
<blockquote>
<div><p>Measure of central tendency to use to combine predictions in the
ensemble. Options include: mean, median, robust_mean.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">numpy.array of predictions</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.predict_to_dataframe">
<code class="descname">predict_to_dataframe</code><span class="sig-paren">(</span><em>peptides</em>, <em>alleles=None</em>, <em>allele=None</em>, <em>throw=True</em>, <em>include_individual_model_predictions=False</em>, <em>include_percentile_ranks=True</em>, <em>centrality_measure='robust_mean'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.predict_to_dataframe"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.predict_to_dataframe" title="Permalink to this definition">¶</a></dt>
<dd><p>Predict nM binding affinities. Gives more detailed output than <a class="reference internal" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.predict" title="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.predict"><code class="xref py py-obj docutils literal"><span class="pre">predict</span></code></a>
method, including 5-95% prediction intervals.</p>
<p>If multiple predictors are available for an allele, the predictions are
the geometric means of the individual model predictions.</p>
<p>One of ‘allele’ or ‘alleles’ must be specified. If ‘allele’ is specified
all predictions will be for the given allele. If ‘alleles’ is specified
it must be the same length as ‘peptides’ and give the allele
corresponding to each peptide.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>peptides</strong> : <code class="xref py py-obj docutils literal"><span class="pre">EncodableSequences</span></code> or list of string</p>
<p><strong>alleles</strong> : list of string</p>
<p><strong>allele</strong> : string</p>
<p><strong>throw</strong> : boolean</p>
<blockquote>
<div><p>If True, a ValueError will be raised in the case of unsupported
alleles or peptide lengths. If False, a warning will be logged and
the predictions for the unsupported alleles or peptides will be NaN.</p>
</div></blockquote>
<p><strong>include_individual_model_predictions</strong> : boolean</p>
<blockquote>
<div><p>If True, the predictions of each individual model are included as
columns in the result dataframe.</p>
</div></blockquote>
<p><strong>include_percentile_ranks</strong> : boolean, default True</p>
<blockquote>
<div><p>If True, a “prediction_percentile” column will be included giving the
percentile ranks. If no percentile rank information is available,
this will be ignored with a warning.</p>
</div></blockquote>
<p><strong>centrality_measure</strong> : string or callable</p>
<blockquote>
<div><p>Measure of central tendency to use to combine predictions in the
ensemble. Options include: mean, median, robust_mean.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last"><code class="xref py py-obj docutils literal"><span class="pre">pandas.DataFrame</span></code> of predictions</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="staticmethod">
<dt id="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.save_weights">
<em class="property">static </em><code class="descname">save_weights</code><span class="sig-paren">(</span><em>weights_list</em>, <em>filename</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.save_weights"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.save_weights" title="Permalink to this definition">¶</a></dt>
<dd><p>Save the model weights to the given filename using numpy’s “.npz”
format.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>weights_list</strong> : list of array</p>
<p><strong>filename</strong> : string</p>
<blockquote class="last">
<div><p>Should end in “.npz”.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="staticmethod">
<dt id="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.load_weights">
<em class="property">static </em><code class="descname">load_weights</code><span class="sig-paren">(</span><em>filename</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.load_weights"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.load_weights" title="Permalink to this definition">¶</a></dt>
<dd><p>Restore model weights from the given filename, which should have been
created with <a class="reference internal" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.save_weights" title="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.save_weights"><code class="xref py py-obj docutils literal"><span class="pre">save_weights</span></code></a>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>filename</strong> : string</p>
<blockquote>
<div><p>Should end in “.npz”.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">list of array</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.calibrate_percentile_ranks">
<code class="descname">calibrate_percentile_ranks</code><span class="sig-paren">(</span><em>peptides=None</em>, <em>num_peptides_per_length=100000</em>, <em>alleles=None</em>, <em>bins=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.calibrate_percentile_ranks"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.calibrate_percentile_ranks" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute the cumulative distribution of ic50 values for a set of alleles
over a large universe of random peptides, to enable computing quantiles in
this distribution later.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>peptides</strong> : sequence of string or EncodableSequences, optional</p>
<blockquote>
<div><p>Peptides to use</p>
</div></blockquote>
<p><strong>num_peptides_per_length</strong> : int, optional</p>
<blockquote>
<div><p>If peptides argument is not specified, then num_peptides_per_length
peptides are randomly sampled from a uniform distribution for each
supported length</p>
</div></blockquote>
<p><strong>alleles</strong> : sequence of string, optional</p>
<blockquote>
<div><p>Alleles to perform calibration for. If not specified all supported
alleles will be calibrated.</p>
</div></blockquote>
<p><strong>bins</strong> : object</p>
<blockquote>
<div><p>Anything that can be passed to numpy.histogram’s “bins” argument
can be used here, i.e. either an integer or a sequence giving bin
edges. This is in ic50 space.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last"><strong>EncodableSequences</strong> : peptides used for calibration</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-mhcflurry.class1_neural_network">
<span id="mhcflurry-class1-neural-network-module"></span><h2>mhcflurry.class1_neural_network module<a class="headerlink" href="#module-mhcflurry.class1_neural_network" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork">
<em class="property">class </em><code class="descclassname">mhcflurry.class1_neural_network.</code><code class="descname">Class1NeuralNetwork</code><span class="sig-paren">(</span><em>**hyperparameters</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<p>Low level class I predictor consisting of a single neural network.</p>
<p>Both single allele and pan-allele prediction are supported, but pan-allele
is in development and not yet well performing.</p>
<p>Users will generally use Class1AffinityPredictor, which gives a higher-level
interface and supports ensembles.</p>
<dl class="attribute">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.network_hyperparameter_defaults">
<code class="descname">network_hyperparameter_defaults</code><em class="property"> = &lt;mhcflurry.hyperparameters.HyperparameterDefaults object&gt;</em><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.network_hyperparameter_defaults" title="Permalink to this definition">¶</a></dt>
<dd><p>Hyperparameters (and their default values) that affect the neural network
architecture.</p>
</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.compile_hyperparameter_defaults">
<code class="descname">compile_hyperparameter_defaults</code><em class="property"> = &lt;mhcflurry.hyperparameters.HyperparameterDefaults object&gt;</em><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.compile_hyperparameter_defaults" title="Permalink to this definition">¶</a></dt>
<dd><p>Loss and optimizer hyperparameters. Any values supported by keras may be
used.</p>
</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.input_encoding_hyperparameter_defaults">
<code class="descname">input_encoding_hyperparameter_defaults</code><em class="property"> = &lt;mhcflurry.hyperparameters.HyperparameterDefaults object&gt;</em><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.input_encoding_hyperparameter_defaults" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of amino acid residues that are given fixed positions on the each
side in the variable length encoding.</p>
</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.fit_hyperparameter_defaults">
<code class="descname">fit_hyperparameter_defaults</code><em class="property"> = &lt;mhcflurry.hyperparameters.HyperparameterDefaults object&gt;</em><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.fit_hyperparameter_defaults" title="Permalink to this definition">¶</a></dt>
<dd><p>Hyperparameters for neural network training.</p>
</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.early_stopping_hyperparameter_defaults">
<code class="descname">early_stopping_hyperparameter_defaults</code><em class="property"> = &lt;mhcflurry.hyperparameters.HyperparameterDefaults object&gt;</em><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.early_stopping_hyperparameter_defaults" title="Permalink to this definition">¶</a></dt>
<dd><p>Hyperparameters for early stopping.</p>
</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.hyperparameter_defaults">
<code class="descname">hyperparameter_defaults</code><em class="property"> = &lt;mhcflurry.hyperparameters.HyperparameterDefaults object&gt;</em><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.hyperparameter_defaults" title="Permalink to this definition">¶</a></dt>
<dd><p>Combined set of all supported hyperparameters and their default values.</p>
</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.hyperparameter_renames">
<code class="descname">hyperparameter_renames</code><em class="property"> = {'use_embedding': None, 'pseudosequence_use_embedding': None}</em><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.hyperparameter_renames" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="classmethod">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.apply_hyperparameter_renames">
<em class="property">classmethod </em><code class="descname">apply_hyperparameter_renames</code><span class="sig-paren">(</span><em>hyperparameters</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.apply_hyperparameter_renames"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.apply_hyperparameter_renames" title="Permalink to this definition">¶</a></dt>
<dd><p>Handle hyperparameter renames.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>hyperparameters</strong> : dict</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><strong>dict</strong> : updated hyperparameters</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.KERAS_MODELS_CACHE">
<code class="descname">KERAS_MODELS_CACHE</code><em class="property"> = {}</em><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.KERAS_MODELS_CACHE" title="Permalink to this definition">¶</a></dt>
<dd><p>Process-wide keras model cache, a map from: architecture JSON string to
(Keras model, existing network weights)</p>
</dd></dl>

<dl class="classmethod">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.clear_model_cache">
<em class="property">classmethod </em><code class="descname">clear_model_cache</code><span class="sig-paren">(</span><em>klass</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.clear_model_cache"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.clear_model_cache" title="Permalink to this definition">¶</a></dt>
<dd><p>Clear the Keras model cache.</p>
</dd></dl>

<dl class="classmethod">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.borrow_cached_network">
<em class="property">classmethod </em><code class="descname">borrow_cached_network</code><span class="sig-paren">(</span><em>klass</em>, <em>network_json</em>, <em>network_weights</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.borrow_cached_network"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.borrow_cached_network" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a keras Model with the specified architecture and weights.
As an optimization, when possible this will reuse architectures from a
process-wide cache.</p>
<p>The returned object is “borrowed” in the sense that its weights can
change later after subsequent calls to this method from other objects.</p>
<p>If you’re using this from a parallel implementation you’ll need to
hold a lock while using the returned object.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>network_json</strong> : string of JSON</p>
<p><strong>network_weights</strong> : list of numpy.array</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">keras.models.Model</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.network">
<code class="descname">network</code><span class="sig-paren">(</span><em>borrow=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.network"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.network" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the keras model associated with this predictor.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>borrow</strong> : bool</p>
<blockquote>
<div><p>Whether to return a cached model if possible. See
borrow_cached_network for details</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">keras.models.Model</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.update_network_description">
<code class="descname">update_network_description</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.update_network_description"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.update_network_description" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.get_config">
<code class="descname">get_config</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.get_config"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.get_config" title="Permalink to this definition">¶</a></dt>
<dd><p>serialize to a dict all attributes except model weights</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">dict</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="classmethod">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.from_config">
<em class="property">classmethod </em><code class="descname">from_config</code><span class="sig-paren">(</span><em>config</em>, <em>weights=None</em>, <em>weights_loader=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.from_config"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.from_config" title="Permalink to this definition">¶</a></dt>
<dd><p>deserialize from a dict returned by get_config().</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>config</strong> : dict</p>
<p><strong>weights</strong> : list of array, optional</p>
<blockquote>
<div><p>Network weights to restore</p>
</div></blockquote>
<p><strong>weights_loader</strong> : callable, optional</p>
<blockquote>
<div><p>Function to call (no arguments) to load weights when needed</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">Class1NeuralNetwork</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.load_weights">
<code class="descname">load_weights</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.load_weights"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.load_weights" title="Permalink to this definition">¶</a></dt>
<dd><p>Load weights by evaluating self.network_weights_loader, if needed.</p>
<p>After calling this, self.network_weights_loader will be None and
self.network_weights will be the weights list, if available.</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.get_weights">
<code class="descname">get_weights</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.get_weights"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.get_weights" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the network weights</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">list of numpy.array giving weights for each layer</p>
<p class="last">or None if there is no network</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.peptides_to_network_input">
<code class="descname">peptides_to_network_input</code><span class="sig-paren">(</span><em>peptides</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.peptides_to_network_input"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.peptides_to_network_input" title="Permalink to this definition">¶</a></dt>
<dd><p>Encode peptides to the fixed-length encoding expected by the neural
network (which depends on the architecture).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>peptides</strong> : EncodableSequences or list of string</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">numpy.array</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.supported_peptide_lengths">
<code class="descname">supported_peptide_lengths</code><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.supported_peptide_lengths" title="Permalink to this definition">¶</a></dt>
<dd><p>(minimum, maximum) lengths of peptides supported, inclusive.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">(int, int) tuple</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.allele_encoding_to_network_input">
<code class="descname">allele_encoding_to_network_input</code><span class="sig-paren">(</span><em>allele_encoding</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.allele_encoding_to_network_input"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.allele_encoding_to_network_input" title="Permalink to this definition">¶</a></dt>
<dd><p>Encode alleles to the fixed-length encoding expected by the neural
network (which depends on the architecture).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>allele_encoding</strong> : AlleleEncoding</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">numpy.array</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.fit">
<code class="descname">fit</code><span class="sig-paren">(</span><em>peptides</em>, <em>affinities</em>, <em>allele_encoding=None</em>, <em>inequalities=None</em>, <em>sample_weights=None</em>, <em>shuffle_permutation=None</em>, <em>verbose=1</em>, <em>progress_preamble=''</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.fit"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.fit" title="Permalink to this definition">¶</a></dt>
<dd><p>Fit the neural network.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>peptides</strong> : EncodableSequences or list of string</p>
<p><strong>affinities</strong> : list of float</p>
<blockquote>
<div><p>nM affinities. Must be same length of as peptides.</p>
</div></blockquote>
<p><strong>allele_encoding</strong> : AlleleEncoding, optional</p>
<blockquote>
<div><p>If not specified, the model will be a single-allele predictor.</p>
</div></blockquote>
<p><strong>inequalities</strong> : list of string, each element one of “&gt;”, “&lt;”, or “=”.</p>
<blockquote>
<div><p>Inequalities to use for fitting. Same length as affinities.
Each element must be one of “&gt;”, “&lt;”, or “=”. For example, a “&gt;”
will train on y_pred &gt; y_true for that element in the training set.
Requires using a custom losses that support inequalities (e.g.
mse_with_ineqalities).
If None all inequalities are taken to be “=”.</p>
</div></blockquote>
<p><strong>sample_weights</strong> : list of float, optional</p>
<blockquote>
<div><p>If not specified, all samples (including random negatives added
during training) will have equal weight. If specified, the random
negatives will be assigned weight=1.0.</p>
</div></blockquote>
<p><strong>shuffle_permutation</strong> : list of int, optional</p>
<blockquote>
<div><p>Permutation (integer list) of same length as peptides and affinities
If None, then a random permutation will be generated.</p>
</div></blockquote>
<p><strong>verbose</strong> : int</p>
<blockquote>
<div><p>Keras verbosity level</p>
</div></blockquote>
<p><strong>progress_preamble</strong> : string</p>
<blockquote class="last">
<div><p>Optional string of information to include in each progress update</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.predict">
<code class="descname">predict</code><span class="sig-paren">(</span><em>peptides</em>, <em>allele_encoding=None</em>, <em>batch_size=4096</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.predict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.predict" title="Permalink to this definition">¶</a></dt>
<dd><p>Predict affinities</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>peptides</strong> : EncodableSequences or list of string</p>
<p><strong>allele_pseudosequences</strong> : AlleleEncoding, optional</p>
<blockquote>
<div><p>Only required when this model is a pan-allele model</p>
</div></blockquote>
<p><strong>batch_size</strong> : int</p>
<blockquote>
<div><p>batch_size passed to Keras</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">numpy.array of nM affinity predictions</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="staticmethod">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.make_network">
<em class="property">static </em><code class="descname">make_network</code><span class="sig-paren">(</span><em>allele_encoding_dims</em>, <em>kmer_size</em>, <em>peptide_amino_acid_encoding</em>, <em>embedding_input_dim</em>, <em>embedding_output_dim</em>, <em>allele_dense_layer_sizes</em>, <em>peptide_dense_layer_sizes</em>, <em>peptide_allele_merge_method</em>, <em>peptide_allele_merge_activation</em>, <em>layer_sizes</em>, <em>dense_layer_l1_regularization</em>, <em>dense_layer_l2_regularization</em>, <em>activation</em>, <em>init</em>, <em>output_activation</em>, <em>dropout_probability</em>, <em>batch_normalization</em>, <em>embedding_init_method</em>, <em>locally_connected_layers</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.make_network"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.make_network" title="Permalink to this definition">¶</a></dt>
<dd><p>Helper function to make a keras network for class1 affinity prediction.</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-mhcflurry.common">
<span id="mhcflurry-common-module"></span><h2>mhcflurry.common module<a class="headerlink" href="#module-mhcflurry.common" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="mhcflurry.common.set_keras_backend">
<code class="descclassname">mhcflurry.common.</code><code class="descname">set_keras_backend</code><span class="sig-paren">(</span><em>backend</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/common.html#set_keras_backend"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.common.set_keras_backend" title="Permalink to this definition">¶</a></dt>
<dd><p>Configure Keras backend to use GPU or CPU. Only tensorflow is supported.</p>
<p>Must be called before Keras has been imported.</p>
<p>backend must be ‘tensorflow-cpu’ or ‘tensorflow-gpu’.</p>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.common.configure_logging">
<code class="descclassname">mhcflurry.common.</code><code class="descname">configure_logging</code><span class="sig-paren">(</span><em>verbose=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/common.html#configure_logging"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.common.configure_logging" title="Permalink to this definition">¶</a></dt>
<dd><p>Configure logging module using defaults.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>verbose</strong> : boolean</p>
<blockquote class="last">
<div><p>If true, output will be at level DEBUG, otherwise, INFO.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.common.amino_acid_distribution">
<code class="descclassname">mhcflurry.common.</code><code class="descname">amino_acid_distribution</code><span class="sig-paren">(</span><em>peptides</em>, <em>smoothing=0.0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/common.html#amino_acid_distribution"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.common.amino_acid_distribution" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute the fraction of each amino acid across a collection of peptides.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>peptides</strong> : list of string</p>
<p><strong>smoothing</strong> : float, optional</p>
<blockquote>
<div><p>Small number (e.g. 0.01) to add to all amino acid fractions. The higher
the number the more uniform the distribution.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">pandas.Series indexed by amino acids</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.common.random_peptides">
<code class="descclassname">mhcflurry.common.</code><code class="descname">random_peptides</code><span class="sig-paren">(</span><em>num</em>, <em>length=9</em>, <em>distribution=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/common.html#random_peptides"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.common.random_peptides" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate random peptides (kmers).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>num</strong> : int</p>
<blockquote>
<div><p>Number of peptides to return</p>
</div></blockquote>
<p><strong>length</strong> : int</p>
<blockquote>
<div><p>Length of each peptide</p>
</div></blockquote>
<p><strong>distribution</strong> : pandas.Series</p>
<blockquote>
<div><p>Maps 1-letter amino acid abbreviations to
probabilities. If not specified a uniform
distribution is used.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">list of string</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="module-mhcflurry.custom_loss">
<span id="mhcflurry-custom-loss-module"></span><h2>mhcflurry.custom_loss module<a class="headerlink" href="#module-mhcflurry.custom_loss" title="Permalink to this headline">¶</a></h2>
<p>Custom loss functions.</p>
<p>For losses supporting inequalities, each training data point is associated with
one of (=), (&lt;), or (&gt;). For e.g. (&gt;) inequalities, penalization is applied only
if the prediction is less than the given value.</p>
<dl class="class">
<dt id="mhcflurry.custom_loss.MSEWithInequalities">
<em class="property">class </em><code class="descclassname">mhcflurry.custom_loss.</code><code class="descname">MSEWithInequalities</code><a class="reference internal" href="_modules/mhcflurry/custom_loss.html#MSEWithInequalities"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.custom_loss.MSEWithInequalities" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<p>Supports training a regressor on data that includes inequalities
(e.g. x &lt; 100). Mean square error is used as the loss for elements with
an (=) inequality. For elements with e.g. a (&gt; 0.5) inequality, then the loss
for that element is (y - 0.5)^2 (standard MSE) if y &lt; 500 and 0 otherwise.</p>
<p>This loss assumes that the normal range for y_true and y_pred is 0 - 1. As a
hack, the implementation uses other intervals for y_pred to encode the
inequality information.</p>
<p>y_true is interpreted as follows:</p>
<dl class="docutils">
<dt>between 0 - 1</dt>
<dd>Regular MSE loss is used. Penality (y_pred - y_true)**2 is applied if
y_pred is greater or less than y_true.</dd>
<dt>between 2 - 3:</dt>
<dd>Treated as a “&lt;” inequality. Penality (y_pred - (y_true - 2))**2 is
applied only if y_pred is greater than y_true - 2.</dd>
<dt>between 4 - 5:</dt>
<dd>Treated as a “&gt;” inequality. Penality (y_pred - (y_true - 4))**2 is
applied only if y_pred is less than y_true - 4.</dd>
</dl>
<dl class="attribute">
<dt id="mhcflurry.custom_loss.MSEWithInequalities.name">
<code class="descname">name</code><em class="property"> = 'mse_with_inequalities'</em><a class="headerlink" href="#mhcflurry.custom_loss.MSEWithInequalities.name" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="mhcflurry.custom_loss.MSEWithInequalities.supports_inequalities">
<code class="descname">supports_inequalities</code><em class="property"> = True</em><a class="headerlink" href="#mhcflurry.custom_loss.MSEWithInequalities.supports_inequalities" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="staticmethod">
<dt id="mhcflurry.custom_loss.MSEWithInequalities.encode_y">
<em class="property">static </em><code class="descname">encode_y</code><span class="sig-paren">(</span><em>y</em>, <em>inequalities=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/custom_loss.html#MSEWithInequalities.encode_y"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.custom_loss.MSEWithInequalities.encode_y" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="staticmethod">
<dt id="mhcflurry.custom_loss.MSEWithInequalities.loss">
<em class="property">static </em><code class="descname">loss</code><span class="sig-paren">(</span><em>y_true</em>, <em>y_pred</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/custom_loss.html#MSEWithInequalities.loss"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.custom_loss.MSEWithInequalities.loss" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.custom_loss.cls">
<code class="descclassname">mhcflurry.custom_loss.</code><code class="descname">cls</code><a class="headerlink" href="#mhcflurry.custom_loss.cls" title="Permalink to this definition">¶</a></dt>
<dd><p>alias of <a class="reference internal" href="#mhcflurry.custom_loss.MSEWithInequalities" title="mhcflurry.custom_loss.MSEWithInequalities"><code class="xref py py-class docutils literal"><span class="pre">MSEWithInequalities</span></code></a></p>
</dd></dl>

</div>
<div class="section" id="module-mhcflurry.downloads">
<span id="mhcflurry-downloads-module"></span><h2>mhcflurry.downloads module<a class="headerlink" href="#module-mhcflurry.downloads" title="Permalink to this headline">¶</a></h2>
<p>Manage local downloaded data.</p>
<dl class="function">
<dt id="mhcflurry.downloads.get_downloads_dir">
<code class="descclassname">mhcflurry.downloads.</code><code class="descname">get_downloads_dir</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/downloads.html#get_downloads_dir"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.downloads.get_downloads_dir" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the path to local downloaded data</p>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.downloads.get_current_release">
<code class="descclassname">mhcflurry.downloads.</code><code class="descname">get_current_release</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/downloads.html#get_current_release"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.downloads.get_current_release" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the current downloaded data release</p>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.downloads.get_downloads_metadata">
<code class="descclassname">mhcflurry.downloads.</code><code class="descname">get_downloads_metadata</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/downloads.html#get_downloads_metadata"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.downloads.get_downloads_metadata" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the contents of downloads.yml as a dict</p>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.downloads.get_default_class1_models_dir">
<code class="descclassname">mhcflurry.downloads.</code><code class="descname">get_default_class1_models_dir</code><span class="sig-paren">(</span><em>test_exists=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/downloads.html#get_default_class1_models_dir"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.downloads.get_default_class1_models_dir" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the absolute path to the default class1 models dir.</p>
<p>If environment variable MHCFLURRY_DEFAULT_CLASS1_MODELS_DIR is set to an
absolute path, return that path. If it’s set to a relative path (i.e. does
not start with /) then return that path taken to be relative to the mhcflurry
downloads dir.</p>
<p>If environment variable MHCFLURRY_DEFAULT_CLASS1_MODELS_DIR is NOT set,
then return the path to downloaded models in the “models_class1” download.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>test_exists</strong> : boolean, optional</p>
<blockquote>
<div><p>Whether to raise an exception of the path does not exist</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last"><strong>string</strong> : absolute path</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.downloads.get_current_release_downloads">
<code class="descclassname">mhcflurry.downloads.</code><code class="descname">get_current_release_downloads</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/downloads.html#get_current_release_downloads"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.downloads.get_current_release_downloads" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a dict of all available downloads in the current release.</p>
<p>The dict keys are the names of the downloads. The values are a dict
with two entries:</p>
<dl class="docutils">
<dt>downloaded <span class="classifier-delimiter">:</span> <span class="classifier">bool</span></dt>
<dd>Whether the download is currently available locally</dd>
<dt>metadata <span class="classifier-delimiter">:</span> <span class="classifier">dict</span></dt>
<dd>Info about the download from downloads.yml such as URL</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.downloads.get_path">
<code class="descclassname">mhcflurry.downloads.</code><code class="descname">get_path</code><span class="sig-paren">(</span><em>download_name</em>, <em>filename=''</em>, <em>test_exists=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/downloads.html#get_path"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.downloads.get_path" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the local path to a file in a MHCflurry download</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>download_name</strong> : string</p>
<p><strong>filename</strong> : string</p>
<blockquote>
<div><p>Relative path within the download to the file of interest</p>
</div></blockquote>
<p><strong>test_exists</strong> : boolean</p>
<blockquote>
<div><p>If True (default) throw an error telling the user how to download the
data if the file does not exist</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">string giving local absolute path</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.downloads.configure">
<code class="descclassname">mhcflurry.downloads.</code><code class="descname">configure</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/downloads.html#configure"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.downloads.configure" title="Permalink to this definition">¶</a></dt>
<dd><p>Setup various global variables based on environment variables.</p>
</dd></dl>

</div>
<div class="section" id="module-mhcflurry.downloads_command">
<span id="mhcflurry-downloads-command-module"></span><h2>mhcflurry.downloads_command module<a class="headerlink" href="#module-mhcflurry.downloads_command" title="Permalink to this headline">¶</a></h2>
<p>Download MHCflurry released datasets and trained models.</p>
<p>Examples</p>
<dl class="docutils">
<dt>Fetch the default downloads:</dt>
<dd>$ mhcflurry-downloads fetch</dd>
<dt>Fetch a specific download:</dt>
<dd>$ mhcflurry-downloads fetch data_kim2014</dd>
<dt>Get the path to a download:</dt>
<dd>$ mhcflurry-downloads path data_kim2014</dd>
<dt>Summarize available and fetched downloads:</dt>
<dd>$ mhcflurry-downloads info</dd>
</dl>
<dl class="function">
<dt id="mhcflurry.downloads_command.run">
<code class="descclassname">mhcflurry.downloads_command.</code><code class="descname">run</code><span class="sig-paren">(</span><em>argv=['-b', 'html', '-d', '_build/doctrees', '.', '_build/html']</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/downloads_command.html#run"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.downloads_command.run" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="mhcflurry.downloads_command.mkdir_p">
<code class="descclassname">mhcflurry.downloads_command.</code><code class="descname">mkdir_p</code><span class="sig-paren">(</span><em>path</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/downloads_command.html#mkdir_p"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.downloads_command.mkdir_p" title="Permalink to this definition">¶</a></dt>
<dd><p>Make directories as needed, similar to mkdir -p in a shell.</p>
<p>From:
<a class="reference external" href="http://stackoverflow.com/questions/600268/mkdir-p-functionality-in-python">http://stackoverflow.com/questions/600268/mkdir-p-functionality-in-python</a></p>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.downloads_command.yes_no">
<code class="descclassname">mhcflurry.downloads_command.</code><code class="descname">yes_no</code><span class="sig-paren">(</span><em>boolean</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/downloads_command.html#yes_no"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.downloads_command.yes_no" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="mhcflurry.downloads_command.TqdmUpTo">
<em class="property">class </em><code class="descclassname">mhcflurry.downloads_command.</code><code class="descname">TqdmUpTo</code><span class="sig-paren">(</span><em>iterable=None</em>, <em>desc=None</em>, <em>total=None</em>, <em>leave=True</em>, <em>file=None</em>, <em>ncols=None</em>, <em>mininterval=0.1</em>, <em>maxinterval=10.0</em>, <em>miniters=None</em>, <em>ascii=None</em>, <em>disable=False</em>, <em>unit='it'</em>, <em>unit_scale=False</em>, <em>dynamic_ncols=False</em>, <em>smoothing=0.3</em>, <em>bar_format=None</em>, <em>initial=0</em>, <em>position=None</em>, <em>postfix=None</em>, <em>unit_divisor=1000</em>, <em>gui=False</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/downloads_command.html#TqdmUpTo"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.downloads_command.TqdmUpTo" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">tqdm._tqdm.tqdm</span></code></p>
<p>Provides <code class="xref py py-obj docutils literal"><span class="pre">update_to(n)</span></code> which uses <code class="xref py py-obj docutils literal"><span class="pre">tqdm.update(delta_n)</span></code>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>iterable</strong> : iterable, optional</p>
<blockquote>
<div><p>Iterable to decorate with a progressbar.
Leave blank to manually manage the updates.</p>
</div></blockquote>
<p><strong>desc</strong> : str, optional</p>
<blockquote>
<div><p>Prefix for the progressbar.</p>
</div></blockquote>
<p><strong>total</strong> : int, optional</p>
<blockquote>
<div><p>The number of expected iterations. If unspecified,
len(iterable) is used if possible. As a last resort, only basic
progress statistics are displayed (no ETA, no progressbar).
If <code class="xref py py-obj docutils literal"><span class="pre">gui</span></code> is True and this parameter needs subsequent updating,
specify an initial arbitrary large positive integer,
e.g. int(9e9).</p>
</div></blockquote>
<p><strong>leave</strong> : bool, optional</p>
<blockquote>
<div><p>If [default: True], keeps all traces of the progressbar
upon termination of iteration.</p>
</div></blockquote>
<p><strong>file</strong> : <code class="xref py py-obj docutils literal"><span class="pre">io.TextIOWrapper</span></code> or <code class="xref py py-obj docutils literal"><span class="pre">io.StringIO</span></code>, optional</p>
<blockquote>
<div><p>Specifies where to output the progress messages
(default: sys.stderr). Uses <code class="xref py py-obj docutils literal"><span class="pre">file.write(str)</span></code> and <code class="xref py py-obj docutils literal"><span class="pre">file.flush()</span></code>
methods.</p>
</div></blockquote>
<p><strong>ncols</strong> : int, optional</p>
<blockquote>
<div><p>The width of the entire output message. If specified,
dynamically resizes the progressbar to stay within this bound.
If unspecified, attempts to use environment width. The
fallback is a meter width of 10 and no limit for the counter and
statistics. If 0, will not print any meter (only stats).</p>
</div></blockquote>
<p><strong>mininterval</strong> : float, optional</p>
<blockquote>
<div><p>Minimum progress display update interval, in seconds [default: 0.1].</p>
</div></blockquote>
<p><strong>maxinterval</strong> : float, optional</p>
<blockquote>
<div><p>Maximum progress display update interval, in seconds [default: 10].
Automatically adjusts <code class="xref py py-obj docutils literal"><span class="pre">miniters</span></code> to correspond to <code class="xref py py-obj docutils literal"><span class="pre">mininterval</span></code>
after long display update lag. Only works if <code class="xref py py-obj docutils literal"><span class="pre">dynamic_miniters</span></code>
or monitor thread is enabled.</p>
</div></blockquote>
<p><strong>miniters</strong> : int, optional</p>
<blockquote>
<div><p>Minimum progress display update interval, in iterations.
If 0 and <code class="xref py py-obj docutils literal"><span class="pre">dynamic_miniters</span></code>, will automatically adjust to equal
<code class="xref py py-obj docutils literal"><span class="pre">mininterval</span></code> (more CPU efficient, good for tight loops).
If &gt; 0, will skip display of specified number of iterations.
Tweak this and <code class="xref py py-obj docutils literal"><span class="pre">mininterval</span></code> to get very efficient loops.
If your progress is erratic with both fast and slow iterations
(network, skipping items, etc) you should set miniters=1.</p>
</div></blockquote>
<p><strong>ascii</strong> : bool, optional</p>
<blockquote>
<div><p>If unspecified or False, use unicode (smooth blocks) to fill
the meter. The fallback is to use ASCII characters <code class="xref py py-obj docutils literal"><span class="pre">1-9</span> <span class="pre">#</span></code>.</p>
</div></blockquote>
<p><strong>disable</strong> : bool, optional</p>
<blockquote>
<div><p>Whether to disable the entire progressbar wrapper
[default: False]. If set to None, disable on non-TTY.</p>
</div></blockquote>
<p><strong>unit</strong> : str, optional</p>
<blockquote>
<div><p>String that will be used to define the unit of each iteration
[default: it].</p>
</div></blockquote>
<p><strong>unit_scale</strong> : bool or int or float, optional</p>
<blockquote>
<div><p>If 1 or True, the number of iterations will be reduced/scaled
automatically and a metric prefix following the
International System of Units standard will be added
(kilo, mega, etc.) [default: False]. If any other non-zero
number, will scale <code class="xref py py-obj docutils literal"><span class="pre">total</span></code> and <code class="xref py py-obj docutils literal"><span class="pre">n</span></code>.</p>
</div></blockquote>
<p><strong>dynamic_ncols</strong> : bool, optional</p>
<blockquote>
<div><p>If set, constantly alters <code class="xref py py-obj docutils literal"><span class="pre">ncols</span></code> to the environment (allowing
for window resizes) [default: False].</p>
</div></blockquote>
<p><strong>smoothing</strong> : float, optional</p>
<blockquote>
<div><p>Exponential moving average smoothing factor for speed estimates
(ignored in GUI mode). Ranges from 0 (average speed) to 1
(current/instantaneous speed) [default: 0.3].</p>
</div></blockquote>
<p><strong>bar_format</strong> : str, optional</p>
<blockquote>
<div><p>Specify a custom bar string formatting. May impact performance.
If unspecified, will use ‘{l_bar}{bar}{r_bar}’, where l_bar is
‘{desc}: {percentage:3.0f}%|’ and r_bar is
‘| {n_fmt}/{total_fmt} [{elapsed}&lt;{remaining}, {rate_fmt}]’
Possible vars: bar, n, n_fmt, total, total_fmt, percentage,
rate, rate_fmt, elapsed, remaining, l_bar, r_bar, desc.
Note that a trailing “: ” is automatically removed after {desc}
if the latter is empty.</p>
</div></blockquote>
<p><strong>initial</strong> : int, optional</p>
<blockquote>
<div><p>The initial counter value. Useful when restarting a progress
bar [default: 0].</p>
</div></blockquote>
<p><strong>position</strong> : int, optional</p>
<blockquote>
<div><p>Specify the line offset to print this bar (starting from 0)
Automatic if unspecified.
Useful to manage multiple bars at once (eg, from threads).</p>
</div></blockquote>
<p><strong>postfix</strong> : dict, optional</p>
<blockquote>
<div><p>Specify additional stats to display at the end of the bar.
Note: postfix is a dict ({‘key’: value} pairs) for this method,
not a string.</p>
</div></blockquote>
<p><strong>unit_divisor</strong> : float, optional</p>
<blockquote>
<div><p>[default: 1000], ignored unless <code class="xref py py-obj docutils literal"><span class="pre">unit_scale</span></code> is True.</p>
</div></blockquote>
<p><strong>gui</strong> : bool, optional</p>
<blockquote>
<div><p>WARNING: internal parameter - do not use.
Use tqdm_gui(…) instead. If set, will attempt to use
matplotlib animations for a graphical output [default: False].</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last"><strong>out</strong> : decorated iterator.</p>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="mhcflurry.downloads_command.TqdmUpTo.update_to">
<code class="descname">update_to</code><span class="sig-paren">(</span><em>b=1</em>, <em>bsize=1</em>, <em>tsize=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/downloads_command.html#TqdmUpTo.update_to"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.downloads_command.TqdmUpTo.update_to" title="Permalink to this definition">¶</a></dt>
<dd><dl class="docutils">
<dt>b <span class="classifier-delimiter">:</span> <span class="classifier">int, optional</span></dt>
<dd>Number of blocks transferred so far [default: 1].</dd>
<dt>bsize <span class="classifier-delimiter">:</span> <span class="classifier">int, optional</span></dt>
<dd>Size of each block (in tqdm units) [default: 1].</dd>
<dt>tsize <span class="classifier-delimiter">:</span> <span class="classifier">int, optional</span></dt>
<dd>Total size (in tqdm units). If [default: None] remains unchanged.</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="mhcflurry.downloads_command.fetch_subcommand">
<code class="descclassname">mhcflurry.downloads_command.</code><code class="descname">fetch_subcommand</code><span class="sig-paren">(</span><em>args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/downloads_command.html#fetch_subcommand"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.downloads_command.fetch_subcommand" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="mhcflurry.downloads_command.info_subcommand">
<code class="descclassname">mhcflurry.downloads_command.</code><code class="descname">info_subcommand</code><span class="sig-paren">(</span><em>args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/downloads_command.html#info_subcommand"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.downloads_command.info_subcommand" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="mhcflurry.downloads_command.path_subcommand">
<code class="descclassname">mhcflurry.downloads_command.</code><code class="descname">path_subcommand</code><span class="sig-paren">(</span><em>args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/downloads_command.html#path_subcommand"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.downloads_command.path_subcommand" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</div>
<div class="section" id="module-mhcflurry.encodable_sequences">
<span id="mhcflurry-encodable-sequences-module"></span><h2>mhcflurry.encodable_sequences module<a class="headerlink" href="#module-mhcflurry.encodable_sequences" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="mhcflurry.encodable_sequences.EncodableSequences">
<em class="property">class </em><code class="descclassname">mhcflurry.encodable_sequences.</code><code class="descname">EncodableSequences</code><span class="sig-paren">(</span><em>sequences</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/encodable_sequences.html#EncodableSequences"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.encodable_sequences.EncodableSequences" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<p>Sequences of amino acids.</p>
<p>This class caches various encodings of a list of sequences.</p>
<dl class="attribute">
<dt id="mhcflurry.encodable_sequences.EncodableSequences.unknown_character">
<code class="descname">unknown_character</code><em class="property"> = 'X'</em><a class="headerlink" href="#mhcflurry.encodable_sequences.EncodableSequences.unknown_character" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="classmethod">
<dt id="mhcflurry.encodable_sequences.EncodableSequences.create">
<em class="property">classmethod </em><code class="descname">create</code><span class="sig-paren">(</span><em>klass</em>, <em>sequences</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/encodable_sequences.html#EncodableSequences.create"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.encodable_sequences.EncodableSequences.create" title="Permalink to this definition">¶</a></dt>
<dd><p>Factory that returns an EncodableSequences given a list of
strings. As a convenience, you can also pass it an EncodableSequences
instance, in which case the object is returned unchanged.</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.encodable_sequences.EncodableSequences.variable_length_to_fixed_length_categorical">
<code class="descname">variable_length_to_fixed_length_categorical</code><span class="sig-paren">(</span><em>left_edge=4</em>, <em>right_edge=4</em>, <em>max_length=15</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/encodable_sequences.html#EncodableSequences.variable_length_to_fixed_length_categorical"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.encodable_sequences.EncodableSequences.variable_length_to_fixed_length_categorical" title="Permalink to this definition">¶</a></dt>
<dd><p>Encode variable-length sequences using a fixed-length encoding designed
for preserving the anchor positions of class I peptides.</p>
<p>The sequences must be of length at least left_edge + right_edge, and at
most max_length.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>left_edge</strong> : int, size of fixed-position left side</p>
<p><strong>right_edge</strong> : int, size of the fixed-position right side</p>
<p><strong>max_length</strong> : sequence length of the resulting encoding</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">numpy.array of integers with shape (num sequences, max_length)</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.encodable_sequences.EncodableSequences.variable_length_to_fixed_length_vector_encoding">
<code class="descname">variable_length_to_fixed_length_vector_encoding</code><span class="sig-paren">(</span><em>vector_encoding_name</em>, <em>left_edge=4</em>, <em>right_edge=4</em>, <em>max_length=15</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/encodable_sequences.html#EncodableSequences.variable_length_to_fixed_length_vector_encoding"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.encodable_sequences.EncodableSequences.variable_length_to_fixed_length_vector_encoding" title="Permalink to this definition">¶</a></dt>
<dd><p>Encode variable-length sequences using a fixed-length encoding designed
for preserving the anchor positions of class I peptides.</p>
<p>The sequences must be of length at least left_edge + right_edge, and at
most max_length.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>vector_encoding_name</strong> : string</p>
<blockquote>
<div><p>How to represent amino acids.
One of “BLOSUM62”, “one-hot”, etc. Full list of supported vector
encodings is given by available_vector_encodings().</p>
</div></blockquote>
<p><strong>left_edge</strong> : int, size of fixed-position left side</p>
<p><strong>right_edge</strong> : int, size of the fixed-position right side</p>
<p><strong>max_length</strong> : sequence length of the resulting encoding</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">numpy.array with shape (num sequences, max_length, m) where m is</p>
<p class="last">vector_encoding_length(vector_encoding_name)</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="classmethod">
<dt id="mhcflurry.encodable_sequences.EncodableSequences.sequences_to_fixed_length_index_encoded_array">
<em class="property">classmethod </em><code class="descname">sequences_to_fixed_length_index_encoded_array</code><span class="sig-paren">(</span><em>klass</em>, <em>sequences</em>, <em>left_edge=4</em>, <em>right_edge=4</em>, <em>max_length=15</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/encodable_sequences.html#EncodableSequences.sequences_to_fixed_length_index_encoded_array"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.encodable_sequences.EncodableSequences.sequences_to_fixed_length_index_encoded_array" title="Permalink to this definition">¶</a></dt>
<dd><p>Transform a sequence of strings, where each string is of length at least
left_edge + right_edge and at most max_length into strings of length
max_length using a scheme designed to preserve the anchor positions of
class I peptides.</p>
<p>The first left_edge characters in the input always map to the first
left_edge characters in the output. Similarly for the last right_edge
characters. The middle characters are filled in based on the length,
with the X character filling in the blanks.</p>
<p>For example, using defaults:</p>
<p>AAAACDDDD -&gt; AAAAXXXCXXXDDDD</p>
<p>The strings are also converted to int categorical amino acid indices.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>sequence</strong> : string</p>
<p><strong>left_edge</strong> : int</p>
<p><strong>right_edge</strong> : int</p>
<p><strong>max_length</strong> : int</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">numpy array of shape (len(sequences), max_length) and dtype int</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-mhcflurry.ensemble_centrality">
<span id="mhcflurry-ensemble-centrality-module"></span><h2>mhcflurry.ensemble_centrality module<a class="headerlink" href="#module-mhcflurry.ensemble_centrality" title="Permalink to this headline">¶</a></h2>
<p>Measures of centrality (e.g. mean) used to combine predictions across an
ensemble. The input to these functions are log affinities, and they are expected
to return a centrality measure also in log-space.</p>
<dl class="function">
<dt id="mhcflurry.ensemble_centrality.robust_mean">
<code class="descclassname">mhcflurry.ensemble_centrality.</code><code class="descname">robust_mean</code><span class="sig-paren">(</span><em>log_values</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/ensemble_centrality.html#robust_mean"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.ensemble_centrality.robust_mean" title="Permalink to this definition">¶</a></dt>
<dd><p>Mean of values falling within the 25-75 percentiles.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>log_values</strong> : 2-d numpy.array</p>
<blockquote>
<div><p>Center is computed along the second axis (i.e. per row).</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last"><strong>center</strong> : numpy.array of length log_values.shape[1]</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="module-mhcflurry.hyperparameters">
<span id="mhcflurry-hyperparameters-module"></span><h2>mhcflurry.hyperparameters module<a class="headerlink" href="#module-mhcflurry.hyperparameters" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="mhcflurry.hyperparameters.HyperparameterDefaults">
<em class="property">class </em><code class="descclassname">mhcflurry.hyperparameters.</code><code class="descname">HyperparameterDefaults</code><span class="sig-paren">(</span><em>**defaults</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/hyperparameters.html#HyperparameterDefaults"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.hyperparameters.HyperparameterDefaults" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<p>Class for managing hyperparameters. Thin wrapper around a dict.</p>
<p>Instances of this class are a specification of the hyperparameters
<em>supported</em> by a model and their defaults. The particular
hyperparameter settings to be used, for example, to train a model
are kept in plain dicts.</p>
<dl class="method">
<dt id="mhcflurry.hyperparameters.HyperparameterDefaults.extend">
<code class="descname">extend</code><span class="sig-paren">(</span><em>other</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/hyperparameters.html#HyperparameterDefaults.extend"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.hyperparameters.HyperparameterDefaults.extend" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a new HyperparameterDefaults instance containing the
hyperparameters from the current instance combined with
those from other.</p>
<p>It is an error if self and other have any hyperparameters in
common.</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.hyperparameters.HyperparameterDefaults.with_defaults">
<code class="descname">with_defaults</code><span class="sig-paren">(</span><em>obj</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/hyperparameters.html#HyperparameterDefaults.with_defaults"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.hyperparameters.HyperparameterDefaults.with_defaults" title="Permalink to this definition">¶</a></dt>
<dd><p>Given a dict of hyperparameter settings, return a dict containing
those settings augmented by the defaults for any keys missing from
the dict.</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.hyperparameters.HyperparameterDefaults.subselect">
<code class="descname">subselect</code><span class="sig-paren">(</span><em>obj</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/hyperparameters.html#HyperparameterDefaults.subselect"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.hyperparameters.HyperparameterDefaults.subselect" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter a dict of hyperparameter settings to only those keys defined
in this HyperparameterDefaults  .</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.hyperparameters.HyperparameterDefaults.check_valid_keys">
<code class="descname">check_valid_keys</code><span class="sig-paren">(</span><em>obj</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/hyperparameters.html#HyperparameterDefaults.check_valid_keys"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.hyperparameters.HyperparameterDefaults.check_valid_keys" title="Permalink to this definition">¶</a></dt>
<dd><p>Given a dict of hyperparameter settings, throw an exception if any
keys are not defined in this HyperparameterDefaults instance.</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.hyperparameters.HyperparameterDefaults.models_grid">
<code class="descname">models_grid</code><span class="sig-paren">(</span><em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/hyperparameters.html#HyperparameterDefaults.models_grid"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.hyperparameters.HyperparameterDefaults.models_grid" title="Permalink to this definition">¶</a></dt>
<dd><p>Make a grid of models by taking the cartesian product of all specified
model parameter lists.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>The valid kwarg parameters are the entries of this</strong></p>
<p><strong>HyperparameterDefaults instance. Each parameter must be a list</strong></p>
<p><strong>giving the values to search across.</strong></p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">list of dict giving the parameters for each model. The length of the</p>
<p class="last">list is the product of the lengths of the input lists.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-mhcflurry.percent_rank_transform">
<span id="mhcflurry-percent-rank-transform-module"></span><h2>mhcflurry.percent_rank_transform module<a class="headerlink" href="#module-mhcflurry.percent_rank_transform" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="mhcflurry.percent_rank_transform.PercentRankTransform">
<em class="property">class </em><code class="descclassname">mhcflurry.percent_rank_transform.</code><code class="descname">PercentRankTransform</code><a class="reference internal" href="_modules/mhcflurry/percent_rank_transform.html#PercentRankTransform"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.percent_rank_transform.PercentRankTransform" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<p>Transform arbitrary values into percent ranks.</p>
<dl class="method">
<dt id="mhcflurry.percent_rank_transform.PercentRankTransform.fit">
<code class="descname">fit</code><span class="sig-paren">(</span><em>values</em>, <em>bins</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/percent_rank_transform.html#PercentRankTransform.fit"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.percent_rank_transform.PercentRankTransform.fit" title="Permalink to this definition">¶</a></dt>
<dd><p>Fit the transform using the given values (in our case ic50s).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>values</strong> : ic50 values</p>
<p><strong>bins</strong> : bins for the cumulative distribution function</p>
<blockquote class="last">
<div><p>Anything that can be passed to numpy.histogram’s “bins” argument
can be used here.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.percent_rank_transform.PercentRankTransform.transform">
<code class="descname">transform</code><span class="sig-paren">(</span><em>values</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/percent_rank_transform.html#PercentRankTransform.transform"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.percent_rank_transform.PercentRankTransform.transform" title="Permalink to this definition">¶</a></dt>
<dd><p>Return percent ranks (range [0, 100]) for the given values.</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.percent_rank_transform.PercentRankTransform.to_series">
<code class="descname">to_series</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/percent_rank_transform.html#PercentRankTransform.to_series"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.percent_rank_transform.PercentRankTransform.to_series" title="Permalink to this definition">¶</a></dt>
<dd><p>Serialize the fit to a pandas.Series.</p>
<p>The index on the series gives the bin edges and the valeus give the CDF.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">pandas.Series</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="staticmethod">
<dt id="mhcflurry.percent_rank_transform.PercentRankTransform.from_series">
<em class="property">static </em><code class="descname">from_series</code><span class="sig-paren">(</span><em>series</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/percent_rank_transform.html#PercentRankTransform.from_series"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.percent_rank_transform.PercentRankTransform.from_series" title="Permalink to this definition">¶</a></dt>
<dd><p>Deseralize a PercentRankTransform the given pandas.Series, as returned
by <a class="reference internal" href="#mhcflurry.percent_rank_transform.PercentRankTransform.to_series" title="mhcflurry.percent_rank_transform.PercentRankTransform.to_series"><code class="xref py py-obj docutils literal"><span class="pre">to_series()</span></code></a>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>series</strong> : pandas.Series</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">PercentRankTransform</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-mhcflurry.predict_command">
<span id="mhcflurry-predict-command-module"></span><h2>mhcflurry.predict_command module<a class="headerlink" href="#module-mhcflurry.predict_command" title="Permalink to this headline">¶</a></h2>
<p>Run MHCflurry predictor on specified peptide/allele pairs.</p>
<p>Examples:</p>
<p>Write a CSV file containing the contents of INPUT.csv plus an
additional column giving MHCflurry binding affinity predictions:</p>
<blockquote>
<div>$ mhcflurry-predict INPUT.csv –out RESULT.csv</div></blockquote>
<p>The input CSV file is expected to contain columns <code class="docutils literal"><span class="pre">allele</span></code> and <code class="docutils literal"><span class="pre">peptide</span></code>.
The predictions are written to a column called <code class="docutils literal"><span class="pre">mhcflurry_prediction</span></code>.
These default column names may be changed with the <code class="xref py py-obj docutils literal"><span class="pre">--allele-column</span></code>,
<code class="xref py py-obj docutils literal"><span class="pre">--peptide-column</span></code>, and <code class="xref py py-obj docutils literal"><span class="pre">--prediction-column</span></code> options.</p>
<p>If <code class="xref py py-obj docutils literal"><span class="pre">--out</span></code> is not specified, results are written to standard out.</p>
<p>You can also run on alleles and peptides specified on the commandline, in
which case predictions are written for all combinations of alleles and
peptides:</p>
<blockquote>
<div>$ mhcflurry-predict –alleles HLA-A0201 H-2Kb –peptides SIINFEKL DENDREKLLL</div></blockquote>
<dl class="function">
<dt id="mhcflurry.predict_command.run">
<code class="descclassname">mhcflurry.predict_command.</code><code class="descname">run</code><span class="sig-paren">(</span><em>argv=['-b', 'html', '-d', '_build/doctrees', '.', '_build/html']</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/predict_command.html#run"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.predict_command.run" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</div>
<div class="section" id="module-mhcflurry.regression_target">
<span id="mhcflurry-regression-target-module"></span><h2>mhcflurry.regression_target module<a class="headerlink" href="#module-mhcflurry.regression_target" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="mhcflurry.regression_target.from_ic50">
<code class="descclassname">mhcflurry.regression_target.</code><code class="descname">from_ic50</code><span class="sig-paren">(</span><em>ic50</em>, <em>max_ic50=50000.0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/regression_target.html#from_ic50"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.regression_target.from_ic50" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert ic50s to regression targets in the range [0.0, 1.0].</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>ic50</strong> : numpy.array of float</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">numpy.array of float</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.regression_target.to_ic50">
<code class="descclassname">mhcflurry.regression_target.</code><code class="descname">to_ic50</code><span class="sig-paren">(</span><em>x</em>, <em>max_ic50=50000.0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/regression_target.html#to_ic50"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.regression_target.to_ic50" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert regression targets in the range [0.0, 1.0] to ic50s in the range
[0, 50000.0].</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>x</strong> : numpy.array of float</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">numpy.array of float</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="module-mhcflurry.scoring">
<span id="mhcflurry-scoring-module"></span><h2>mhcflurry.scoring module<a class="headerlink" href="#module-mhcflurry.scoring" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="mhcflurry.scoring.make_scores">
<code class="descclassname">mhcflurry.scoring.</code><code class="descname">make_scores</code><span class="sig-paren">(</span><em>ic50_y</em>, <em>ic50_y_pred</em>, <em>sample_weight=None</em>, <em>threshold_nm=500</em>, <em>max_ic50=50000</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/scoring.html#make_scores"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.scoring.make_scores" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate AUC, F1, and Kendall Tau scores.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>ic50_y</strong> : float list</p>
<blockquote>
<div><p>true IC50s (i.e. affinities)</p>
</div></blockquote>
<p><strong>ic50_y_pred</strong> : float list</p>
<blockquote>
<div><p>predicted IC50s</p>
</div></blockquote>
<p><strong>sample_weight</strong> : float list [optional]</p>
<p><strong>threshold_nm</strong> : float [optional]</p>
<p><strong>max_ic50</strong> : float [optional]</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">dict with entries “auc”, “f1”, “tau”</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="module-mhcflurry.train_allele_specific_models_command">
<span id="mhcflurry-train-allele-specific-models-command-module"></span><h2>mhcflurry.train_allele_specific_models_command module<a class="headerlink" href="#module-mhcflurry.train_allele_specific_models_command" title="Permalink to this headline">¶</a></h2>
<p>Train Class1 single allele models.</p>
<dl class="function">
<dt id="mhcflurry.train_allele_specific_models_command.run">
<code class="descclassname">mhcflurry.train_allele_specific_models_command.</code><code class="descname">run</code><span class="sig-paren">(</span><em>argv=['-b', 'html', '-d', '_build/doctrees', '.', '_build/html']</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/train_allele_specific_models_command.html#run"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.train_allele_specific_models_command.run" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="mhcflurry.train_allele_specific_models_command.train_model_entrypoint">
<code class="descclassname">mhcflurry.train_allele_specific_models_command.</code><code class="descname">train_model_entrypoint</code><span class="sig-paren">(</span><em>item</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/train_allele_specific_models_command.html#train_model_entrypoint"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.train_allele_specific_models_command.train_model_entrypoint" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="mhcflurry.train_allele_specific_models_command.train_model">
<code class="descclassname">mhcflurry.train_allele_specific_models_command.</code><code class="descname">train_model</code><span class="sig-paren">(</span><em>model_group</em>, <em>n_models</em>, <em>allele_num</em>, <em>n_alleles</em>, <em>hyperparameter_set_num</em>, <em>num_hyperparameter_sets</em>, <em>allele</em>, <em>data</em>, <em>hyperparameters</em>, <em>verbose</em>, <em>predictor</em>, <em>save_to</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/train_allele_specific_models_command.html#train_model"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.train_allele_specific_models_command.train_model" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="mhcflurry.train_allele_specific_models_command.calibrate_percentile_ranks">
<code class="descclassname">mhcflurry.train_allele_specific_models_command.</code><code class="descname">calibrate_percentile_ranks</code><span class="sig-paren">(</span><em>allele</em>, <em>predictor</em>, <em>peptides=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/train_allele_specific_models_command.html#calibrate_percentile_ranks"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.train_allele_specific_models_command.calibrate_percentile_ranks" title="Permalink to this definition">¶</a></dt>
<dd><p>Private helper function.</p>
</dd></dl>

</div>
<div class="section" id="module-mhcflurry.version">
<span id="mhcflurry-version-module"></span><h2>mhcflurry.version module<a class="headerlink" href="#module-mhcflurry.version" title="Permalink to this headline">¶</a></h2>
</div>
</div>


           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
      
        <a href="commandline_tools.html" class="btn btn-neutral" title="Command-line reference" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, Timothy O&#39;Donnell.
      Last updated on Feb 06, 2018.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'1.1.0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>