

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Command-line tutorial &mdash; MHCflurry 2.0.0 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Python library tutorial" href="python_tutorial.html" />
    <link rel="prev" title="Introduction and setup" href="intro.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home" alt="Documentation Home"> MHCflurry
          

          
          </a>

          
            
            
              <div class="version">
                2.0.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="intro.html">Introduction and setup</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Command-line tutorial</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#downloading-models">Downloading models</a></li>
<li class="toctree-l2"><a class="reference internal" href="#generating-predictions">Generating predictions</a></li>
<li class="toctree-l2"><a class="reference internal" href="#using-the-older-allele-specific-models">Using the older, allele-specific models</a></li>
<li class="toctree-l2"><a class="reference internal" href="#scanning-protein-sequences-for-predicted-mhc-i-ligands">Scanning protein sequences for predicted MHC I ligands</a></li>
<li class="toctree-l2"><a class="reference internal" href="#fitting-your-own-models">Fitting your own models</a></li>
<li class="toctree-l2"><a class="reference internal" href="#environment-variables">Environment variables</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="python_tutorial.html">Python library tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="commandline_tools.html">Command-line reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="api.html">API Documentation</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">MHCflurry</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Command-line tutorial</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/commandline_tutorial.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="command-line-tutorial">
<span id="commandline-tutorial"></span><h1>Command-line tutorial<a class="headerlink" href="#command-line-tutorial" title="Permalink to this headline">¶</a></h1>
<div class="section" id="downloading-models">
<span id="downloading"></span><h2>Downloading models<a class="headerlink" href="#downloading-models" title="Permalink to this headline">¶</a></h2>
<p>Most users will use pre-trained MHCflurry models that we release. These models
are distributed separately from the pip package and may be downloaded with the
<a class="reference internal" href="commandline_tools.html#mhcflurry-downloads"><span class="std std-ref">mhcflurry-downloads</span></a> tool:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ mhcflurry-downloads fetch models_class1_presentation
</pre></div>
</div>
<p>Files downloaded with <a class="reference internal" href="commandline_tools.html#mhcflurry-downloads"><span class="std std-ref">mhcflurry-downloads</span></a> are stored in a platform-specific
directory. To get the path to downloaded data, you can use:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ mhcflurry-downloads path models_class1_presentation
/Users/tim/Library/Application Support/mhcflurry/4/2.0.0/models_class1_presentation/
</pre></div>
</div>
<p>We also release a number of other “downloads,” such as curated training data and some
experimental models. To see what’s available and what you have downloaded, run
<code class="docutils literal notranslate"><span class="pre">mhcflurry-downloads</span> <span class="pre">info</span></code>.</p>
<p>Most users will only need <code class="docutils literal notranslate"><span class="pre">models_class1_presentation</span></code>, however, as the
presentation predictor includes a peptide / MHC I binding affinity (BA) predictor
as well as an antigen processing (AP) predictor.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The code we use for <em>generating</em> the downloads is in the
<code class="docutils literal notranslate"><span class="pre">downloads_generation</span></code> directory in the repository (<a class="reference external" href="https://github.com/openvax/mhcflurry/tree/master/downloads-generation">https://github.com/openvax/mhcflurry/tree/master/downloads-generation</a>)</p>
</div>
</div>
<div class="section" id="generating-predictions">
<h2>Generating predictions<a class="headerlink" href="#generating-predictions" title="Permalink to this headline">¶</a></h2>
<p>The <a class="reference internal" href="commandline_tools.html#mhcflurry-predict"><span class="std std-ref">mhcflurry-predict</span></a> command generates predictions for individual peptides
(see the next section for how to scan protein sequences for epitopes). By
default it will use the pre-trained models you downloaded above. Other
models can be used by specifying the <code class="docutils literal notranslate"><span class="pre">--models</span></code> argument.</p>
<p>Running:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ mhcflurry-predict
    --alleles HLA-A0201 HLA-A0301
    --peptides SIINFEKL SIINFEKD SIINFEKQ
    --out /tmp/predictions.csv
Forcing tensorflow backend.
Predicting processing.
Predicting affinities.
Wrote: /tmp/predictions.csv
</pre></div>
</div>
<p>results in a file like this:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ cat /tmp/predictions.csv
allele,peptide,mhcflurry_affinity,mhcflurry_affinity_percentile,mhcflurry_processing_score,mhcflurry_presentation_score,mhcflurry_presentation_percentile
HLA-A0201,SIINFEKL,11927.158941532069,6.296000000000002,0.26470961794257164,0.020690210604074,11.6303260869565
HLA-A0201,SIINFEKD,30039.79614570185,41.398,0.024963302072137594,0.0036743984764636087,99.28660326086957
HLA-A0201,SIINFEKQ,28026.90550690346,30.226250000000004,0.06154933862853795,0.00447008680074548,62.74467391304348
HLA-A0301,SIINFEKL,29871.585202288912,23.19225,0.26470961794257164,0.008586191856042273,27.617391304347805
HLA-A0301,SIINFEKD,33158.802891745494,62.27625,0.024963302072137594,0.0033393306217936363,99.28660326086957
HLA-A0301,SIINFEKQ,29705.412189610724,21.8915,0.06154933862853795,0.004225574919955567,62.74467391304348
</pre></div>
</div>
<p>The binding affinity predictions are given as affinities (KD) in nM in the
<code class="docutils literal notranslate"><span class="pre">mhcflurry_affinity</span></code> column. Lower values indicate stronger binders. A commonly-used
threshold for peptides with a reasonable chance of being immunogenic is 500 nM.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">mhcflurry_affinity_percentile</span></code> gives the percentile of the affinity
prediction among a large number of random peptides tested on that allele (range
0 - 100). Lower is stronger. Two percent is a commonly-used threshold.</p>
<p>The last two columns give the antigen processing and presentation scores,
respectively. These range from 0 to 1 with higher values indicating more
favorable processing or presentation.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The processing predictor is experimental. It models allele-independent
effects that influence whether a
peptide will be detected in a mass spec experiment. The presentation score is
a simple logistic regression model that combines the (log) binding affinity
prediction with the processing score to give a composite prediction. The resulting
prediction may be useful for prioritizing potential epitopes, but no
thresholds have been established for what constitutes a “high enough”
presentation score.</p>
</div>
<p>In most cases you’ll want to specify the input as a CSV file instead of passing
peptides and alleles as commandline arguments. If you’re relying on the
processing or presentation scores, you may also want to pass the upstream and
downstream sequences of the peptides from their source proteins for potentially more
accurate cleavage prediction. See the <a class="reference internal" href="commandline_tools.html#mhcflurry-predict"><span class="std std-ref">mhcflurry-predict</span></a> docs.</p>
</div>
<div class="section" id="using-the-older-allele-specific-models">
<h2>Using the older, allele-specific models<a class="headerlink" href="#using-the-older-allele-specific-models" title="Permalink to this headline">¶</a></h2>
<p>Previous versions of MHCflurry (described in the 2018 paper) used models
trained on affinity measurements, one allele per model (i.e. allele-specific).
Mass spec datasets were incorporated in the model selection step.</p>
<p>These models are still available to use with the latest version of MHCflurry.
To download these predictors, run:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ mhcflurry-downloads fetch models_class1
</pre></div>
</div>
<p>and specify <code class="docutils literal notranslate"><span class="pre">--models</span></code> when you call <code class="docutils literal notranslate"><span class="pre">mhcflurry-predict</span></code>:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ mhcflurry-predict <span class="se">\</span>
    --alleles HLA-A0201 HLA-A0301 <span class="se">\</span>
    --peptides SIINFEKL SIINFEKD SIINFEKQ <span class="se">\</span>
    --models <span class="s2">&quot;</span><span class="k">$(</span>mhcflurry-downloads path models_class1<span class="k">)</span><span class="s2">/models&quot;</span>
    --out /tmp/predictions.csv
</pre></div>
</div>
</div>
<div class="section" id="scanning-protein-sequences-for-predicted-mhc-i-ligands">
<h2>Scanning protein sequences for predicted MHC I ligands<a class="headerlink" href="#scanning-protein-sequences-for-predicted-mhc-i-ligands" title="Permalink to this headline">¶</a></h2>
<p>Starting in version 1.6.0, MHCflurry supports scanning proteins for MHC-binding
peptides using the <code class="docutils literal notranslate"><span class="pre">mhcflurry-predict-scan</span></code> command.</p>
<p>We’ll generate predictions across <code class="docutils literal notranslate"><span class="pre">example.fasta</span></code>, a FASTA file with two short
sequences:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span><span class="n">protein1</span>
<span class="n">MSSSSTPVCPNGPGNCQV</span>
<span class="o">&gt;</span><span class="n">protein2</span>
<span class="n">MVENKRLLEGMEMIFGQVIPGA</span>
</pre></div>
</div>
<p>Here’s the <code class="docutils literal notranslate"><span class="pre">mhcflurry-predict-scan</span></code> invocation to scan the proteins for
binders to either of two MHC I genotypes (using a 100 nM threshold):</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ mhcflurry-predict-scan
    example.fasta
    --alleles
        HLA-A*02:01,HLA-A*03:01,HLA-B*57:01,HLA-B*45:01,HLA-C*02:02,HLA-C*07:02
        HLA-A*01:01,HLA-A*02:06,HLA-B*44:02,HLA-B*07:02,HLA-C*01:02,HLA-C*03:01
    --results-filtered affinity
    --threshold-affinity 100
Forcing tensorflow backend.
Guessed input file format: fasta
Read input fasta with 2 sequences
  sequence_id                sequence
0    protein1      MSSSSTPVCPNGPGNCQV
1    protein2  MVENKRLLEGMEMIFGQVIPGA
Predicting processing.
Predicting affinities.
sequence_name,pos,peptide,n_flank,c_flank,sample_name,affinity,best_allele,affinity_percentile,processing_score,presentation_score,presentation_percentile
protein2,5,RLLEGMEMI,MVENK,FGQVI,genotype_00,20.48976379349174,HLA-A*02:01,0.10225000000000008,0.6447581760585308,0.975528093093756,0.013097826086976738
protein2,5,RLLEGMEMI,MVENK,FGQVI,genotype_01,23.010403529711986,HLA-A*02:06,0.1372500000000001,0.6447581760585308,0.972681225756668,0.016576086956561653
protein2,12,MIFGQVIPGA,LEGME,,genotype_01,40.981431478935065,HLA-A*02:06,0.3232499999999999,0.8904485031962395,0.9810106587410801,0.0061684782608750766
protein2,4,KRLLEGMEM,MVEN,IFGQV,genotype_00,49.677106614054956,HLA-C*07:02,0.09350000000000007,0.12594258785247803,0.7006409502993601,0.43820652173911867
protein2,12,MIFGQVIPGA,LEGME,,genotype_00,62.967656603774884,HLA-A*02:01,0.4323749999999997,0.8904485031962395,0.9714322924052166,0.017663043478293616
protein2,10,MEMIFGQVI,RLLEG,PGA,genotype_00,82.40893767472276,HLA-B*45:01,0.18000000000000002,0.3939148336648941,0.7984213046411495,0.2809510869564775
protein1,8,CPNGPGNCQV,SSTPV,,genotype_01,87.84821064140868,HLA-B*07:02,0.297,0.057573866099119186,0.5088251513715745,0.8024184782607335
</pre></div>
</div>
<p>See the <a class="reference internal" href="commandline_tools.html#mhcflurry-predict-scan"><span class="std std-ref">mhcflurry-predict-scan</span></a> docs for more options.</p>
</div>
<div class="section" id="fitting-your-own-models">
<h2>Fitting your own models<a class="headerlink" href="#fitting-your-own-models" title="Permalink to this headline">¶</a></h2>
<p>If you have your own data and want to fit your own MHCflurry models, you have
a few options. If you have data for only one or a few MHC I alleles, the best
approach is to use the
<a class="reference internal" href="commandline_tools.html#mhcflurry-class1-train-allele-specific-models"><span class="std std-ref">mhcflurry-class1-train-allele-specific-models</span></a> command to fit an
“allele-specific” predictor, in which separate neural networks are used for
each allele.</p>
<p>To call <a class="reference internal" href="commandline_tools.html#mhcflurry-class1-train-allele-specific-models"><span class="std std-ref">mhcflurry-class1-train-allele-specific-models</span></a> you’ll need some
training data. The data we use for our released predictors can be downloaded with
<a class="reference internal" href="commandline_tools.html#mhcflurry-downloads"><span class="std std-ref">mhcflurry-downloads</span></a>:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ mhcflurry-downloads fetch data_curated
</pre></div>
</div>
<p>It looks like this:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ bzcat &quot;$(mhcflurry-downloads path data_curated)/curated_training_data.csv.bz2&quot; | head -n 3
allele,peptide,measurement_value,measurement_inequality,measurement_type,measurement_kind,measurement_source,original_allele
BoLA-1*21:01,AENDTLVVSV,7817.0,=,quantitative,affinity,Barlow - purified MHC/competitive/fluorescence,BoLA-1*02101
BoLA-1*21:01,NQFNGGCLLV,1086.0,=,quantitative,affinity,Barlow - purified MHC/direct/fluorescence,BoLA-1*02101
</pre></div>
</div>
<p>Here’s an example invocation to fit a predictor:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ mhcflurry-class1-train-allele-specific-models <span class="se">\</span>
    --data curated_training_data.csv.bz2 <span class="se">\</span>
    --hyperparameters hyperparameters.yaml <span class="se">\</span>
    --min-measurements-per-allele <span class="m">75</span> <span class="se">\</span>
    --out-models-dir models
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">hyperparameters.yaml</span></code> file gives the list of neural network architectures
to train models for. Here’s an example specifying a single architecture:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="p p-Indicator">-</span> <span class="nt">activation</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">tanh</span>
  <span class="nt">dense_layer_l1_regularization</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">0.0</span>
  <span class="nt">dropout_probability</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">0.0</span>
  <span class="nt">early_stopping</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">true</span>
  <span class="nt">layer_sizes</span><span class="p">:</span> <span class="p p-Indicator">[</span><span class="nv">8</span><span class="p p-Indicator">]</span>
  <span class="nt">locally_connected_layers</span><span class="p">:</span> <span class="p p-Indicator">[]</span>
  <span class="nt">loss</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">custom:mse_with_inequalities</span>
  <span class="nt">max_epochs</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">500</span>
  <span class="nt">minibatch_size</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">128</span>
  <span class="nt">n_models</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">4</span>
  <span class="nt">output_activation</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">sigmoid</span>
  <span class="nt">patience</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">20</span>
  <span class="nt">peptide_amino_acid_encoding</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">BLOSUM62</span>
  <span class="nt">random_negative_affinity_max</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">50000.0</span>
  <span class="nt">random_negative_affinity_min</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">20000.0</span>
  <span class="nt">random_negative_constant</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">25</span>
  <span class="nt">random_negative_rate</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">0.0</span>
  <span class="nt">validation_split</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">0.1</span>
</pre></div>
</div>
<p>The available hyperparameters for binding predictors are defined in
<a class="reference internal" href="api.html#mhcflurry.Class1NeuralNetwork" title="mhcflurry.Class1NeuralNetwork"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1NeuralNetwork</span></code></a>. To see exactly how
these are used you will need to read the source code.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>MHCflurry predictors are serialized to disk as many files in a directory. The
model training command above will write the models to the output directory specified by the
<code class="docutils literal notranslate"><span class="pre">--out-models-dir</span></code> argument. This directory has files like:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>info.txt
manifest.csv
model_selection.csv.bz2
model_selection_data.csv.bz2
...
weights_PATR-B*24:01-6-da42511a2164a8fe.npz
weights_PATR-B*24:01-7-4e3f5ded9cc1d851.npz
weights_PATR-B*24:01-8-491d1b4d85da0dc4.npz
weights_PATR-B*24:01-9-8f295e814502ffa1.npz
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">manifest.csv</span></code> file gives metadata for all the models used in the predictor.
There will be a <code class="docutils literal notranslate"><span class="pre">weights_...</span></code> file for each model giving its weights
(the parameters for the neural network). The <code class="docutils literal notranslate"><span class="pre">percent_ranks.csv</span></code> stores a
histogram of model predictions for each allele over a large number of random
peptides. It is used for generating the percent ranks at prediction time.</p>
</div>
<p>To fit pan-allele models like the ones released with MHCflurry, you can use
a similar tool, <a class="reference internal" href="commandline_tools.html#mhcflurry-class1-train-pan-allele-models"><span class="std std-ref">mhcflurry-class1-train-pan-allele-models</span></a>. You’ll probably
also want to take a look at the scripts used to generate the production models,
which are available in the <em>downloads-generation</em> directory in the MHCflurry
repository. See the scripts in the <em>models_class1_pan</em> subdirectory to see how the
fitting and model selection was done for models currently distributed with MHCflurry.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The production MHCflurry models were fit using a cluster with several
dozen GPUs over a period of about two days. If you model select over fewer
architectures, however, it should be possible to fit a predictor using less
resources.</p>
</div>
</div>
<div class="section" id="environment-variables">
<h2>Environment variables<a class="headerlink" href="#environment-variables" title="Permalink to this headline">¶</a></h2>
<p>MHCflurry behavior can be modified using these environment variables:</p>
<dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">MHCFLURRY_DEFAULT_CLASS1_MODELS</span></code></dt><dd><p>Path to models directory. If you call <code class="docutils literal notranslate"><span class="pre">Class1AffinityPredictor.load()</span></code>
with no arguments, the models specified in this environment variable will be
used. If this environment variable is undefined, the downloaded models for
the current MHCflurry release are used.</p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">MHCFLURRY_OPTIMIZATION_LEVEL</span></code></dt><dd><p>The pan-allele models can be somewhat slow. As an optimization, when this
variable is greater than 0 (default is 1), we “stitch” the pan-allele models in
the ensemble into one large tensorflow graph. In our experiments
it gives about a 30% speed improvement. It has no effect on allele-specific
models. Set this variable to 0 to disable this behavior. This may be helpful
if you are running out of memory using the pan-allele models.</p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">MHCFLURRY_DEFAULT_PREDICT_BATCH_SIZE</span></code></dt><dd><p>For large prediction tasks, it can be helpful to increase the prediction batch
size, which is set by this environment variable (default is 4096). This
affects both allele-specific and pan-allele predictors. It can have large
effects on performance. Alternatively, if you are running out of memory,
you can try decreasing the batch size.</p>
</dd>
</dl>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="python_tutorial.html" class="btn btn-neutral float-right" title="Python library tutorial" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="intro.html" class="btn btn-neutral float-left" title="Introduction and setup" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright Timothy O&#39;Donnell
      <span class="lastupdated">
        Last updated on Jul 13, 2020.
      </span>

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>