

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Command-line tutorial &mdash; MHCflurry 1.0.0 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="search" title="Search" href="search.html"/>
    <link rel="top" title="MHCflurry 1.0.0 documentation" href="index.html"/>
        <link rel="next" title="Python library tutorial" href="python_tutorial.html"/>
        <link rel="prev" title="Introduction and setup" href="intro.html"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> MHCflurry
          

          
          </a>

          
            
            
              <div class="version">
                1.0.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="intro.html">Introduction and setup</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Command-line tutorial</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#downloading-models">Downloading models</a></li>
<li class="toctree-l2"><a class="reference internal" href="#generating-predictions">Generating predictions</a></li>
<li class="toctree-l2"><a class="reference internal" href="#fitting-your-own-models">Fitting your own models</a></li>
<li class="toctree-l2"><a class="reference internal" href="#scanning-protein-sequences-for-predicted-epitopes">Scanning protein sequences for predicted epitopes</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="python_tutorial.html">Python library tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="models_supported_alleles.html">Supported alleles and peptide lengths</a></li>
<li class="toctree-l1"><a class="reference internal" href="models.html">Details on the released models</a></li>
<li class="toctree-l1"><a class="reference internal" href="commandline_tools.html">Command-line reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="api.html">API Documentation</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">MHCflurry</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Command-line tutorial</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/commandline_tutorial.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="command-line-tutorial">
<span id="commandline-tutorial"></span><h1>Command-line tutorial<a class="headerlink" href="#command-line-tutorial" title="Permalink to this headline">¶</a></h1>
<div class="section" id="downloading-models">
<span id="downloading"></span><h2>Downloading models<a class="headerlink" href="#downloading-models" title="Permalink to this headline">¶</a></h2>
<p>Most users will use pre-trained MHCflurry models that we release. These models
are distributed separately from the pip package and may be downloaded with the
<a class="reference internal" href="commandline_tools.html#mhcflurry-downloads"><span class="std std-ref">mhcflurry-downloads</span></a> tool:</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>$ mhcflurry-downloads fetch models_class1
</pre></div>
</div>
<p>Files downloaded with <a class="reference internal" href="commandline_tools.html#mhcflurry-downloads"><span class="std std-ref">mhcflurry-downloads</span></a> are stored in a platform-specific
directory. To get the path to downloaded data, you can use:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ mhcflurry-downloads path models_class1
/Users/tim/Library/Application Support/mhcflurry/4/1.0.0/models_class1/
</pre></div>
</div>
<p>We also release a few other “downloads,” such as curated training data and some
experimental models. To see what’s available and what you have downloaded, run:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ mhcflurry-downloads info
Environment variables
  MHCFLURRY_DATA_DIR                  [unset or empty]
  MHCFLURRY_DOWNLOADS_CURRENT_RELEASE [unset or empty]
  MHCFLURRY_DOWNLOADS_DIR             [unset or empty]

Configuration
  current release                     = 1.0.0                
  downloads dir                       = &#39;/Users/tim/Library/Application Support/mhcflurry/4/1.0.0&#39; [exists]

DOWNLOAD NAME                             DOWNLOADED?    DEFAULT?      URL                  
models_class1                             YES            YES           http://github.com/hammerlab/mhcflurry/releases/download/pre-1.0/models_class1.tar.bz2 
models_class1_experiments1                NO             NO            http://github.com/hammerlab/mhcflurry/releases/download/pre-1.0/models_class1_experiments1.tar.bz2 
cross_validation_class1                   YES            NO            http://github.com/hammerlab/mhcflurry/releases/download/pre-1.0/cross_validation_class1.tar.bz2 
data_iedb                                 NO             NO            https://github.com/hammerlab/mhcflurry/releases/download/pre-1.0/data_iedb.tar.bz2 
data_kim2014                              NO             NO            http://github.com/hammerlab/mhcflurry/releases/download/0.9.1/data_kim2014.tar.bz2 
data_curated                              YES            YES           https://github.com/hammerlab/mhcflurry/releases/download/pre-1.0/data_curated.tar.bz2
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The code we use for <em>generating</em> the downloads is in the
<code class="docutils literal"><span class="pre">downloads_generation</span></code> directory in the repository.</p>
</div>
</div>
<div class="section" id="generating-predictions">
<h2>Generating predictions<a class="headerlink" href="#generating-predictions" title="Permalink to this headline">¶</a></h2>
<p>The <a class="reference internal" href="commandline_tools.html#mhcflurry-predict"><span class="std std-ref">mhcflurry-predict</span></a> command generates predictions from the command-line.
By default it will use the pre-trained models you downloaded above; other
models can be used by specifying the <code class="docutils literal"><span class="pre">--models</span></code> argument.</p>
<p>Running:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ mhcflurry-predict
    --alleles HLA-A0201 HLA-A0301
    --peptides SIINFEKL SIINFEKD SIINFEKQ
    --out /tmp/predictions.csv
Wrote: /tmp/predictions.csv
</pre></div>
</div>
<p>results in a file like this:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ cat /tmp/predictions.csv
allele,peptide,mhcflurry_prediction,mhcflurry_prediction_low,mhcflurry_prediction_high,mhcflurry_prediction_percentile
HLA-A0201,SIINFEKL,4899.04784343,2767.76365395,7269.68364294,6.5097875
HLA-A0201,SIINFEKD,21050.420243,16834.6585914,24129.0460917,34.297175
HLA-A0201,SIINFEKQ,21048.4726578,16736.5612549,24111.0131144,34.297175
HLA-A0301,SIINFEKL,28227.2989092,24826.3079098,32714.285974,33.9512125
HLA-A0301,SIINFEKD,30816.7212184,27685.5084708,36037.3259046,41.225775
HLA-A0301,SIINFEKQ,24183.0210465,19346.154182,32263.7124753,24.8109625
</pre></div>
</div>
<p>The predictions are given as affinities (KD) in nM in the <code class="docutils literal"><span class="pre">mhcflurry_prediction</span></code>
column. The other fields give the 5-95 percentile predictions across
the models in the ensemble and the quantile of the affinity prediction among
a large number of random peptides tested on that allele.</p>
<p>The predictions shown above were generated with MHCflurry 1.0.0. Different versions of
MHCflurry can give considerably different results. Even
on the same version, exact predictions may vary (up to about 1 nM) depending
on the Keras backend and other details.</p>
<p>In most cases you’ll want to specify the input as a CSV file instead of passing
peptides and alleles as commandline arguments. See <a class="reference internal" href="commandline_tools.html#mhcflurry-predict"><span class="std std-ref">mhcflurry-predict</span></a> docs.</p>
</div>
<div class="section" id="fitting-your-own-models">
<h2>Fitting your own models<a class="headerlink" href="#fitting-your-own-models" title="Permalink to this headline">¶</a></h2>
<p>The <a class="reference internal" href="commandline_tools.html#mhcflurry-class1-train-allele-specific-models"><span class="std std-ref">mhcflurry-class1-train-allele-specific-models</span></a> command is used to
fit models to training data. The models we release with MHCflurry are trained
with a command like:</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>$ mhcflurry-class1-train-allele-specific-models <span class="se">\</span>
    --data TRAINING_DATA.csv <span class="se">\</span>
    --hyperparameters hyperparameters.yaml <span class="se">\</span>
    --percent-rank-calibration-num-peptides-per-length <span class="m">1000000</span> <span class="se">\</span>
    --min-measurements-per-allele <span class="m">75</span> <span class="se">\</span>
    --out-models-dir models
</pre></div>
</div>
<p>MHCflurry predictors are serialized to disk as many files in a directory. The
command above will write the models to the output directory specified by the
<code class="docutils literal"><span class="pre">--out-models-dir</span></code> argument. This directory has files like:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>manifest.csv
percent_ranks.csv
weights_BOLA-6*13:01-0-1e6e7c0610ac68f8.npz
...
weights_PATR-B*24:01-0-e12e0ee723833110.npz
weights_PATR-B*24:01-0-ec4a36529321d868.npz
weights_PATR-B*24:01-0-fd5a340098d3a9f4.npz
</pre></div>
</div>
<p>The <code class="docutils literal"><span class="pre">manifest.csv</span></code> file gives metadata for all the models used in the predictor.
There will be a <code class="docutils literal"><span class="pre">weights_...</span></code> file for each model giving its weights
(the parameters for the neural network). The <code class="docutils literal"><span class="pre">percent_ranks.csv</span></code> stores a
histogram of model predictions for each allele over a large number of random
peptides. It is used for generating the percent ranks at prediction time.</p>
<p>To call <a class="reference internal" href="commandline_tools.html#mhcflurry-class1-train-allele-specific-models"><span class="std std-ref">mhcflurry-class1-train-allele-specific-models</span></a> you’ll need some
training data. The data we use for our released predictors can be downloaded with
<a class="reference internal" href="commandline_tools.html#mhcflurry-downloads"><span class="std std-ref">mhcflurry-downloads</span></a>:</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>$ mhcflurry-downloads fetch data_curated
</pre></div>
</div>
<p>It looks like this:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ bzcat &quot;$(mhcflurry-downloads path data_curated)/curated_training_data.csv.bz2&quot; | head -n 3
allele,peptide,measurement_value,measurement_type,measurement_source,original_allele
BoLA-1*21:01,AENDTLVVSV,7817.0,quantitative,Barlow - purified MHC/competitive/fluorescence,BoLA-1*02101
BoLA-1*21:01,NQFNGGCLLV,1086.0,quantitative,Barlow - purified MHC/direct/fluorescence,BoLA-1*02101
</pre></div>
</div>
</div>
<div class="section" id="scanning-protein-sequences-for-predicted-epitopes">
<h2>Scanning protein sequences for predicted epitopes<a class="headerlink" href="#scanning-protein-sequences-for-predicted-epitopes" title="Permalink to this headline">¶</a></h2>
<p>The <a class="reference external" href="https://github.com/hammerlab/mhctools">mhctools</a> package
provides support for scanning protein sequences to find predicted
epitopes. It supports MHCflurry as well as other binding predictors.
Here is an example.</p>
<p>First, install <code class="docutils literal"><span class="pre">mhctools</span></code> if it is not already installed:</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>$ pip install mhctools
</pre></div>
</div>
<p>We’ll generate predictions across <code class="docutils literal"><span class="pre">example.fasta</span></code>, a FASTA file with two short
sequences:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="o">&gt;</span><span class="n">protein1</span>
<span class="n">MDSKGSSQKGSRLLLLLVVSNLLLCQGVVSTPVCPNGPGNCQV</span>
<span class="n">EMFNEFDKRYAQGKGFITMALNSCHTSSLPTPEDKEQAQQTHH</span>
<span class="o">&gt;</span><span class="n">protein2</span>
<span class="n">VTEVRGMKGAPDAILSRAIEIEEENKRLLEGMEMIFGQVIPGA</span>
<span class="n">ARYSAFYNLLHCLRRDSSKIDTYLKLLNCRIIYNNNC</span>
</pre></div>
</div>
<p>Here’s the <code class="docutils literal"><span class="pre">mhctools</span></code> invocation. See <code class="docutils literal"><span class="pre">mhctools</span> <span class="pre">-h</span></code> for more information.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ mhctools
    --mhc-predictor mhcflurry
    --input-fasta-file example.fasta
    --mhc-alleles A02:01,A03:01
    --mhc-peptide-lengths 8,9,10,11
    --extract-subsequences
    --output-csv /tmp/subsequence_predictions.csv
2017-12-22 01:12:44,974 - mhctools.cli.args - INFO - Building MHC binding prediction type for alleles [&#39;HLA-A*02:01&#39;, &#39;HLA-A*03:01&#39;] and epitope lengths [8, 9, 10, 11]
2017-12-22 01:12:48,868 - mhctools.mhcflurry - INFO - BindingPrediction(peptide=&#39;AARYSAFY&#39;, allele=&#39;HLA-A*03:01&#39;, affinity=5744.3443, percentile_rank=None, source_sequence_name=None, offset=0, prediction_method_name=&#39;mhcflurry&#39;)
...

[1192 rows x 8 columns]
</pre></div>
</div>
<p>This will write a file giving predictions for all subsequences of the specified lengths:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ head -n 3 /tmp/subsequence_predictions.csv
,source_sequence_name,offset,peptide,allele,affinity,percentile_rank,prediction_method_name,length
0,protein2,42,AARYSAFY,HLA-A*03:01,5744.3442744,,mhcflurry,8
1,protein2,42,AARYSAFYN,HLA-A*03:01,10576.5364408,,mhcflurry,9
</pre></div>
</div>
</div>
</div>


           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="python_tutorial.html" class="btn btn-neutral float-right" title="Python library tutorial" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="intro.html" class="btn btn-neutral" title="Introduction and setup" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, Timothy O&#39;Donnell.
      Last updated on Dec 22, 2017.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'1.0.0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>