

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Python library tutorial &mdash; MHCflurry 1.6.0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Command-line reference" href="commandline_tools.html" />
    <link rel="prev" title="Command-line tutorial" href="commandline_tutorial.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> MHCflurry
          

          
          </a>

          
            
            
              <div class="version">
                1.6.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="intro.html">Introduction and setup</a></li>
<li class="toctree-l1"><a class="reference internal" href="commandline_tutorial.html">Command-line tutorial</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Python library tutorial</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#loading-a-predictor">Loading a predictor</a></li>
<li class="toctree-l2"><a class="reference internal" href="#predicting-for-individual-peptides">Predicting for individual peptides</a></li>
<li class="toctree-l2"><a class="reference internal" href="#scanning-protein-sequences">Scanning protein sequences</a></li>
<li class="toctree-l2"><a class="reference internal" href="#lower-level-interfaces">Lower level interfaces</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="commandline_tools.html">Command-line reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="api.html">API Documentation</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">MHCflurry</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Python library tutorial</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/python_tutorial.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="python-library-tutorial">
<h1>Python library tutorial<a class="headerlink" href="#python-library-tutorial" title="Permalink to this headline">¶</a></h1>
<p>The MHCflurry Python API exposes additional options and features beyond those
supported by the commandline tools and can be more convenient for interactive
analyses and bioinformatic pipelines. This tutorial gives a basic overview
of the most important functionality. See the <a class="reference internal" href="api.html#api-documentation"><span class="std std-ref">API Documentation</span></a> for further
details.</p>
<div class="section" id="loading-a-predictor">
<h2>Loading a predictor<a class="headerlink" href="#loading-a-predictor" title="Permalink to this headline">¶</a></h2>
<p>Most prediction tasks can be performed using the
<a class="reference internal" href="api.html#mhcflurry.Class1PresentationPredictor" title="mhcflurry.Class1PresentationPredictor"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1PresentationPredictor</span></code></a> class, which provides a programmatic API
to the functionality in the <a class="reference internal" href="commandline_tools.html#mhcflurry-predict"><span class="std std-ref">mhcflurry-predict</span></a> and
<a class="reference internal" href="commandline_tools.html#mhcflurry-predict-scan"><span class="std std-ref">mhcflurry-predict-scan</span></a> commands.</p>
<p>Instances of <a class="reference internal" href="api.html#mhcflurry.Class1PresentationPredictor" title="mhcflurry.Class1PresentationPredictor"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1PresentationPredictor</span></code></a> wrap a
<a class="reference internal" href="api.html#mhcflurry.Class1AffinityPredictor" title="mhcflurry.Class1AffinityPredictor"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1AffinityPredictor</span></code></a> to generate binding affinity predictions
and a <a class="reference internal" href="api.html#mhcflurry.Class1ProcessingPredictor" title="mhcflurry.Class1ProcessingPredictor"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1ProcessingPredictor</span></code></a> to generate antigen processing
predictions. The presentation score is computed using a logistic regression
model over binding affinity and processing predictions.</p>
<p>Use the <a class="reference internal" href="api.html#mhcflurry.Class1PresentationPredictor.load" title="mhcflurry.Class1PresentationPredictor.load"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load</span></code></a> static method to load a
trained predictor from disk. With no arguments this method will load the predictor
released with MHCflurry (see <a class="reference internal" href="commandline_tutorial.html#downloading"><span class="std std-ref">Downloading models</span></a>). If you pass a path to a
models directory, then it will load that predictor instead.</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">mhcflurry</span> <span class="k">import</span> <span class="n">Class1PresentationPredictor</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">predictor</span> <span class="o">=</span> <span class="n">Class1PresentationPredictor</span><span class="o">.</span><span class="n">load</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">predictor</span><span class="o">.</span><span class="n">supported_alleles</span><span class="p">[:</span><span class="mi">5</span><span class="p">]</span>
<span class="go">[&#39;Atbe-B*01:01&#39;, &#39;Atbe-E*03:01&#39;, &#39;Atbe-G*03:01&#39;, &#39;Atbe-G*03:02&#39;, &#39;Atbe-G*06:01&#39;]</span>
</pre></div>
</div>
</div>
<div class="section" id="predicting-for-individual-peptides">
<h2>Predicting for individual peptides<a class="headerlink" href="#predicting-for-individual-peptides" title="Permalink to this headline">¶</a></h2>
<p>To generate predictions for individual peptides, we can use the
<a class="reference internal" href="api.html#mhcflurry.Class1AffinityPredictor.predict" title="mhcflurry.Class1AffinityPredictor.predict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">predict</span></code></a> method of the <a class="reference internal" href="api.html#mhcflurry.Class1PresentationPredictor" title="mhcflurry.Class1PresentationPredictor"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1PresentationPredictor</span></code></a>,
loaded above. This method returns a <code class="xref py py-obj docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> with binding affinity, processing, and presentation
predictions:</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">predictor</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">peptides</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;SIINFEKL&quot;</span><span class="p">,</span> <span class="s2">&quot;NLVPMVATV&quot;</span><span class="p">],</span>
<span class="gp">... </span>    <span class="n">alleles</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;HLA-A0201&quot;</span><span class="p">,</span> <span class="s2">&quot;HLA-A0301&quot;</span><span class="p">],</span>
<span class="gp">... </span>    <span class="n">verbose</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="go">     peptide  peptide_num sample_name      affinity best_allele  processing_score  presentation_score</span>
<span class="go">0   SIINFEKL            0     sample1  12906.786173   HLA-A0201          0.101473            0.012503</span>
<span class="go">1  NLVPMVATV            1     sample1     15.038358   HLA-A0201          0.676289            0.975463</span>
</pre></div>
</div>
<p>Here, the list of alleles is taken to be an individual’s MHC I genotype (i.e. up
to 6 alleles), and the strongest binder across alleles for each peptide is
reported.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>MHCflurry normalizes allele names using the <a class="reference external" href="https://github.com/openvax/mhcnames">mhcnames</a>
package. Names like <code class="docutils literal notranslate"><span class="pre">HLA-A0201</span></code> or <code class="docutils literal notranslate"><span class="pre">A*02:01</span></code> will be
normalized to <code class="docutils literal notranslate"><span class="pre">HLA-A*02:01</span></code>, so most naming conventions can be used
with methods such as <a class="reference internal" href="api.html#mhcflurry.Class1PresentationPredictor.predict" title="mhcflurry.Class1PresentationPredictor.predict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">predict</span></code></a>.</p>
</div>
<p>If you have multiple sample genotypes, you can pass a dict, where the
keys are arbitrary sample names:</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">predictor</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">peptides</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;KSEYMTSWFY&quot;</span><span class="p">,</span> <span class="s2">&quot;NLVPMVATV&quot;</span><span class="p">],</span>
<span class="gp">... </span>    <span class="n">alleles</span><span class="o">=</span><span class="p">{</span>
<span class="gp">... </span>       <span class="s2">&quot;sample1&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;A0201&quot;</span><span class="p">,</span> <span class="s2">&quot;A0301&quot;</span><span class="p">,</span> <span class="s2">&quot;B0702&quot;</span><span class="p">,</span> <span class="s2">&quot;B4402&quot;</span><span class="p">,</span> <span class="s2">&quot;C0201&quot;</span><span class="p">,</span> <span class="s2">&quot;C0702&quot;</span><span class="p">],</span>
<span class="gp">... </span>       <span class="s2">&quot;sample2&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;A0101&quot;</span><span class="p">,</span> <span class="s2">&quot;A0206&quot;</span><span class="p">,</span> <span class="s2">&quot;B5701&quot;</span><span class="p">,</span> <span class="s2">&quot;C0202&quot;</span><span class="p">],</span>
<span class="gp">... </span>    <span class="p">},</span>
<span class="gp">... </span>    <span class="n">verbose</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="go">      peptide  peptide_num sample_name      affinity best_allele  processing_score  presentation_score</span>
<span class="go">0  KSEYMTSWFY            0     sample1  16737.745268       A0301          0.381632            0.026550</span>
<span class="go">1   NLVPMVATV            1     sample1     15.038358       A0201          0.676289            0.975463</span>
<span class="go">2  KSEYMTSWFY            0     sample2     62.540779       A0101          0.381632            0.796731</span>
<span class="go">3   NLVPMVATV            1     sample2     15.765500       A0206          0.676289            0.974439</span>
</pre></div>
</div>
<p>Here the strongest binder for each sample / peptide pair is returned.</p>
<p>Many users will focus on the binding affinity predictions, as the
processing and presentation predictions are experimental. If you do use the latter
scores, however, when available you should provide the upstream (N-flank)
and downstream (C-flank) sequences from the source proteins of the peptides for
a small boost in accuracy. To do so, specify the <code class="docutils literal notranslate"><span class="pre">n_flank</span></code> and <code class="docutils literal notranslate"><span class="pre">c_flank</span></code>
arguments, which give the flanking sequences for the corresponding peptides:</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">predictor</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">peptides</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;KSEYMTSWFY&quot;</span><span class="p">,</span> <span class="s2">&quot;NLVPMVATV&quot;</span><span class="p">],</span>
<span class="gp">... </span>    <span class="n">n_flanks</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;NNNNNNN&quot;</span><span class="p">,</span> <span class="s2">&quot;SSSSSSSS&quot;</span><span class="p">],</span>
<span class="gp">... </span>    <span class="n">c_flanks</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;CCCCCCCC&quot;</span><span class="p">,</span> <span class="s2">&quot;YYYAAAA&quot;</span><span class="p">],</span>
<span class="gp">... </span>    <span class="n">alleles</span><span class="o">=</span><span class="p">{</span>
<span class="gp">... </span>       <span class="s2">&quot;sample1&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;A0201&quot;</span><span class="p">,</span> <span class="s2">&quot;A0301&quot;</span><span class="p">,</span> <span class="s2">&quot;B0702&quot;</span><span class="p">,</span> <span class="s2">&quot;B4402&quot;</span><span class="p">,</span> <span class="s2">&quot;C0201&quot;</span><span class="p">,</span> <span class="s2">&quot;C0702&quot;</span><span class="p">],</span>
<span class="gp">... </span>       <span class="s2">&quot;sample2&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;A0101&quot;</span><span class="p">,</span> <span class="s2">&quot;A0206&quot;</span><span class="p">,</span> <span class="s2">&quot;B5701&quot;</span><span class="p">,</span> <span class="s2">&quot;C0202&quot;</span><span class="p">],</span>
<span class="gp">... </span>    <span class="p">},</span>
<span class="gp">... </span>    <span class="n">verbose</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="go">      peptide   n_flank   c_flank  peptide_num sample_name      affinity best_allele  processing_score  presentation_score</span>
<span class="go">0  KSEYMTSWFY   NNNNNNN  CCCCCCCC            0     sample1  16737.745268       A0301          0.605816            0.056190</span>
<span class="go">1   NLVPMVATV  SSSSSSSS   YYYAAAA            1     sample1     15.038358       A0201          0.824994            0.986719</span>
<span class="go">2  KSEYMTSWFY   NNNNNNN  CCCCCCCC            0     sample2     62.540779       A0101          0.605816            0.897493</span>
<span class="go">3   NLVPMVATV  SSSSSSSS   YYYAAAA            1     sample2     15.765500       A0206          0.824994            0.986155</span>
</pre></div>
</div>
</div>
<div class="section" id="scanning-protein-sequences">
<h2>Scanning protein sequences<a class="headerlink" href="#scanning-protein-sequences" title="Permalink to this headline">¶</a></h2>
<p>The <a class="reference internal" href="api.html#mhcflurry.Class1PresentationPredictor.predict_sequences" title="mhcflurry.Class1PresentationPredictor.predict_sequences"><code class="xref py py-obj docutils literal notranslate"><span class="pre">predict_sequences</span></code></a> method supports
scanning protein sequences for MHC ligands. Here’s an example to identify all
peptides with a predicted binding affinity of 500 nM or tighter to any allele
across two sample genotypes and two short peptide sequences.</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">predictor</span><span class="o">.</span><span class="n">predict_sequences</span><span class="p">(</span>
<span class="gp">... </span>   <span class="n">sequences</span><span class="o">=</span><span class="p">{</span>
<span class="gp">... </span>       <span class="s1">&#39;protein1&#39;</span><span class="p">:</span> <span class="s2">&quot;MDSKGSSQKGSRLLLLLVVSNLL&quot;</span><span class="p">,</span>
<span class="gp">... </span>       <span class="s1">&#39;protein2&#39;</span><span class="p">:</span> <span class="s2">&quot;SSLPTPEDKEQAQQTHH&quot;</span><span class="p">,</span>
<span class="gp">... </span>   <span class="p">},</span>
<span class="gp">... </span>   <span class="n">alleles</span><span class="o">=</span><span class="p">{</span>
<span class="gp">... </span>       <span class="s2">&quot;sample1&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;A0201&quot;</span><span class="p">,</span> <span class="s2">&quot;A0301&quot;</span><span class="p">,</span> <span class="s2">&quot;B0702&quot;</span><span class="p">],</span>
<span class="gp">... </span>       <span class="s2">&quot;sample2&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;A0101&quot;</span><span class="p">,</span> <span class="s2">&quot;C0202&quot;</span><span class="p">],</span>
<span class="gp">... </span>   <span class="p">},</span>
<span class="gp">... </span>   <span class="n">result</span><span class="o">=</span><span class="s2">&quot;filtered&quot;</span><span class="p">,</span>
<span class="gp">... </span>   <span class="n">comparison_quantity</span><span class="o">=</span><span class="s2">&quot;affinity&quot;</span><span class="p">,</span>
<span class="gp">... </span>   <span class="n">filter_value</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span>
<span class="gp">... </span>   <span class="n">verbose</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="go">  sequence_name  pos     peptide         n_flank     c_flank sample_name    affinity best_allele  affinity_percentile  processing_score  presentation_score</span>
<span class="go">0      protein1   13   LLLLVVSNL   MDSKGSSQKGSRL           L     sample1   38.206225       A0201             0.380125          0.017644            0.571060</span>
<span class="go">1      protein1   14   LLLVVSNLL  MDSKGSSQKGSRLL                 sample1   42.243472       A0201             0.420250          0.090984            0.619213</span>
<span class="go">2      protein1    5   SSQKGSRLL           MDSKG   LLLVVSNLL     sample2   66.749223       C0202             0.803375          0.383608            0.774468</span>
<span class="go">3      protein1    6   SQKGSRLLL          MDSKGS    LLVVSNLL     sample2  178.033467       C0202             1.820000          0.275019            0.482206</span>
<span class="go">4      protein1   13  LLLLVVSNLL   MDSKGSSQKGSRL                 sample1  202.208167       A0201             1.112500          0.058782            0.261320</span>
<span class="go">5      protein1   12  LLLLLVVSNL    MDSKGSSQKGSR           L     sample1  202.506582       A0201             1.112500          0.010025            0.225648</span>
<span class="go">6      protein2    0   SSLPTPEDK                    EQAQQTHH     sample1  335.529377       A0301             1.011750          0.010443            0.156798</span>
<span class="go">7      protein2    0   SSLPTPEDK                    EQAQQTHH     sample2  353.451759       C0202             2.674250          0.010443            0.150753</span>
<span class="go">8      protein1    8   KGSRLLLLL        MDSKGSSQ      VVSNLL     sample2  410.327286       C0202             2.887000          0.121374            0.194081</span>
<span class="go">9      protein1    5    SSQKGSRL           MDSKG  LLLLVVSNLL     sample2  477.285937       C0202             3.107375          0.111982            0.168572</span>
</pre></div>
</div>
<p>When using <code class="docutils literal notranslate"><span class="pre">predict_sequences</span></code>, the flanking sequences for each peptide are
automatically included in the processing and presentation predictions.</p>
<p>See the documentation for <a class="reference internal" href="api.html#mhcflurry.Class1PresentationPredictor" title="mhcflurry.Class1PresentationPredictor"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1PresentationPredictor</span></code></a> for other
useful methods.</p>
</div>
<div class="section" id="lower-level-interfaces">
<h2>Lower level interfaces<a class="headerlink" href="#lower-level-interfaces" title="Permalink to this headline">¶</a></h2>
<p>The <a class="reference internal" href="api.html#mhcflurry.Class1PresentationPredictor" title="mhcflurry.Class1PresentationPredictor"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1PresentationPredictor</span></code></a> delegates to a
<a class="reference internal" href="api.html#mhcflurry.Class1AffinityPredictor" title="mhcflurry.Class1AffinityPredictor"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1AffinityPredictor</span></code></a> instance for binding affinity predictions.
If all you need are binding affinities, you can use this instance directly.</p>
<p>Here’s an example:</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">mhcflurry</span> <span class="k">import</span> <span class="n">Class1AffinityPredictor</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">predictor</span> <span class="o">=</span> <span class="n">Class1AffinityPredictor</span><span class="o">.</span><span class="n">load</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">predictor</span><span class="o">.</span><span class="n">predict_to_dataframe</span><span class="p">(</span><span class="n">allele</span><span class="o">=</span><span class="s2">&quot;HLA-A0201&quot;</span><span class="p">,</span> <span class="n">peptides</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;SIINFEKL&quot;</span><span class="p">,</span> <span class="s2">&quot;SIINFEQL&quot;</span><span class="p">])</span>
<span class="go">    peptide     allele    prediction  prediction_low  prediction_high  prediction_percentile</span>
<span class="go">0  SIINFEKL  HLA-A0201  12906.786173     8829.460289     18029.923061               6.566375</span>
<span class="go">1  SIINFEQL  HLA-A0201  13025.300796     9050.056312     18338.004869               6.623625</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">prediction_low</span></code> and <code class="docutils literal notranslate"><span class="pre">prediction_high</span></code> fields give the 5-95 percentile
predictions across the models in the ensemble. This detailed information is not
available through the higher-level <a class="reference internal" href="api.html#mhcflurry.Class1PresentationPredictor" title="mhcflurry.Class1PresentationPredictor"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1PresentationPredictor</span></code></a>
interface.</p>
<p>Under the hood, <code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1AffinityPredictor</span></code> itself delegates to an ensemble of
of <a class="reference internal" href="api.html#mhcflurry.Class1NeuralNetwork" title="mhcflurry.Class1NeuralNetwork"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1NeuralNetwork</span></code></a> instances, which implement the neural network
models used for prediction. To fit your own affinity prediction models, call
<a class="reference internal" href="api.html#mhcflurry.Class1NeuralNetwork.fit" title="mhcflurry.Class1NeuralNetwork.fit"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fit</span></code></a>.</p>
<p>You can similarly use <a class="reference internal" href="api.html#mhcflurry.Class1ProcessingPredictor" title="mhcflurry.Class1ProcessingPredictor"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1ProcessingPredictor</span></code></a> directly for
antigen processing prediction, and there is a low-level
<a class="reference internal" href="api.html#mhcflurry.Class1ProcessingNeuralNetwork" title="mhcflurry.Class1ProcessingNeuralNetwork"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1ProcessingNeuralNetwork</span></code></a> with a <a class="reference internal" href="api.html#mhcflurry.Class1ProcessingNeuralNetwork.fit" title="mhcflurry.Class1ProcessingNeuralNetwork.fit"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fit</span></code></a> method.</p>
<p>See the API documentation of these classes for details.</p>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="commandline_tools.html" class="btn btn-neutral float-right" title="Command-line reference" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="commandline_tutorial.html" class="btn btn-neutral float-left" title="Command-line tutorial" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright Timothy O&#39;Donnell
      <span class="lastupdated">
        Last updated on Mar 23, 2020.
      </span>

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>