

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>mhcflurry.calibrate_percentile_ranks_command &mdash; MHCflurry 1.6.0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../index.html" class="icon icon-home"> MHCflurry
          

          
          </a>

          
            
            
              <div class="version">
                1.6.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../intro.html">Introduction and setup</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../commandline_tutorial.html">Command-line tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../python_tutorial.html">Python library tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../commandline_tools.html">Command-line reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../api.html">API Documentation</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">MHCflurry</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../index.html">Module code</a> &raquo;</li>
        
      <li>mhcflurry.calibrate_percentile_ranks_command</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for mhcflurry.calibrate_percentile_ranks_command</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Calibrate percentile ranks for models. Runs in-place.</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="kn">import</span> <span class="nn">argparse</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">signal</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">time</span>
<span class="kn">import</span> <span class="nn">traceback</span>
<span class="kn">import</span> <span class="nn">collections</span>
<span class="kn">from</span> <span class="nn">functools</span> <span class="k">import</span> <span class="n">partial</span>

<span class="kn">import</span> <span class="nn">pandas</span>

<span class="kn">from</span> <span class="nn">mhcnames</span> <span class="k">import</span> <span class="n">normalize_allele_name</span>
<span class="kn">import</span> <span class="nn">tqdm</span>  <span class="c1"># progress bar</span>
<span class="n">tqdm</span><span class="o">.</span><span class="n">monitor_interval</span> <span class="o">=</span> <span class="mi">0</span>  <span class="c1"># see https://github.com/tqdm/tqdm/issues/481</span>

<span class="kn">from</span> <span class="nn">.class1_affinity_predictor</span> <span class="k">import</span> <span class="n">Class1AffinityPredictor</span>
<span class="kn">from</span> <span class="nn">.encodable_sequences</span> <span class="k">import</span> <span class="n">EncodableSequences</span>
<span class="kn">from</span> <span class="nn">.common</span> <span class="k">import</span> <span class="n">configure_logging</span><span class="p">,</span> <span class="n">random_peptides</span><span class="p">,</span> <span class="n">amino_acid_distribution</span>
<span class="kn">from</span> <span class="nn">.local_parallelism</span> <span class="k">import</span> <span class="p">(</span>
    <span class="n">add_local_parallelism_args</span><span class="p">,</span>
    <span class="n">worker_pool_with_gpu_assignments_from_args</span><span class="p">,</span>
    <span class="n">call_wrapped_kwargs</span><span class="p">)</span>
<span class="kn">from</span> <span class="nn">.cluster_parallelism</span> <span class="k">import</span> <span class="p">(</span>
    <span class="n">add_cluster_parallelism_args</span><span class="p">,</span>
    <span class="n">cluster_results_from_args</span><span class="p">)</span>


<span class="c1"># To avoid pickling large matrices to send to child processes when running in</span>
<span class="c1"># parallel, we use this global variable as a place to store data. Data that is</span>
<span class="c1"># stored here before creating the thread pool will be inherited to the child</span>
<span class="c1"># processes upon fork() call, allowing us to share large data with the workers</span>
<span class="c1"># via shared memory.</span>
<span class="n">GLOBAL_DATA</span> <span class="o">=</span> <span class="p">{}</span>

<span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span><span class="n">usage</span><span class="o">=</span><span class="vm">__doc__</span><span class="p">)</span>

<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
    <span class="s2">&quot;--models-dir&quot;</span><span class="p">,</span>
    <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;DIR&quot;</span><span class="p">,</span>
    <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Directory to read and write models&quot;</span><span class="p">)</span>
<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
    <span class="s2">&quot;--allele&quot;</span><span class="p">,</span>
    <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">nargs</span><span class="o">=</span><span class="s2">&quot;+&quot;</span><span class="p">,</span>
    <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Alleles to calibrate percentile ranks for. If not specified all &quot;</span>
    <span class="s2">&quot;alleles are used&quot;</span><span class="p">)</span>
<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
    <span class="s2">&quot;--match-amino-acid-distribution-data&quot;</span><span class="p">,</span>
    <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Sample random peptides from the amino acid distribution of the &quot;</span>
    <span class="s2">&quot;peptides listed in the supplied CSV file, which must have a &#39;peptide&#39; &quot;</span>
    <span class="s2">&quot;column. If not specified a uniform distribution is used.&quot;</span><span class="p">)</span>
<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
    <span class="s2">&quot;--num-peptides-per-length&quot;</span><span class="p">,</span>
    <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span>
    <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;N&quot;</span><span class="p">,</span>
    <span class="n">default</span><span class="o">=</span><span class="nb">int</span><span class="p">(</span><span class="mf">1e5</span><span class="p">),</span>
    <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Number of peptides per length to use to calibrate percent ranks. &quot;</span>
    <span class="s2">&quot;Default: </span><span class="si">%(default)s</span><span class="s2">.&quot;</span><span class="p">)</span>
<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
    <span class="s2">&quot;--motif-summary&quot;</span><span class="p">,</span>
    <span class="n">default</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Calculate motifs and length preferences for each allele&quot;</span><span class="p">)</span>
<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
    <span class="s2">&quot;--summary-top-peptide-fraction&quot;</span><span class="p">,</span>
    <span class="n">default</span><span class="o">=</span><span class="p">[</span><span class="mf">0.0001</span><span class="p">,</span> <span class="mf">0.001</span><span class="p">,</span> <span class="mf">0.01</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">],</span>
    <span class="n">nargs</span><span class="o">=</span><span class="s2">&quot;+&quot;</span><span class="p">,</span>
    <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span>
    <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;X&quot;</span><span class="p">,</span>
    <span class="n">help</span><span class="o">=</span><span class="s2">&quot;The top X fraction of predictions (i.e. tightest binders) to use to &quot;</span>
    <span class="s2">&quot;generate motifs and length preferences. Default: </span><span class="si">%(default)s</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
    <span class="s2">&quot;--length-range&quot;</span><span class="p">,</span>
    <span class="n">default</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">15</span><span class="p">),</span>
    <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span>
    <span class="n">nargs</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
    <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Min and max peptide length to calibrate, inclusive. &quot;</span>
    <span class="s2">&quot;Default: </span><span class="si">%(default)s</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
    <span class="s2">&quot;--prediction-batch-size&quot;</span><span class="p">,</span>
    <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span>
    <span class="n">default</span><span class="o">=</span><span class="mi">4096</span><span class="p">,</span>
    <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Keras batch size for predictions&quot;</span><span class="p">)</span>
<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
    <span class="s2">&quot;--alleles-per-work-chunk&quot;</span><span class="p">,</span>
    <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span>
    <span class="n">metavar</span><span class="o">=</span><span class="s2">&quot;N&quot;</span><span class="p">,</span>
    <span class="n">default</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
    <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Number of alleles per work chunk. Default: </span><span class="si">%(default)s</span><span class="s2">.&quot;</span><span class="p">)</span>
<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
    <span class="s2">&quot;--verbosity&quot;</span><span class="p">,</span>
    <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span>
    <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Keras verbosity. Default: </span><span class="si">%(default)s</span><span class="s2">&quot;</span><span class="p">,</span>
    <span class="n">default</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

<span class="n">add_local_parallelism_args</span><span class="p">(</span><span class="n">parser</span><span class="p">)</span>
<span class="n">add_cluster_parallelism_args</span><span class="p">(</span><span class="n">parser</span><span class="p">)</span>


<div class="viewcode-block" id="run"><a class="viewcode-back" href="../../api.html#mhcflurry.calibrate_percentile_ranks_command.run">[docs]</a><span class="k">def</span> <span class="nf">run</span><span class="p">(</span><span class="n">argv</span><span class="o">=</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">:]):</span>
    <span class="k">global</span> <span class="n">GLOBAL_DATA</span>

    <span class="c1"># On sigusr1 print stack trace</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;To show stack trace, run:</span><span class="se">\n</span><span class="s2">kill -s USR1 </span><span class="si">%d</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">os</span><span class="o">.</span><span class="n">getpid</span><span class="p">())</span>
    <span class="n">signal</span><span class="o">.</span><span class="n">signal</span><span class="p">(</span><span class="n">signal</span><span class="o">.</span><span class="n">SIGUSR1</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">sig</span><span class="p">,</span> <span class="n">frame</span><span class="p">:</span> <span class="n">traceback</span><span class="o">.</span><span class="n">print_stack</span><span class="p">())</span>

    <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">(</span><span class="n">argv</span><span class="p">)</span>

    <span class="n">args</span><span class="o">.</span><span class="n">models_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">models_dir</span><span class="p">)</span>

    <span class="n">configure_logging</span><span class="p">(</span><span class="n">verbose</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">verbosity</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">)</span>

    <span class="c1"># It&#39;s important that we don&#39;t trigger a Keras import here since that breaks</span>
    <span class="c1"># local parallelism (tensorflow backend). So we set optimization_level=0.</span>
    <span class="n">predictor</span> <span class="o">=</span> <span class="n">Class1AffinityPredictor</span><span class="o">.</span><span class="n">load</span><span class="p">(</span>
        <span class="n">args</span><span class="o">.</span><span class="n">models_dir</span><span class="p">,</span>
        <span class="n">optimization_level</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">allele</span><span class="p">:</span>
        <span class="n">alleles</span> <span class="o">=</span> <span class="p">[</span><span class="n">normalize_allele_name</span><span class="p">(</span><span class="n">a</span><span class="p">)</span> <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">args</span><span class="o">.</span><span class="n">allele</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">alleles</span> <span class="o">=</span> <span class="n">predictor</span><span class="o">.</span><span class="n">supported_alleles</span>

    <span class="n">allele_set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">alleles</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">predictor</span><span class="o">.</span><span class="n">allele_to_sequence</span><span class="p">:</span>
        <span class="c1"># Remove alleles that have the same sequence.</span>
        <span class="n">new_allele_set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="n">sequence_to_allele</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">set</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">allele</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">allele_set</span><span class="p">):</span>
            <span class="n">sequence_to_allele</span><span class="p">[</span><span class="n">predictor</span><span class="o">.</span><span class="n">allele_to_sequence</span><span class="p">[</span><span class="n">allele</span><span class="p">]]</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">allele</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">equivalent_alleles</span> <span class="ow">in</span> <span class="n">sequence_to_allele</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="n">equivalent_alleles</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">equivalent_alleles</span><span class="p">)</span>
            <span class="n">keep</span> <span class="o">=</span> <span class="n">equivalent_alleles</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
            <span class="n">new_allele_set</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">keep</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span>
            <span class="s2">&quot;Sequence comparison reduced num alleles from&quot;</span><span class="p">,</span>
            <span class="nb">len</span><span class="p">(</span><span class="n">allele_set</span><span class="p">),</span>
            <span class="s2">&quot;to&quot;</span><span class="p">,</span>
            <span class="nb">len</span><span class="p">(</span><span class="n">new_allele_set</span><span class="p">))</span>
        <span class="n">allele_set</span> <span class="o">=</span> <span class="n">new_allele_set</span>

    <span class="n">alleles</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">allele_set</span><span class="p">)</span>

    <span class="n">distribution</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">match_amino_acid_distribution_data</span><span class="p">:</span>
        <span class="n">distribution_peptides</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
            <span class="n">args</span><span class="o">.</span><span class="n">match_amino_acid_distribution_data</span><span class="p">)</span><span class="o">.</span><span class="n">peptide</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
        <span class="n">distribution</span> <span class="o">=</span> <span class="n">amino_acid_distribution</span><span class="p">(</span><span class="n">distribution_peptides</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Using amino acid distribution:&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">distribution</span><span class="p">)</span>

    <span class="n">start</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Percent rank calibration for </span><span class="si">%d</span><span class="s2"> alleles. Generating peptides.&quot;</span> <span class="o">%</span> <span class="p">(</span>
        <span class="nb">len</span><span class="p">(</span><span class="n">alleles</span><span class="p">)))</span>
    <span class="n">peptides</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">lengths</span> <span class="o">=</span> <span class="nb">range</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">length_range</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">args</span><span class="o">.</span><span class="n">length_range</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">length</span> <span class="ow">in</span> <span class="n">lengths</span><span class="p">:</span>
        <span class="n">peptides</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span>
            <span class="n">random_peptides</span><span class="p">(</span>
                <span class="n">args</span><span class="o">.</span><span class="n">num_peptides_per_length</span><span class="p">,</span> <span class="n">length</span><span class="p">,</span> <span class="n">distribution</span><span class="o">=</span><span class="n">distribution</span><span class="p">))</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Done generating peptides in </span><span class="si">%0.2f</span><span class="s2"> sec.&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span> <span class="o">-</span> <span class="n">start</span><span class="p">))</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Encoding </span><span class="si">%d</span><span class="s2"> peptides.&quot;</span> <span class="o">%</span> <span class="nb">len</span><span class="p">(</span><span class="n">peptides</span><span class="p">))</span>
    <span class="n">start</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>

    <span class="n">encoded_peptides</span> <span class="o">=</span> <span class="n">EncodableSequences</span><span class="o">.</span><span class="n">create</span><span class="p">(</span><span class="n">peptides</span><span class="p">)</span>
    <span class="k">del</span> <span class="n">peptides</span>

    <span class="c1"># Now we encode the peptides for each neural network, so the encoding</span>
    <span class="c1"># becomes cached.</span>
    <span class="k">for</span> <span class="n">network</span> <span class="ow">in</span> <span class="n">predictor</span><span class="o">.</span><span class="n">neural_networks</span><span class="p">:</span>
        <span class="n">network</span><span class="o">.</span><span class="n">peptides_to_network_input</span><span class="p">(</span><span class="n">encoded_peptides</span><span class="p">)</span>
    <span class="k">assert</span> <span class="n">encoded_peptides</span><span class="o">.</span><span class="n">encoding_cache</span>  <span class="c1"># must have cached the encoding</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Finished encoding peptides in </span><span class="si">%0.2f</span><span class="s2"> sec.&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span> <span class="o">-</span> <span class="n">start</span><span class="p">))</span>

    <span class="c1"># Store peptides in global variable so they are in shared memory</span>
    <span class="c1"># after fork, instead of needing to be pickled (when doing a parallel run).</span>
    <span class="n">GLOBAL_DATA</span><span class="p">[</span><span class="s2">&quot;calibration_peptides&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">encoded_peptides</span>
    <span class="n">GLOBAL_DATA</span><span class="p">[</span><span class="s2">&quot;predictor&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">predictor</span>
    <span class="n">GLOBAL_DATA</span><span class="p">[</span><span class="s2">&quot;args&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s1">&#39;motif_summary&#39;</span><span class="p">:</span> <span class="n">args</span><span class="o">.</span><span class="n">motif_summary</span><span class="p">,</span>
        <span class="s1">&#39;summary_top_peptide_fractions&#39;</span><span class="p">:</span> <span class="n">args</span><span class="o">.</span><span class="n">summary_top_peptide_fraction</span><span class="p">,</span>
        <span class="s1">&#39;verbose&#39;</span><span class="p">:</span> <span class="n">args</span><span class="o">.</span><span class="n">verbosity</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">,</span>
        <span class="s1">&#39;model_kwargs&#39;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s1">&#39;batch_size&#39;</span><span class="p">:</span> <span class="n">args</span><span class="o">.</span><span class="n">prediction_batch_size</span><span class="p">,</span>
        <span class="p">}</span>
    <span class="p">}</span>
    <span class="k">del</span> <span class="n">encoded_peptides</span>

    <span class="n">serial_run</span> <span class="o">=</span> <span class="ow">not</span> <span class="n">args</span><span class="o">.</span><span class="n">cluster_parallelism</span> <span class="ow">and</span> <span class="n">args</span><span class="o">.</span><span class="n">num_jobs</span> <span class="o">==</span> <span class="mi">0</span>
    <span class="n">worker_pool</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">start</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>

    <span class="n">work_items</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">allele</span> <span class="ow">in</span> <span class="n">alleles</span><span class="p">:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">work_items</span> <span class="ow">or</span> <span class="nb">len</span><span class="p">(</span>
                <span class="n">work_items</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;alleles&#39;</span><span class="p">])</span> <span class="o">&gt;=</span> <span class="n">args</span><span class="o">.</span><span class="n">alleles_per_work_chunk</span><span class="p">:</span>
            <span class="n">work_items</span><span class="o">.</span><span class="n">append</span><span class="p">({</span><span class="s2">&quot;alleles&quot;</span><span class="p">:</span> <span class="p">[]})</span>
        <span class="n">work_items</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;alleles&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">allele</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">serial_run</span><span class="p">:</span>
        <span class="c1"># Serial run</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Running in serial.&quot;</span><span class="p">)</span>
        <span class="n">results</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">do_calibrate_percentile_ranks</span><span class="p">(</span><span class="o">**</span><span class="n">item</span><span class="p">)</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">work_items</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">args</span><span class="o">.</span><span class="n">cluster_parallelism</span><span class="p">:</span>
        <span class="c1"># Run using separate processes HPC cluster.</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Running on cluster.&quot;</span><span class="p">)</span>
        <span class="n">results</span> <span class="o">=</span> <span class="n">cluster_results_from_args</span><span class="p">(</span>
            <span class="n">args</span><span class="p">,</span>
            <span class="n">work_function</span><span class="o">=</span><span class="n">do_calibrate_percentile_ranks</span><span class="p">,</span>
            <span class="n">work_items</span><span class="o">=</span><span class="n">work_items</span><span class="p">,</span>
            <span class="n">constant_data</span><span class="o">=</span><span class="n">GLOBAL_DATA</span><span class="p">,</span>
            <span class="n">result_serialization_method</span><span class="o">=</span><span class="s2">&quot;pickle&quot;</span><span class="p">,</span>
            <span class="n">clear_constant_data</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">worker_pool</span> <span class="o">=</span> <span class="n">worker_pool_with_gpu_assignments_from_args</span><span class="p">(</span><span class="n">args</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Worker pool&quot;</span><span class="p">,</span> <span class="n">worker_pool</span><span class="p">)</span>
        <span class="k">assert</span> <span class="n">worker_pool</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
        <span class="n">results</span> <span class="o">=</span> <span class="n">worker_pool</span><span class="o">.</span><span class="n">imap_unordered</span><span class="p">(</span>
            <span class="n">partial</span><span class="p">(</span><span class="n">call_wrapped_kwargs</span><span class="p">,</span> <span class="n">do_calibrate_percentile_ranks</span><span class="p">),</span>
            <span class="n">work_items</span><span class="p">,</span>
            <span class="n">chunksize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="n">summary_results_lists</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">work_item</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="o">.</span><span class="n">tqdm</span><span class="p">(</span><span class="n">results</span><span class="p">,</span> <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">work_items</span><span class="p">)):</span>
        <span class="k">for</span> <span class="p">(</span><span class="n">transforms</span><span class="p">,</span> <span class="n">summary_results</span><span class="p">)</span> <span class="ow">in</span> <span class="n">work_item</span><span class="p">:</span>
            <span class="n">predictor</span><span class="o">.</span><span class="n">allele_to_percent_rank_transform</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">transforms</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">summary_results</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">for</span> <span class="p">(</span><span class="n">item</span><span class="p">,</span> <span class="n">value</span><span class="p">)</span> <span class="ow">in</span> <span class="n">summary_results</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                    <span class="n">summary_results_lists</span><span class="p">[</span><span class="n">item</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">value</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Done calibrating </span><span class="si">%d</span><span class="s2"> alleles.&quot;</span> <span class="o">%</span> <span class="nb">len</span><span class="p">(</span><span class="n">alleles</span><span class="p">))</span>
    <span class="k">if</span> <span class="n">summary_results_lists</span><span class="p">:</span>
        <span class="k">for</span> <span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">lst</span><span class="p">)</span> <span class="ow">in</span> <span class="n">summary_results_lists</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">lst</span><span class="p">,</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">predictor</span><span class="o">.</span><span class="n">metadata_dataframes</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Including summary result: </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">name</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>

    <span class="n">predictor</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">models_dir</span><span class="p">,</span> <span class="n">model_names_to_write</span><span class="o">=</span><span class="p">[])</span>

    <span class="n">percent_rank_calibration_time</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span> <span class="o">-</span> <span class="n">start</span>

    <span class="k">if</span> <span class="n">worker_pool</span><span class="p">:</span>
        <span class="n">worker_pool</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
        <span class="n">worker_pool</span><span class="o">.</span><span class="n">join</span><span class="p">()</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Percent rank calibration time: </span><span class="si">%0.2f</span><span class="s2"> min.&quot;</span> <span class="o">%</span> <span class="p">(</span>
        <span class="n">percent_rank_calibration_time</span> <span class="o">/</span> <span class="mf">60.0</span><span class="p">))</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Predictor written to: </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">args</span><span class="o">.</span><span class="n">models_dir</span><span class="p">)</span></div>


<div class="viewcode-block" id="do_calibrate_percentile_ranks"><a class="viewcode-back" href="../../api.html#mhcflurry.calibrate_percentile_ranks_command.do_calibrate_percentile_ranks">[docs]</a><span class="k">def</span> <span class="nf">do_calibrate_percentile_ranks</span><span class="p">(</span><span class="n">alleles</span><span class="p">,</span> <span class="n">constant_data</span><span class="o">=</span><span class="n">GLOBAL_DATA</span><span class="p">):</span>
    <span class="n">result_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">allele</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">alleles</span><span class="p">):</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Processing allele&quot;</span><span class="p">,</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;of&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">alleles</span><span class="p">))</span>
        <span class="n">result_item</span> <span class="o">=</span> <span class="n">calibrate_percentile_ranks</span><span class="p">(</span>
            <span class="n">allele</span><span class="p">,</span>
            <span class="n">constant_data</span><span class="p">[</span><span class="s1">&#39;predictor&#39;</span><span class="p">],</span>
            <span class="n">peptides</span><span class="o">=</span><span class="n">constant_data</span><span class="p">[</span><span class="s1">&#39;calibration_peptides&#39;</span><span class="p">],</span>
            <span class="o">**</span><span class="n">constant_data</span><span class="p">[</span><span class="s2">&quot;args&quot;</span><span class="p">])</span>
        <span class="n">result_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">result_item</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">result_list</span></div>


<div class="viewcode-block" id="calibrate_percentile_ranks"><a class="viewcode-back" href="../../api.html#mhcflurry.calibrate_percentile_ranks_command.calibrate_percentile_ranks">[docs]</a><span class="k">def</span> <span class="nf">calibrate_percentile_ranks</span><span class="p">(</span>
        <span class="n">allele</span><span class="p">,</span>
        <span class="n">predictor</span><span class="p">,</span>
        <span class="n">peptides</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">motif_summary</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">summary_top_peptide_fractions</span><span class="o">=</span><span class="p">[</span><span class="mf">0.001</span><span class="p">],</span>
        <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">model_kwargs</span><span class="o">=</span><span class="p">{}):</span>
    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Calibrating&quot;</span><span class="p">,</span> <span class="n">allele</span><span class="p">)</span>
    <span class="n">predictor</span><span class="o">.</span><span class="n">optimize</span><span class="p">()</span>  <span class="c1"># since we loaded with optimization_level=0</span>
    <span class="n">start</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
    <span class="n">summary_results</span> <span class="o">=</span> <span class="n">predictor</span><span class="o">.</span><span class="n">calibrate_percentile_ranks</span><span class="p">(</span>
        <span class="n">peptides</span><span class="o">=</span><span class="n">peptides</span><span class="p">,</span>
        <span class="n">alleles</span><span class="o">=</span><span class="p">[</span><span class="n">allele</span><span class="p">],</span>
        <span class="n">motif_summary</span><span class="o">=</span><span class="n">motif_summary</span><span class="p">,</span>
        <span class="n">summary_top_peptide_fractions</span><span class="o">=</span><span class="n">summary_top_peptide_fractions</span><span class="p">,</span>
        <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span><span class="p">,</span>
        <span class="n">model_kwargs</span><span class="o">=</span><span class="n">model_kwargs</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Done calibrating&quot;</span><span class="p">,</span> <span class="n">allele</span><span class="p">,</span> <span class="s2">&quot;in&quot;</span><span class="p">,</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span> <span class="o">-</span> <span class="n">start</span><span class="p">,</span> <span class="s2">&quot;sec&quot;</span><span class="p">)</span>
    <span class="n">transforms</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">allele</span><span class="p">:</span> <span class="n">predictor</span><span class="o">.</span><span class="n">allele_to_percent_rank_transform</span><span class="p">[</span><span class="n">allele</span><span class="p">],</span>
    <span class="p">}</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">transforms</span><span class="p">,</span> <span class="n">summary_results</span><span class="p">)</span></div>


<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s1">&#39;__main__&#39;</span><span class="p">:</span>
    <span class="n">run</span><span class="p">()</span>
</pre></div>

           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright Timothy O&#39;Donnell
      <span class="lastupdated">
        Last updated on Mar 23, 2020.
      </span>

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>