

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Installation (pip) &mdash; MHCflurry 1.0.0 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="search" title="Search" href="../search.html"/>
    <link rel="top" title="MHCflurry 1.0.0 documentation" href="../index.html"/> 

  
  <script src="../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../index.html" class="icon icon-home"> MHCflurry
          

          
          </a>

          
            
            
              <div class="version">
                1.0.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../intro.html">Introduction and setup</a></li>
<li class="toctree-l1"><a class="reference internal" href="../commandline_tutorial.html">Command-line tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="../python_tutorial.html">Python library tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="../models_supported_alleles.html">Supported alleles and peptide lengths</a></li>
<li class="toctree-l1"><a class="reference internal" href="../models.html">Details on the released models</a></li>
<li class="toctree-l1"><a class="reference internal" href="../commandline_tools.html">Command-line reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html">API Documentation</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">MHCflurry</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
      <li>Installation (pip)</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/package_readme/readme.template.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <p>MHCflurry is an open source package for peptide/MHC I binding affinity prediction. It
provides competitive accuracy with a fast and documented implementation.</p>
<p>You can download pre-trained MHCflurry models fit to affinity measurements
deposited in IEDB or train a MHCflurry predictor on your own data.</p>
<p>Currently only allele-specific prediction is implemented, in which separate models
are trained for each allele. The released models therefore support a fixed set of common
class I alleles for which sufficient published training data is available
(see <a class="reference internal" href="#models-supported-alleles"><span class="std std-ref">Supported alleles and peptide lengths</span></a>).</p>
<p>MHCflurry supports Python versions 2.7 and 3.4+. It uses the <a class="reference external" href="https://keras.io">keras</a>
neural network library via either the Tensorflow or Theano backends. GPUs may
optionally be used for a generally modest speed improvement.</p>
<p>If you find MHCflurry useful in your research please cite:</p>
<blockquote>
<div>O’Donnell, T. et al., 2017. MHCflurry: open-source class I MHC
binding affinity prediction. bioRxiv. Available at:
<a class="reference external" href="http://www.biorxiv.org/content/early/2017/08/09/174243">http://www.biorxiv.org/content/early/2017/08/09/174243</a>.</div></blockquote>
<div class="section" id="installation-pip">
<h1>Installation (pip)<a class="headerlink" href="#installation-pip" title="Permalink to this headline">¶</a></h1>
<p>Install the package:</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>$ pip install mhcflurry
</pre></div>
</div>
<p>Then download our datasets and trained models:</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>$ mhcflurry-downloads fetch
</pre></div>
</div>
<p>From a checkout you can run the unit tests with:</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>$ pip install nose
$ nosetests .
</pre></div>
</div>
</div>
<div class="section" id="using-conda">
<h1>Using conda<a class="headerlink" href="#using-conda" title="Permalink to this headline">¶</a></h1>
<p>You can alternatively get up and running with a <a class="reference external" href="https://conda.io/docs/">conda</a>
environment as follows. Some users have reported that this can avoid problems installing
tensorflow.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>$ conda create -q -n mhcflurry-env <span class="nv">python</span><span class="o">=</span><span class="m">3</span>.6 <span class="s1">&#39;tensorflow&gt;=1.1.2&#39;</span>
$ <span class="nb">source</span> activate mhcflurry-env
</pre></div>
</div>
<p>Then continue as above:</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>$ pip install mhcflurry
$ mhcflurry-downloads fetch
</pre></div>
</div>
<div class="section" id="command-line-tutorial">
<span id="commandline-tutorial"></span><h2>Command-line tutorial<a class="headerlink" href="#command-line-tutorial" title="Permalink to this headline">¶</a></h2>
</div>
</div>
<div class="section" id="downloading-models">
<span id="downloading"></span><h1>Downloading models<a class="headerlink" href="#downloading-models" title="Permalink to this headline">¶</a></h1>
<p>Most users will use pre-trained MHCflurry models that we release. These models
are distributed separately from the pip package and may be downloaded with the
<a class="reference internal" href="../commandline_tools.html#mhcflurry-downloads"><span class="std std-ref">mhcflurry-downloads</span></a> tool:</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>$ mhcflurry-downloads fetch models_class1
</pre></div>
</div>
<p>Files downloaded with <a class="reference internal" href="../commandline_tools.html#mhcflurry-downloads"><span class="std std-ref">mhcflurry-downloads</span></a> are stored in a platform-specific
directory. To get the path to downloaded data, you can use:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ mhcflurry-downloads path models_class1
/Users/tim/Library/Application Support/mhcflurry/4/1.0.0/models_class1/
</pre></div>
</div>
<p>We also release a few other “downloads,” such as curated training data and some
experimental models. To see what’s available and what you have downloaded, run:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ mhcflurry-downloads info
Environment variables
  MHCFLURRY_DATA_DIR                  [unset or empty]
  MHCFLURRY_DOWNLOADS_CURRENT_RELEASE [unset or empty]
  MHCFLURRY_DOWNLOADS_DIR             [unset or empty]

Configuration
  current release                     = 1.0.0                
  downloads dir                       = &#39;/Users/tim/Library/Application Support/mhcflurry/4/1.0.0&#39; [exists]

DOWNLOAD NAME                             DOWNLOADED?    DEFAULT?      URL                  
models_class1                             YES            YES           http://github.com/hammerlab/mhcflurry/releases/download/pre-1.0/models_class1.tar.bz2 
models_class1_experiments1                NO             NO            http://github.com/hammerlab/mhcflurry/releases/download/pre-1.0/models_class1_experiments1.tar.bz2 
cross_validation_class1                   YES            NO            http://github.com/hammerlab/mhcflurry/releases/download/pre-1.0/cross_validation_class1.tar.bz2 
data_iedb                                 NO             NO            https://github.com/hammerlab/mhcflurry/releases/download/pre-1.0/data_iedb.tar.bz2 
data_kim2014                              NO             NO            http://github.com/hammerlab/mhcflurry/releases/download/0.9.1/data_kim2014.tar.bz2 
data_curated                              YES            YES           https://github.com/hammerlab/mhcflurry/releases/download/pre-1.0/data_curated.tar.bz2
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The code we use for <em>generating</em> the downloads is in the
<code class="docutils literal"><span class="pre">downloads_generation</span></code> directory in the repository.</p>
</div>
</div>
<div class="section" id="generating-predictions">
<h1>Generating predictions<a class="headerlink" href="#generating-predictions" title="Permalink to this headline">¶</a></h1>
<p>The <a class="reference internal" href="../commandline_tools.html#mhcflurry-predict"><span class="std std-ref">mhcflurry-predict</span></a> command generates predictions from the command-line.
By default it will use the pre-trained models you downloaded above; other
models can be used by specifying the <code class="docutils literal"><span class="pre">--models</span></code> argument.</p>
<p>Running:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ mhcflurry-predict
    --alleles HLA-A0201 HLA-A0301
    --peptides SIINFEKL SIINFEKD SIINFEKQ
    --out /tmp/predictions.csv
Wrote: /tmp/predictions.csv
</pre></div>
</div>
<p>results in a file like this:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ cat /tmp/predictions.csv
allele,peptide,mhcflurry_prediction,mhcflurry_prediction_low,mhcflurry_prediction_high,mhcflurry_prediction_percentile
HLA-A0201,SIINFEKL,4899.04784343,2767.76365395,7269.68364294,6.5097875
HLA-A0201,SIINFEKD,21050.420243,16834.6585914,24129.0460917,34.297175
HLA-A0201,SIINFEKQ,21048.4726578,16736.5612549,24111.0131144,34.297175
HLA-A0301,SIINFEKL,28227.2989092,24826.3079098,32714.285974,33.9512125
HLA-A0301,SIINFEKD,30816.7212184,27685.5084708,36037.3259046,41.225775
HLA-A0301,SIINFEKQ,24183.0210465,19346.154182,32263.7124753,24.8109625
</pre></div>
</div>
<p>The predictions are given as affinities (KD) in nM in the <code class="docutils literal"><span class="pre">mhcflurry_prediction</span></code>
column. The other fields give the 5-95 percentile predictions across
the models in the ensemble and the quantile of the affinity prediction among
a large number of random peptides tested on that allele.</p>
<p>The predictions shown above were generated with MHCflurry 1.0.0. Different versions of
MHCflurry can give considerably different results. Even
on the same version, exact predictions may vary (up to about 1 nM) depending
on the Keras backend and other details.</p>
<p>In most cases you’ll want to specify the input as a CSV file instead of passing
peptides and alleles as commandline arguments. See <a class="reference internal" href="../commandline_tools.html#mhcflurry-predict"><span class="std std-ref">mhcflurry-predict</span></a> docs.</p>
</div>
<div class="section" id="fitting-your-own-models">
<h1>Fitting your own models<a class="headerlink" href="#fitting-your-own-models" title="Permalink to this headline">¶</a></h1>
<p>The <a class="reference internal" href="../commandline_tools.html#mhcflurry-class1-train-allele-specific-models"><span class="std std-ref">mhcflurry-class1-train-allele-specific-models</span></a> command is used to
fit models to training data. The models we release with MHCflurry are trained
with a command like:</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>$ mhcflurry-class1-train-allele-specific-models <span class="se">\</span>
    --data TRAINING_DATA.csv <span class="se">\</span>
    --hyperparameters hyperparameters.yaml <span class="se">\</span>
    --percent-rank-calibration-num-peptides-per-length <span class="m">1000000</span> <span class="se">\</span>
    --min-measurements-per-allele <span class="m">75</span> <span class="se">\</span>
    --out-models-dir models
</pre></div>
</div>
<p>MHCflurry predictors are serialized to disk as many files in a directory. The
command above will write the models to the output directory specified by the
<code class="docutils literal"><span class="pre">--out-models-dir</span></code> argument. This directory has files like:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>manifest.csv
percent_ranks.csv
weights_BOLA-6*13:01-0-1e6e7c0610ac68f8.npz
...
weights_PATR-B*24:01-0-e12e0ee723833110.npz
weights_PATR-B*24:01-0-ec4a36529321d868.npz
weights_PATR-B*24:01-0-fd5a340098d3a9f4.npz
</pre></div>
</div>
<p>The <code class="docutils literal"><span class="pre">manifest.csv</span></code> file gives metadata for all the models used in the predictor.
There will be a <code class="docutils literal"><span class="pre">weights_...</span></code> file for each model giving its weights
(the parameters for the neural network). The <code class="docutils literal"><span class="pre">percent_ranks.csv</span></code> stores a
histogram of model predictions for each allele over a large number of random
peptides. It is used for generating the percent ranks at prediction time.</p>
<p>To call <a class="reference internal" href="../commandline_tools.html#mhcflurry-class1-train-allele-specific-models"><span class="std std-ref">mhcflurry-class1-train-allele-specific-models</span></a> you’ll need some
training data. The data we use for our released predictors can be downloaded with
<a class="reference internal" href="../commandline_tools.html#mhcflurry-downloads"><span class="std std-ref">mhcflurry-downloads</span></a>:</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>$ mhcflurry-downloads fetch data_curated
</pre></div>
</div>
<p>It looks like this:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ bzcat &quot;$(mhcflurry-downloads path data_curated)/curated_training_data.csv.bz2&quot; | head -n 3
allele,peptide,measurement_value,measurement_type,measurement_source,original_allele
BoLA-1*21:01,AENDTLVVSV,7817.0,quantitative,Barlow - purified MHC/competitive/fluorescence,BoLA-1*02101
BoLA-1*21:01,NQFNGGCLLV,1086.0,quantitative,Barlow - purified MHC/direct/fluorescence,BoLA-1*02101
</pre></div>
</div>
</div>
<div class="section" id="scanning-protein-sequences-for-predicted-epitopes">
<h1>Scanning protein sequences for predicted epitopes<a class="headerlink" href="#scanning-protein-sequences-for-predicted-epitopes" title="Permalink to this headline">¶</a></h1>
<p>The <a class="reference external" href="https://github.com/hammerlab/mhctools">mhctools</a> package
provides support for scanning protein sequences to find predicted
epitopes. It supports MHCflurry as well as other binding predictors.
Here is an example.</p>
<p>First, install <code class="docutils literal"><span class="pre">mhctools</span></code> if it is not already installed:</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>$ pip install mhctools
</pre></div>
</div>
<p>We’ll generate predictions across <code class="docutils literal"><span class="pre">example.fasta</span></code>, a FASTA file with two short
sequences:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="o">&gt;</span><span class="n">protein1</span>
<span class="n">MDSKGSSQKGSRLLLLLVVSNLLLCQGVVSTPVCPNGPGNCQV</span>
<span class="n">EMFNEFDKRYAQGKGFITMALNSCHTSSLPTPEDKEQAQQTHH</span>
<span class="o">&gt;</span><span class="n">protein2</span>
<span class="n">VTEVRGMKGAPDAILSRAIEIEEENKRLLEGMEMIFGQVIPGA</span>
<span class="n">ARYSAFYNLLHCLRRDSSKIDTYLKLLNCRIIYNNNC</span>
</pre></div>
</div>
<p>Here’s the <code class="docutils literal"><span class="pre">mhctools</span></code> invocation. See <code class="docutils literal"><span class="pre">mhctools</span> <span class="pre">-h</span></code> for more information.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ mhctools
    --mhc-predictor mhcflurry
    --input-fasta-file example.fasta
    --mhc-alleles A02:01,A03:01
    --mhc-peptide-lengths 8,9,10,11
    --extract-subsequences
    --output-csv /tmp/subsequence_predictions.csv
2017-12-22 01:01:36,343 - mhctools.cli.args - INFO - Building MHC binding prediction type for alleles [&#39;HLA-A*02:01&#39;, &#39;HLA-A*03:01&#39;] and epitope lengths [8, 9, 10, 11]
2017-12-22 01:01:40,257 - mhctools.mhcflurry - INFO - BindingPrediction(peptide=&#39;AARYSAFY&#39;, allele=&#39;HLA-A*03:01&#39;, affinity=5744.3443, percentile_rank=None, source_sequence_name=None, offset=0, prediction_method_name=&#39;mhcflurry&#39;)
...

[1192 rows x 8 columns]
</pre></div>
</div>
<p>This will write a file giving predictions for all subsequences of the specified lengths:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ head -n 3 /tmp/subsequence_predictions.csv
,source_sequence_name,offset,peptide,allele,affinity,percentile_rank,prediction_method_name,length
0,protein2,42,AARYSAFY,HLA-A*03:01,5744.3442744,,mhcflurry,8
1,protein2,42,AARYSAFYN,HLA-A*03:01,10576.5364408,,mhcflurry,9
</pre></div>
</div>
<div class="section" id="python-library-tutorial">
<h2>Python library tutorial<a class="headerlink" href="#python-library-tutorial" title="Permalink to this headline">¶</a></h2>
</div>
</div>
<div class="section" id="predicting">
<h1>Predicting<a class="headerlink" href="#predicting" title="Permalink to this headline">¶</a></h1>
<p>The MHCflurry Python API exposes additional options and features beyond those
supported by the commandline tools. This tutorial gives a basic overview
of the most important functionality. See the <a class="reference internal" href="../api.html#api-documentation"><span class="std std-ref">API Documentation</span></a> for further details.</p>
<p>The <a class="reference internal" href="../api.html#mhcflurry.Class1AffinityPredictor" title="mhcflurry.Class1AffinityPredictor"><code class="xref py py-obj docutils literal"><span class="pre">Class1AffinityPredictor</span></code></a> class is the primary user-facing interface.
Use the <a class="reference internal" href="../api.html#mhcflurry.Class1AffinityPredictor.load" title="mhcflurry.Class1AffinityPredictor.load"><code class="xref py py-obj docutils literal"><span class="pre">load</span></code></a> static method to load a
trained predictor from disk. With no arguments this method will load the predictor
released with MHCflurry (see <a class="reference internal" href="#downloading"><span class="std std-ref">Downloading models</span></a>). If you pass a path to a
models directory, then it will load that predictor instead.</p>
<div class="highlight-pycon"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">mhcflurry</span> <span class="kn">import</span> <span class="n">Class1AffinityPredictor</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">predictor</span> <span class="o">=</span> <span class="n">Class1AffinityPredictor</span><span class="o">.</span><span class="n">load</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">predictor</span><span class="o">.</span><span class="n">supported_alleles</span><span class="p">[:</span><span class="mi">10</span><span class="p">]</span>
<span class="go">[&#39;BoLA-6*13:01&#39;, &#39;Eqca-1*01:01&#39;, &#39;H-2-Db&#39;, &#39;H-2-Dd&#39;, &#39;H-2-Kb&#39;, &#39;H-2-Kd&#39;, &#39;H-2-Kk&#39;, &#39;H-2-Ld&#39;, &#39;HLA-A*01:01&#39;, &#39;HLA-A*02:01&#39;]</span>
</pre></div>
</div>
<p>With a predictor loaded we can now generate some binding predictions:</p>
<div class="highlight-pycon"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">predictor</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">allele</span><span class="o">=</span><span class="s2">&quot;HLA-A0201&quot;</span><span class="p">,</span> <span class="n">peptides</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;SIINFEKL&quot;</span><span class="p">,</span> <span class="s2">&quot;SIINFEQL&quot;</span><span class="p">])</span>
<span class="go">/Users/tim/miniconda3/envs/py2k/lib/python2.7/site-packages/h5py/__init__.py:34: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 80</span>
<span class="go">  from ._conv import register_converters as _register_converters</span>
<span class="go">/Users/tim/miniconda3/envs/py2k/lib/python2.7/site-packages/h5py/__init__.py:43: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 80</span>
<span class="go">  from . import h5a, h5d, h5ds, h5f, h5fd, h5g, h5r, h5s, h5t, h5p, h5z</span>
<span class="go">/Users/tim/miniconda3/envs/py2k/lib/python2.7/site-packages/h5py/_hl/group.py:21: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 80</span>
<span class="go">  from .. import h5g, h5i, h5o, h5r, h5t, h5l, h5p</span>
<span class="go">Using TensorFlow backend.</span>
<span class="go">array([ 4899.04784343,  5685.25682682])</span>
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">MHCflurry normalizes allele names using the <a class="reference external" href="https://github.com/hammerlab/mhcnames">mhcnames</a>
package. Names like <code class="docutils literal"><span class="pre">HLA-A0201</span></code> or <code class="docutils literal"><span class="pre">A*02:01</span></code> will be
normalized to <code class="docutils literal"><span class="pre">HLA-A*02:01</span></code>, so most naming conventions can be used
with methods such as <a class="reference internal" href="../api.html#mhcflurry.Class1AffinityPredictor.predict" title="mhcflurry.Class1AffinityPredictor.predict"><code class="xref py py-obj docutils literal"><span class="pre">predict</span></code></a>.</p>
</div>
<p>For more detailed results, we can use
<a class="reference internal" href="../api.html#mhcflurry.Class1AffinityPredictor.predict_to_dataframe" title="mhcflurry.Class1AffinityPredictor.predict_to_dataframe"><code class="xref py py-obj docutils literal"><span class="pre">predict_to_dataframe</span></code></a>.</p>
<div class="highlight-pycon"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">predictor</span><span class="o">.</span><span class="n">predict_to_dataframe</span><span class="p">(</span><span class="n">allele</span><span class="o">=</span><span class="s2">&quot;HLA-A0201&quot;</span><span class="p">,</span> <span class="n">peptides</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;SIINFEKL&quot;</span><span class="p">,</span> <span class="s2">&quot;SIINFEQL&quot;</span><span class="p">])</span>
<span class="go">      allele   peptide   prediction  prediction_low  prediction_high  \</span>
<span class="go">0  HLA-A0201  SIINFEKL  4899.047843     2767.763654      7269.683643   </span>
<span class="go">1  HLA-A0201  SIINFEQL  5685.256827     3815.923563      7476.714466   </span>

<span class="go">   prediction_percentile  </span>
<span class="go">0               6.509787  </span>
<span class="go">1               7.436687  </span>
</pre></div>
</div>
<p>Instead of a single allele and multiple peptides, we may need predictions for
allele/peptide pairs. We can predict across pairs by specifying
the <code class="xref py py-obj docutils literal"><span class="pre">alleles</span></code> argument instead of <code class="xref py py-obj docutils literal"><span class="pre">allele</span></code>. The list of alleles
must be the same length as the list of peptides (i.e. it is predicting over pairs,
<em>not</em> taking the cross product).</p>
<div class="highlight-pycon"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">predictor</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">alleles</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;HLA-A0201&quot;</span><span class="p">,</span> <span class="s2">&quot;HLA-B*57:01&quot;</span><span class="p">],</span> <span class="n">peptides</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;SIINFEKL&quot;</span><span class="p">,</span> <span class="s2">&quot;SIINFEQL&quot;</span><span class="p">])</span>
<span class="go">array([  4899.04794216,  26704.22011499])</span>
</pre></div>
</div>
</div>
<div class="section" id="training">
<h1>Training<a class="headerlink" href="#training" title="Permalink to this headline">¶</a></h1>
<p>Let’s fit our own MHCflurry predictor. First we need some training data. If you
haven’t already, run this in a shell to download the MHCflurry training data:</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>$ mhcflurry-downloads fetch data_curated
</pre></div>
</div>
<p>We can get the path to this data from Python using <a class="reference internal" href="../api.html#mhcflurry.downloads.get_path" title="mhcflurry.downloads.get_path"><code class="xref py py-obj docutils literal"><span class="pre">mhcflurry.downloads.get_path</span></code></a>:</p>
<div class="highlight-pycon"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">mhcflurry.downloads</span> <span class="kn">import</span> <span class="n">get_path</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">data_path</span> <span class="o">=</span> <span class="n">get_path</span><span class="p">(</span><span class="s2">&quot;data_curated&quot;</span><span class="p">,</span> <span class="s2">&quot;curated_training_data.csv.bz2&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">data_path</span>
<span class="go">&#39;/Users/tim/Library/Application Support/mhcflurry/4/1.0.0/data_curated/curated_training_data.csv.bz2&#39;</span>
</pre></div>
</div>
<p>Now let’s load it with pandas and filter to reasonably-sized peptides:</p>
<div class="highlight-pycon"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pandas</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">data_path</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[(</span><span class="n">df</span><span class="o">.</span><span class="n">peptide</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">len</span><span class="p">()</span> <span class="o">&gt;=</span> <span class="mi">8</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">peptide</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">len</span><span class="p">()</span> <span class="o">&lt;=</span> <span class="mi">15</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
<span class="go">         allele     peptide  measurement_value measurement_type  \</span>
<span class="go">0  BoLA-1*21:01  AENDTLVVSV             7817.0     quantitative   </span>
<span class="go">1  BoLA-1*21:01  NQFNGGCLLV             1086.0     quantitative   </span>
<span class="go">2  BoLA-2*08:01   AAHCIHAEW               21.0     quantitative   </span>
<span class="go">3  BoLA-2*08:01   AAKHMSNTY             1299.0     quantitative   </span>
<span class="go">4  BoLA-2*08:01  DSYAYMRNGW                2.0     quantitative   </span>

<span class="go">                               measurement_source original_allele  </span>
<span class="go">0  Barlow - purified MHC/competitive/fluorescence    BoLA-1*02101  </span>
<span class="go">1       Barlow - purified MHC/direct/fluorescence    BoLA-1*02101  </span>
<span class="go">2       Barlow - purified MHC/direct/fluorescence    BoLA-2*00801  </span>
<span class="go">3       Barlow - purified MHC/direct/fluorescence    BoLA-2*00801  </span>
<span class="go">4       Barlow - purified MHC/direct/fluorescence    BoLA-2*00801  </span>
</pre></div>
</div>
<p>We’ll make an untrained <a class="reference internal" href="../api.html#mhcflurry.Class1AffinityPredictor" title="mhcflurry.Class1AffinityPredictor"><code class="xref py py-obj docutils literal"><span class="pre">Class1AffinityPredictor</span></code></a> and then call
<a class="reference internal" href="../api.html#mhcflurry.Class1AffinityPredictor.fit_allele_specific_predictors" title="mhcflurry.Class1AffinityPredictor.fit_allele_specific_predictors"><code class="xref py py-obj docutils literal"><span class="pre">fit_allele_specific_predictors</span></code></a> to fit
some models.</p>
<div class="highlight-pycon"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">new_predictor</span> <span class="o">=</span> <span class="n">Class1AffinityPredictor</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">single_allele_train_data</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">df</span><span class="o">.</span><span class="n">allele</span> <span class="o">==</span> <span class="s2">&quot;HLA-B*57:01&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="mi">100</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">new_predictor</span><span class="o">.</span><span class="n">fit_allele_specific_predictors</span><span class="p">(</span>
<span class="gp">... </span>   <span class="n">n_models</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
<span class="gp">... </span>   <span class="n">architecture_hyperparameters</span><span class="o">=</span><span class="p">{</span>
<span class="gp">... </span>        <span class="s2">&quot;layer_sizes&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">16</span><span class="p">],</span>
<span class="gp">... </span>        <span class="s2">&quot;max_epochs&quot;</span><span class="p">:</span> <span class="mi">5</span><span class="p">,</span>
<span class="gp">... </span>        <span class="s2">&quot;random_negative_constant&quot;</span><span class="p">:</span> <span class="mi">5</span><span class="p">,</span>
<span class="gp">... </span>   <span class="p">},</span>
<span class="gp">... </span>   <span class="n">peptides</span><span class="o">=</span><span class="n">single_allele_train_data</span><span class="o">.</span><span class="n">peptide</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
<span class="gp">... </span>   <span class="n">affinities</span><span class="o">=</span><span class="n">single_allele_train_data</span><span class="o">.</span><span class="n">measurement_value</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
<span class="gp">... </span>   <span class="n">allele</span><span class="o">=</span><span class="s2">&quot;HLA-B*57:01&quot;</span><span class="p">)</span>
<span class="go">Train on 112 samples, validate on 28 samples</span>
<span class="go">Epoch 1/1</span>

<span class="go">112/112 [==============================] - 0s 3ms/step - loss: 0.3525 - val_loss: 0.3200</span>
<span class="go">Epoch   0 /   5: loss=0.352491. Min val loss (None) at epoch None</span>
<span class="go">Train on 112 samples, validate on 28 samples</span>
<span class="go">Epoch 1/1</span>

<span class="go">112/112 [==============================] - 0s 38us/step - loss: 0.3264 - val_loss: 0.3042</span>
<span class="go">Train on 112 samples, validate on 28 samples</span>
<span class="go">Epoch 1/1</span>

<span class="go">112/112 [==============================] - 0s 38us/step - loss: 0.3110 - val_loss: 0.2915</span>
<span class="go">Train on 112 samples, validate on 28 samples</span>
<span class="go">Epoch 1/1</span>

<span class="go">112/112 [==============================] - 0s 39us/step - loss: 0.2948 - val_loss: 0.2811</span>
<span class="go">Train on 112 samples, validate on 28 samples</span>
<span class="go">Epoch 1/1</span>

<span class="go">112/112 [==============================] - 0s 36us/step - loss: 0.2817 - val_loss: 0.2721</span>
<span class="go">[&lt;mhcflurry.class1_neural_network.Class1NeuralNetwork object at 0x11e0add10&gt;]</span>
</pre></div>
</div>
<p>The <a class="reference internal" href="../api.html#mhcflurry.Class1AffinityPredictor.fit_allele_specific_predictors" title="mhcflurry.Class1AffinityPredictor.fit_allele_specific_predictors"><code class="xref py py-obj docutils literal"><span class="pre">fit_allele_specific_predictors</span></code></a> method
can be called any number of times on the same instance to build up ensembles
of models across alleles. The <code class="xref py py-obj docutils literal"><span class="pre">architecture_hyperparameters</span></code> we specified are
for demonstration purposes; to fit real models you would usually train for
more epochs.</p>
<p>Now we can generate predictions:</p>
<div class="highlight-pycon"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">new_predictor</span><span class="o">.</span><span class="n">predict</span><span class="p">([</span><span class="s2">&quot;SYNPEPII&quot;</span><span class="p">],</span> <span class="n">allele</span><span class="o">=</span><span class="s2">&quot;HLA-B*57:01&quot;</span><span class="p">)</span>
<span class="go">array([ 904.97690009])</span>
</pre></div>
</div>
<p>We can save our predictor to the specified directory on disk by running:</p>
<div class="highlight-pycon"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">new_predictor</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="s2">&quot;/tmp/new-predictor&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>and restore it:</p>
<div class="highlight-pycon"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">new_predictor2</span> <span class="o">=</span> <span class="n">Class1AffinityPredictor</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;/tmp/new-predictor&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">new_predictor2</span><span class="o">.</span><span class="n">supported_alleles</span>
<span class="go">[&#39;HLA-B*57:01&#39;]</span>
</pre></div>
</div>
</div>
<div class="section" id="lower-level-interface">
<h1>Lower level interface<a class="headerlink" href="#lower-level-interface" title="Permalink to this headline">¶</a></h1>
<p>The high-level <code class="xref py py-obj docutils literal"><span class="pre">Class1AffinityPredictor</span></code> delegates to low-level
<a class="reference internal" href="../api.html#mhcflurry.Class1NeuralNetwork" title="mhcflurry.Class1NeuralNetwork"><code class="xref py py-obj docutils literal"><span class="pre">Class1NeuralNetwork</span></code></a> objects, each of which represents
a single neural network. The purpose of <a class="reference internal" href="../api.html#mhcflurry.Class1AffinityPredictor" title="mhcflurry.Class1AffinityPredictor"><code class="xref py py-obj docutils literal"><span class="pre">Class1AffinityPredictor</span></code></a>
is to implement several important features:</p>
<dl class="docutils">
<dt>ensembles</dt>
<dd>More than one neural network can be used to generate each prediction. The
predictions returned to the user are the geometric mean of the individual
model predictions. This gives higher accuracy in most situations</dd>
<dt>multiple alleles</dt>
<dd>A <a class="reference internal" href="../api.html#mhcflurry.Class1NeuralNetwork" title="mhcflurry.Class1NeuralNetwork"><code class="xref py py-obj docutils literal"><span class="pre">Class1NeuralNetwork</span></code></a> generates predictions for only a single
allele. The <a class="reference internal" href="../api.html#mhcflurry.Class1AffinityPredictor" title="mhcflurry.Class1AffinityPredictor"><code class="xref py py-obj docutils literal"><span class="pre">Class1AffinityPredictor</span></code></a> maps alleles to the
relevant <a class="reference internal" href="../api.html#mhcflurry.Class1NeuralNetwork" title="mhcflurry.Class1NeuralNetwork"><code class="xref py py-obj docutils literal"><span class="pre">Class1NeuralNetwork</span></code></a> instances</dd>
<dt>serialization</dt>
<dd>Loading and saving predictors is implemented in <a class="reference internal" href="../api.html#mhcflurry.Class1AffinityPredictor" title="mhcflurry.Class1AffinityPredictor"><code class="xref py py-obj docutils literal"><span class="pre">Class1AffinityPredictor</span></code></a>.</dd>
</dl>
<p>Sometimes it’s easiest to work directly with <a class="reference internal" href="../api.html#mhcflurry.Class1NeuralNetwork" title="mhcflurry.Class1NeuralNetwork"><code class="xref py py-obj docutils literal"><span class="pre">Class1NeuralNetwork</span></code></a>.
Here is a simple example of doing so:</p>
<div class="highlight-pycon"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">mhcflurry</span> <span class="kn">import</span> <span class="n">Class1NeuralNetwork</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">network</span> <span class="o">=</span> <span class="n">Class1NeuralNetwork</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">network</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span>
<span class="gp">... </span>   <span class="n">single_allele_train_data</span><span class="o">.</span><span class="n">peptide</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
<span class="gp">... </span>   <span class="n">single_allele_train_data</span><span class="o">.</span><span class="n">measurement_value</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
<span class="gp">... </span>   <span class="n">verbose</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="go">Epoch   0 / 500: loss=0.53638. Min val loss (None) at epoch None</span>
<span class="go">Early stopping at epoch 169 / 500: loss=0.00989471. Min val loss (0.0899693146348) at epoch 158</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">network</span><span class="o">.</span><span class="n">predict</span><span class="p">([</span><span class="s2">&quot;SIINFEKLL&quot;</span><span class="p">])</span>
<span class="go">array([ 19371.89996653])</span>
</pre></div>
</div>
<div class="section" id="supported-alleles-and-peptide-lengths">
<span id="models-supported-alleles"></span><h2>Supported alleles and peptide lengths<a class="headerlink" href="#supported-alleles-and-peptide-lengths" title="Permalink to this headline">¶</a></h2>
<p>Models released with the current version of MHCflurry (1.0.0) support peptides of length 8-15 and the following 124 alleles:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">BoLA</span><span class="o">-</span><span class="mi">6</span><span class="o">*</span><span class="mi">13</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">Eqca</span><span class="o">-</span><span class="mi">1</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">H</span><span class="o">-</span><span class="mi">2</span><span class="o">-</span><span class="n">Db</span><span class="p">,</span> <span class="n">H</span><span class="o">-</span><span class="mi">2</span><span class="o">-</span><span class="n">Dd</span><span class="p">,</span> <span class="n">H</span><span class="o">-</span><span class="mi">2</span><span class="o">-</span><span class="n">Kb</span><span class="p">,</span> <span class="n">H</span><span class="o">-</span><span class="mi">2</span><span class="o">-</span><span class="n">Kd</span><span class="p">,</span> <span class="n">H</span><span class="o">-</span><span class="mi">2</span><span class="o">-</span><span class="n">Kk</span><span class="p">,</span>
<span class="n">H</span><span class="o">-</span><span class="mi">2</span><span class="o">-</span><span class="n">Ld</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">02</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">02</span><span class="p">:</span><span class="mi">02</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">02</span><span class="p">:</span><span class="mi">03</span><span class="p">,</span>
<span class="n">HLA</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">02</span><span class="p">:</span><span class="mi">05</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">02</span><span class="p">:</span><span class="mi">06</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">02</span><span class="p">:</span><span class="mi">07</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">02</span><span class="p">:</span><span class="mi">11</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">02</span><span class="p">:</span><span class="mi">12</span><span class="p">,</span>
<span class="n">HLA</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">02</span><span class="p">:</span><span class="mi">16</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">02</span><span class="p">:</span><span class="mi">17</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">02</span><span class="p">:</span><span class="mi">19</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">02</span><span class="p">:</span><span class="mi">50</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">03</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span>
<span class="n">HLA</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">11</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">23</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">24</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">24</span><span class="p">:</span><span class="mi">02</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">24</span><span class="p">:</span><span class="mi">03</span><span class="p">,</span>
<span class="n">HLA</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">25</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">26</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">26</span><span class="p">:</span><span class="mi">02</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">26</span><span class="p">:</span><span class="mi">03</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">29</span><span class="p">:</span><span class="mi">02</span><span class="p">,</span>
<span class="n">HLA</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">30</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">30</span><span class="p">:</span><span class="mi">02</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">31</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">32</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">32</span><span class="p">:</span><span class="mi">07</span><span class="p">,</span>
<span class="n">HLA</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">33</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">66</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">68</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">68</span><span class="p">:</span><span class="mi">02</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">68</span><span class="p">:</span><span class="mi">23</span><span class="p">,</span>
<span class="n">HLA</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">69</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">80</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">07</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">07</span><span class="p">:</span><span class="mi">02</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">08</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span>
<span class="n">HLA</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">08</span><span class="p">:</span><span class="mi">02</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">08</span><span class="p">:</span><span class="mi">03</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">14</span><span class="p">:</span><span class="mi">02</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">15</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">15</span><span class="p">:</span><span class="mi">02</span><span class="p">,</span>
<span class="n">HLA</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">15</span><span class="p">:</span><span class="mi">03</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">15</span><span class="p">:</span><span class="mi">09</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">15</span><span class="p">:</span><span class="mi">17</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">15</span><span class="p">:</span><span class="mi">42</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">18</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span>
<span class="n">HLA</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">27</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">27</span><span class="p">:</span><span class="mi">03</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">27</span><span class="p">:</span><span class="mi">04</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">27</span><span class="p">:</span><span class="mi">05</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">27</span><span class="p">:</span><span class="mi">06</span><span class="p">,</span>
<span class="n">HLA</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">27</span><span class="p">:</span><span class="mi">20</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">35</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">35</span><span class="p">:</span><span class="mi">03</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">35</span><span class="p">:</span><span class="mi">08</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">37</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span>
<span class="n">HLA</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">38</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">39</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">40</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">40</span><span class="p">:</span><span class="mi">02</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">42</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span>
<span class="n">HLA</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">44</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">44</span><span class="p">:</span><span class="mi">02</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">44</span><span class="p">:</span><span class="mi">03</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">45</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">45</span><span class="p">:</span><span class="mi">06</span><span class="p">,</span>
<span class="n">HLA</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">46</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">48</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">51</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">53</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">54</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span>
<span class="n">HLA</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">57</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">58</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">73</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">83</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">C</span><span class="o">*</span><span class="mi">03</span><span class="p">:</span><span class="mi">03</span><span class="p">,</span>
<span class="n">HLA</span><span class="o">-</span><span class="n">C</span><span class="o">*</span><span class="mi">03</span><span class="p">:</span><span class="mi">04</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">C</span><span class="o">*</span><span class="mi">04</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">C</span><span class="o">*</span><span class="mi">05</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">C</span><span class="o">*</span><span class="mi">06</span><span class="p">:</span><span class="mi">02</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">C</span><span class="o">*</span><span class="mi">07</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span>
<span class="n">HLA</span><span class="o">-</span><span class="n">C</span><span class="o">*</span><span class="mi">07</span><span class="p">:</span><span class="mi">02</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">C</span><span class="o">*</span><span class="mi">08</span><span class="p">:</span><span class="mi">02</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">C</span><span class="o">*</span><span class="mi">12</span><span class="p">:</span><span class="mi">03</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">C</span><span class="o">*</span><span class="mi">14</span><span class="p">:</span><span class="mi">02</span><span class="p">,</span> <span class="n">HLA</span><span class="o">-</span><span class="n">C</span><span class="o">*</span><span class="mi">15</span><span class="p">:</span><span class="mi">02</span><span class="p">,</span>
<span class="n">Mamu</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">Mamu</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">02</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">Mamu</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">02</span><span class="p">:</span><span class="mi">0102</span><span class="p">,</span> <span class="n">Mamu</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">07</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span>
<span class="n">Mamu</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">07</span><span class="p">:</span><span class="mi">0103</span><span class="p">,</span> <span class="n">Mamu</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">11</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">Mamu</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">22</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">Mamu</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">26</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span>
<span class="n">Mamu</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">Mamu</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">03</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">Mamu</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">08</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">Mamu</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">10</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">Mamu</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">17</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span>
<span class="n">Mamu</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">17</span><span class="p">:</span><span class="mi">04</span><span class="p">,</span> <span class="n">Mamu</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">39</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">Mamu</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">52</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">Mamu</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">66</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">Mamu</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">83</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span>
<span class="n">Mamu</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">87</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">Patr</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">Patr</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">03</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">Patr</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">04</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">Patr</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">07</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span>
<span class="n">Patr</span><span class="o">-</span><span class="n">A</span><span class="o">*</span><span class="mi">09</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">Patr</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">Patr</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">13</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span> <span class="n">Patr</span><span class="o">-</span><span class="n">B</span><span class="o">*</span><span class="mi">24</span><span class="p">:</span><span class="mi">01</span>
</pre></div>
</div>
</div>
<div class="section" id="mhcflurry">
<h2>mhcflurry<a class="headerlink" href="#mhcflurry" title="Permalink to this headline">¶</a></h2>
<p>Open source neural network models for peptide-MHC binding affinity
prediction</p>
<p>The <a class="reference external" href="https://en.wikipedia.org/wiki/Adaptive_immune_system">adaptive immune
system</a> depends
on the presentation of protein fragments by
<a class="reference external" href="https://en.wikipedia.org/wiki/Major_histocompatibility_complex">MHC</a>
molecules. Machine learning models of this interaction are used in
studies of infectious diseases, autoimmune diseases, vaccine
development, and cancer immunotherapy.</p>
<p>MHCflurry supports Class I peptide/MHC binding affinity prediction using
ensembles of allele-specific models. You can fit MHCflurry models to
your own data or download models that we fit to data from
<a class="reference external" href="http://www.iedb.org/home_v3.php">IEDB</a> and <a class="reference external" href="http://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-15-241">Kim
2014</a>.
Our combined dataset is available for download
<a class="reference external" href="https://github.com/hammerlab/mhcflurry/releases/download/pre-1.0.0-alpha/data_curated.tar.bz2">here</a>.</p>
<p>Pan-allelic prediction is supported in principle but is not yet
performing accurately. Infrastructure for modeling other aspects of
antigen processing is also implemented but experimental.</p>
<p>If you find MHCflurry useful in your research please cite:</p>
<blockquote>
<div>O’Donnell, T. et al., 2017. MHCflurry: open-source class I MHC
binding affinity prediction. bioRxiv. Available at:
<a class="reference external" href="http://www.biorxiv.org/content/early/2017/08/09/174243">http://www.biorxiv.org/content/early/2017/08/09/174243</a>.</div></blockquote>
</div>
</div>
<div class="section" id="setup-pip">
<h1>Setup (pip)<a class="headerlink" href="#setup-pip" title="Permalink to this headline">¶</a></h1>
<p>Install the package:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">pip</span> <span class="n">install</span> <span class="n">mhcflurry</span>
</pre></div>
</div>
<p>Then download our datasets and trained models:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">mhcflurry</span><span class="o">-</span><span class="n">downloads</span> <span class="n">fetch</span>
</pre></div>
</div>
<p>From a checkout you can run the unit tests with:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">nosetests</span> <span class="o">.</span>
</pre></div>
</div>
<p>The MHCflurry predictors are implemented in Python using
<a class="reference external" href="https://keras.io">keras</a>.</p>
<p>MHCflurry works with both the tensorflow and theano keras backends. The
tensorflow backend gives faster model-loading time but is undergoing
more rapid development and sometimes hits issues. If you encounter
tensorflow errors running MHCflurry, try setting this environment
variable to switch to the theano backend:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">export</span> <span class="n">KERAS_BACKEND</span><span class="o">=</span><span class="n">theano</span>
</pre></div>
</div>
<p>You may also needs to <code class="docutils literal"><span class="pre">pip</span> <span class="pre">install</span> <span class="pre">theano</span></code>.</p>
</div>
<div class="section" id="setup-conda">
<h1>Setup (conda)<a class="headerlink" href="#setup-conda" title="Permalink to this headline">¶</a></h1>
<p>You can alternatively get up and running with a
<a class="reference external" href="https://conda.io/docs/">conda</a> environment as follows:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">create</span> <span class="o">-</span><span class="n">q</span> <span class="o">-</span><span class="n">n</span> <span class="n">mhcflurry</span><span class="o">-</span><span class="n">env</span> <span class="n">python</span><span class="o">=</span><span class="mf">3.6</span> <span class="s1">&#39;tensorflow&gt;=1.1.0&#39;</span>
<span class="n">source</span> <span class="n">activate</span> <span class="n">mhcflurry</span><span class="o">-</span><span class="n">env</span>
</pre></div>
</div>
<p>Then continue as above:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">pip</span> <span class="n">install</span> <span class="n">mhcflurry</span>
<span class="n">mhcflurry</span><span class="o">-</span><span class="n">downloads</span> <span class="n">fetch</span>
</pre></div>
</div>
<p>If you wish to test your installation, you can install <code class="docutils literal"><span class="pre">nose</span></code> and run
the tests from a checkout:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">pip</span> <span class="n">install</span> <span class="n">nose</span>
<span class="n">nosetests</span> <span class="o">.</span>
</pre></div>
</div>
</div>
<div class="section" id="making-predictions-from-the-command-line">
<h1>Making predictions from the command-line<a class="headerlink" href="#making-predictions-from-the-command-line" title="Permalink to this headline">¶</a></h1>
<div class="code shell highlight-default"><div class="highlight"><pre><span></span>$ mhcflurry-predict --alleles HLA-A0201 HLA-A0301 --peptides SIINFEKL SIINFEKD SIINFEKQ
allele,peptide,mhcflurry_prediction,mhcflurry_prediction_low,mhcflurry_prediction_high
HLA-A0201,SIINFEKL,5326.541919062165,3757.86675352994,7461.37693353508
HLA-A0201,SIINFEKD,18763.70298522213,13140.82000240037,23269.82139560844
HLA-A0201,SIINFEKQ,18620.10057358322,13096.425874678192,23223.148184869413
HLA-A0301,SIINFEKL,24481.726678691946,21035.52779725433,27245.371837497867
HLA-A0301,SIINFEKD,24687.529360239587,21582.590014592537,27749.39869616437
HLA-A0301,SIINFEKQ,25923.062203902562,23522.5793450799,28079.456657427705
</pre></div>
</div>
<p>The predictions returned are affinities (KD) in nM. The
<code class="docutils literal"><span class="pre">prediction_low</span></code> and <code class="docutils literal"><span class="pre">prediction_high</span></code> fields give the 5-95
percentile predictions across the models in the ensemble. The
predictions above were generated with MHCflurry 0.9.2. Your exact
predictions may vary slightly from these (up to about 1 nM) depending on
the Keras backend in use and other numerical details. Different versions
of MHCflurry can of course give results considerably different from
these.</p>
<p>You can also specify the input and output as CSV files. Run
<code class="docutils literal"><span class="pre">mhcflurry-predict</span> <span class="pre">-h</span></code> for details.</p>
</div>
<div class="section" id="making-predictions-from-python">
<h1>Making predictions from Python<a class="headerlink" href="#making-predictions-from-python" title="Permalink to this headline">¶</a></h1>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">mhcflurry</span> <span class="k">import</span> <span class="n">Class1AffinityPredictor</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">predictor</span> <span class="o">=</span> <span class="n">Class1AffinityPredictor</span><span class="o">.</span><span class="n">load</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">predictor</span><span class="o">.</span><span class="n">predict_to_dataframe</span><span class="p">(</span><span class="n">peptides</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;SIINFEKL&#39;</span><span class="p">],</span> <span class="n">allele</span><span class="o">=</span><span class="s1">&#39;A0201&#39;</span><span class="p">)</span>


<span class="go">  allele   peptide   prediction  prediction_low  prediction_high</span>
<span class="go">  A0201  SIINFEKL  6029.084473     4474.103253      7771.297702</span>
</pre></div>
</div>
<p>See the
<a class="reference external" href="https://github.com/hammerlab/mhcflurry/blob/master/examples/class1_allele_specific_models.ipynb">class1_allele_specific_models.ipynb</a>
notebook for an overview of the Python API, including fitting your own
predictors.</p>
</div>
<div class="section" id="id1">
<h1>Scanning protein sequences for predicted epitopes<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h1>
<p>The <a class="reference external" href="https://github.com/hammerlab/mhctools">mhctools</a> package
provides support for scanning protein sequences to find predicted
epitopes. It supports MHCflurry as well as other binding predictors.
Here is an example:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="c1"># First install mhctools if needed:</span>
<span class="n">pip</span> <span class="n">install</span> <span class="n">mhctools</span>

<span class="c1"># Now generate predictions for protein sequences in FASTA format:</span>
<span class="n">mhctools</span> \
    <span class="o">--</span><span class="n">mhc</span><span class="o">-</span><span class="n">predictor</span> <span class="n">mhcflurry</span> \
    <span class="o">--</span><span class="nb">input</span><span class="o">-</span><span class="n">fasta</span><span class="o">-</span><span class="n">file</span> <span class="n">INPUT</span><span class="o">.</span><span class="n">fasta</span> \
    <span class="o">--</span><span class="n">mhc</span><span class="o">-</span><span class="n">alleles</span> <span class="n">A02</span><span class="p">:</span><span class="mi">01</span><span class="p">,</span><span class="n">A03</span><span class="p">:</span><span class="mi">01</span> \
    <span class="o">--</span><span class="n">mhc</span><span class="o">-</span><span class="n">peptide</span><span class="o">-</span><span class="n">lengths</span> <span class="mi">8</span><span class="p">,</span><span class="mi">9</span><span class="p">,</span><span class="mi">10</span><span class="p">,</span><span class="mi">11</span> \
    <span class="o">--</span><span class="n">extract</span><span class="o">-</span><span class="n">subsequences</span> \
    <span class="o">--</span><span class="n">out</span> <span class="n">RESULT</span><span class="o">.</span><span class="n">csv</span>
</pre></div>
</div>
</div>
<div class="section" id="details-on-the-downloadable-models">
<h1>Details on the downloadable models<a class="headerlink" href="#details-on-the-downloadable-models" title="Permalink to this headline">¶</a></h1>
</div>
<div class="section" id="environment-variables">
<h1>Environment variables<a class="headerlink" href="#environment-variables" title="Permalink to this headline">¶</a></h1>
<p>The path where MHCflurry looks for model weights and data can be set
with the <code class="docutils literal"><span class="pre">MHCFLURRY_DOWNLOADS_DIR</span></code> environment variable. This
directory should contain subdirectories like “models_class1”.</p>
</div>


           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, Timothy O&#39;Donnell.
      Last updated on Dec 22, 2017.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../',
            VERSION:'1.0.0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../_static/jquery.js"></script>
      <script type="text/javascript" src="../_static/underscore.js"></script>
      <script type="text/javascript" src="../_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>